{"cells":[{"cell_type":"markdown","source":["![JohnSnowLabs](https://nlp.johnsnowlabs.com/assets/images/logo.png)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"aec851fe-e39a-4743-b983-9930d110e2c6"}}},{"cell_type":"markdown","source":["# 2. Clinical Assertion Status Model v2.7.0"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"8e5732f0-beec-44b9-b4e5-42023927f9d4"}}},{"cell_type":"code","source":["import os\nimport json\nimport string\nimport numpy as np\nimport pandas as pd\n\n\nimport sparknlp\nimport sparknlp_jsl\nfrom sparknlp.base import *\nfrom sparknlp.util import *\nfrom sparknlp.annotator import *\nfrom sparknlp_jsl.annotator import *\nfrom sparknlp.pretrained import ResourceDownloader\n\nfrom pyspark.sql import functions as F\nfrom pyspark.ml import Pipeline, PipelineModel\n\npd.set_option('max_colwidth', 100)\npd.set_option('display.max_columns', 100)  \npd.set_option('display.expand_frame_repr', False)\n\n\nprint('sparknlp_jsl.version : ',sparknlp_jsl.version())\n\nspark"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"c0f3e14c-398a-4ecd-864c-3b1c939acafd"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">sparknlp version:  2.6.3\n</div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">sparknlp version:  2.6.3\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["spark"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"d7da6149-ee0a-46fd-9f7b-91e784b2e4f4"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"\n            <div>\n                <p><b>SparkSession - hive</b></p>\n                \n        <div>\n            <p><b>SparkContext</b></p>\n\n            <p><a href=\"http://10.164.245.220:43034\">Spark UI</a></p>\n\n            <dl>\n              <dt>Version</dt>\n                <dd><code>v2.4.5</code></dd>\n              <dt>Master</dt>\n                <dd><code>spark://10.164.245.220:7077</code></dd>\n              <dt>AppName</dt>\n                <dd><code>Databricks Shell</code></dd>\n            </dl>\n        </div>\n        \n            </div>\n        ","textData":"<div class=\"ansiout\">Out[2]: </div>","removedWidgets":[],"addedWidgets":{},"type":"htmlSandbox","arguments":{}}},"output_type":"display_data","data":{"text/html":["\n            <div>\n                <p><b>SparkSession - hive</b></p>\n                \n        <div>\n            <p><b>SparkContext</b></p>\n\n            <p><a href=\"http://10.164.245.220:43034\">Spark UI</a></p>\n\n            <dl>\n              <dt>Version</dt>\n                <dd><code>v2.4.5</code></dd>\n              <dt>Master</dt>\n                <dd><code>spark://10.164.245.220:7077</code></dd>\n              <dt>AppName</dt>\n                <dd><code>Databricks Shell</code></dd>\n            </dl>\n        </div>\n        \n            </div>\n        "]}}],"execution_count":0},{"cell_type":"markdown","source":["# Clinical Assertion Model (with pretrained models)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"7868685d-d1a9-4896-8527-45da3ec252e7"}}},{"cell_type":"code","source":["# Annotator that transforms a text column from dataframe into an Annotation ready for NLP\n\nfrom sparknlp_jsl.annotator import *\n\ndocumentAssembler = DocumentAssembler()\\\n  .setInputCol(\"text\")\\\n  .setOutputCol(\"document\")\n\n# Sentence Detector annotator, processes various sentences per line\n\nsentenceDetector = SentenceDetector()\\\n  .setInputCols([\"document\"])\\\n  .setOutputCol(\"sentence\")\n\n# Tokenizer splits words in a relevant format for NLP\n\ntokenizer = Tokenizer()\\\n  .setInputCols([\"sentence\"])\\\n  .setOutputCol(\"token\")\n\n# Clinical word embeddings trained on PubMED dataset\nword_embeddings = WordEmbeddingsModel.pretrained(\"embeddings_clinical\", \"en\", \"clinical/models\")\\\n  .setInputCols([\"sentence\", \"token\"])\\\n  .setOutputCol(\"embeddings\")\n\n# NER model trained on i2b2 (sampled from MIMIC) dataset\nclinical_ner = NerDLModel.pretrained(\"ner_clinical\", \"en\", \"clinical/models\") \\\n  .setInputCols([\"sentence\", \"token\", \"embeddings\"]) \\\n  .setOutputCol(\"ner\")\n\nner_converter = NerConverter() \\\n  .setInputCols([\"sentence\", \"token\", \"ner\"]) \\\n  .setOutputCol(\"ner_chunk\")\n\n# Assertion model trained on i2b2 (sampled from MIMIC) dataset\n# coming from sparknlp_jsl.annotator !!\nclinical_assertion = AssertionDLModel.pretrained(\"assertion_dl\", \"en\", \"clinical/models\") \\\n    .setInputCols([\"sentence\", \"ner_chunk\", \"embeddings\"]) \\\n    .setOutputCol(\"assertion\")\n    \nnlpPipeline = Pipeline(stages=[\n    documentAssembler, \n    sentenceDetector,\n    tokenizer,\n    word_embeddings,\n    clinical_ner,\n    ner_converter,\n    clinical_assertion\n    ])\n\nempty_data = spark.createDataFrame([[\"\"]]).toDF(\"text\")\n\nmodel = nlpPipeline.fit(empty_data)\n"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"064d576d-3e0b-497b-810d-c6e8c363a0c5"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[{"name":"empty_data","typeStr":"pyspark.sql.dataframe.DataFrame","schema":{"fields":[{"metadata":{},"name":"text","nullable":true,"type":"string"}],"type":"struct"},"tableIdentifier":null}],"data":"<div class=\"ansiout\">embeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[OK!]\nner_clinical download started this may take some time.\nApproximate size to download 13.8 MB\n\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[ / ]\r[OK!]\n</div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">embeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[OK!]\nner_clinical download started this may take some time.\nApproximate size to download 13.8 MB\n\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[ / ]\r[OK!]\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["# we also have a LogReg based Assertion Model.\n'''\nclinical_assertion_ml = AssertionLogRegModel.pretrained(\"assertion_ml\", \"en\", \"clinical/models\") \\\n    .setInputCols([\"sentence\", \"ner_chunk\", \"embeddings\"]) \\\n    .setOutputCol(\"assertion\")\n'''"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"0bede1e8-a98e-4774-ad2e-5863b56bd636"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">Out[4]: &#39;\\nclinical_assertion_ml = AssertionLogRegModel.pretrained(&#34;assertion_ml&#34;, &#34;en&#34;, &#34;clinical/models&#34;)     .setInputCols([&#34;sentence&#34;, &#34;ner_chunk&#34;, &#34;embeddings&#34;])     .setOutputCol(&#34;assertion&#34;)\\n&#39;</div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Out[4]: &#39;\\nclinical_assertion_ml = AssertionLogRegModel.pretrained(&#34;assertion_ml&#34;, &#34;en&#34;, &#34;clinical/models&#34;)     .setInputCols([&#34;sentence&#34;, &#34;ner_chunk&#34;, &#34;embeddings&#34;])     .setOutputCol(&#34;assertion&#34;)\\n&#39;</div>"]}}],"execution_count":0},{"cell_type":"code","source":["import pandas as pd\n\ntext = 'Patient has a headache for the last 2 weeks and appears anxious when she walks fast. No alopecia noted. She denies pain'\n\nprint (text)\n\nlight_model = LightPipeline(model)\n\nlight_result = light_model.fullAnnotate(text)[0]\n\nchunks=[]\nentities=[]\nstatus=[]\n\nfor n,m in zip(light_result['ner_chunk'],light_result['assertion']):\n    \n    chunks.append(n.result)\n    entities.append(n.metadata['entity']) \n    status.append(m.result)\n        \ndf = pd.DataFrame({'chunks':chunks, 'entities':entities, 'assertion':status})\n\ndf"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"6137f486-8004-4553-bb12-b2f01df9ab93"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chunks</th>\n      <th>entities</th>\n      <th>assertion</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>a headache</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>anxious</td>\n      <td>PROBLEM</td>\n      <td>conditional</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>alopecia</td>\n      <td>PROBLEM</td>\n      <td>absent</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>pain</td>\n      <td>PROBLEM</td>\n      <td>absent</td>\n    </tr>\n  </tbody>\n</table>\n</div>","textData":"<div class=\"ansiout\">Patient has a headache for the last 2 weeks and appears anxious when she walks fast. No alopecia noted. She denies pain\nOut[5]: </div>","removedWidgets":[],"addedWidgets":{},"type":"htmlSandbox","arguments":{}}},"output_type":"display_data","data":{"text/html":["<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chunks</th>\n      <th>entities</th>\n      <th>assertion</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>a headache</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>anxious</td>\n      <td>PROBLEM</td>\n      <td>conditional</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>alopecia</td>\n      <td>PROBLEM</td>\n      <td>absent</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>pain</td>\n      <td>PROBLEM</td>\n      <td>absent</td>\n    </tr>\n  </tbody>\n</table>\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["! wget \thttps://s3.amazonaws.com/auxdata.johnsnowlabs.com/public/resources/en/pubmed/pubmed_sample_text_small.csv\n"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"f1e1c581-7c0e-4b56-be6e-d604278e3d18"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["import pyspark.sql.functions as F\n\npubMedDF = spark.read\\\n                .option(\"header\", \"true\")\\\n                .csv(\"pubmed_sample_text_small.csv\")\\\n                \npubMedDF.show(truncate=50)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"1321721d-d896-4179-87c9-6716a68ccba4"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[{"name":"pubMedDF","typeStr":"pyspark.sql.dataframe.DataFrame","schema":{"fields":[{"metadata":{},"name":"text","nullable":true,"type":"string"}],"type":"struct"},"tableIdentifier":null}],"data":"<div class=\"ansiout\">+--------------------------------------------------+\n|                                              text|\n+--------------------------------------------------+\n|The human KCNJ9 (Kir 3.3, GIRK3) is a member of...|\n|BACKGROUND: At present, it is one of the most i...|\n|OBJECTIVE: To investigate the relationship betw...|\n|Combined EEG/fMRI recording has been used to lo...|\n|Kohlschutter syndrome is a rare neurodegenerati...|\n|Statistical analysis of neuroimages is commonly...|\n|The synthetic DOX-LNA conjugate was characteriz...|\n|Our objective was to compare three different me...|\n|We conducted a phase II study to assess the eff...|\n|&#34;&#34;&#34;Monomeric sarcosine oxidase (MSOX) is a flav...|\n|We presented the tachinid fly Exorista japonica...|\n|The literature dealing with the water conductin...|\n|A novel approach to synthesize chitosan-O-isopr...|\n|An HPLC-ESI-MS-MS method has been developed for...|\n|The localizing and lateralizing values of eye a...|\n|OBJECTIVE: To evaluate the effectiveness and ac...|\n|For the construction of new combinatorial libra...|\n|We report the results of a screen for genetic a...|\n|Intraparenchymal pericatheter cyst is rarely re...|\n|It is known that patients with Klinefelter&#39;s sy...|\n+--------------------------------------------------+\nonly showing top 20 rows\n\n</div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">+--------------------------------------------------+\n                                              text|\n+--------------------------------------------------+\nThe human KCNJ9 (Kir 3.3, GIRK3) is a member of...|\nBACKGROUND: At present, it is one of the most i...|\nOBJECTIVE: To investigate the relationship betw...|\nCombined EEG/fMRI recording has been used to lo...|\nKohlschutter syndrome is a rare neurodegenerati...|\nStatistical analysis of neuroimages is commonly...|\nThe synthetic DOX-LNA conjugate was characteriz...|\nOur objective was to compare three different me...|\nWe conducted a phase II study to assess the eff...|\n&#34;&#34;&#34;Monomeric sarcosine oxidase (MSOX) is a flav...|\nWe presented the tachinid fly Exorista japonica...|\nThe literature dealing with the water conductin...|\nA novel approach to synthesize chitosan-O-isopr...|\nAn HPLC-ESI-MS-MS method has been developed for...|\nThe localizing and lateralizing values of eye a...|\nOBJECTIVE: To evaluate the effectiveness and ac...|\nFor the construction of new combinatorial libra...|\nWe report the results of a screen for genetic a...|\nIntraparenchymal pericatheter cyst is rarely re...|\nIt is known that patients with Klinefelter&#39;s sy...|\n+--------------------------------------------------+\nonly showing top 20 rows\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["result = model.transform(pubMedDF.limit(100))"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"dd660faf-5874-42eb-99ef-a083952d91ab"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[{"name":"result","typeStr":"pyspark.sql.dataframe.DataFrame","schema":{"fields":[{"metadata":{},"name":"text","nullable":true,"type":"string"},{"metadata":{"annotatorType":"document"},"name":"document","nullable":true,"type":{"containsNull":true,"elementType":{"fields":[{"metadata":{},"name":"annotatorType","nullable":true,"type":"string"},{"metadata":{},"name":"begin","nullable":false,"type":"integer"},{"metadata":{},"name":"end","nullable":false,"type":"integer"},{"metadata":{},"name":"result","nullable":true,"type":"string"},{"metadata":{},"name":"metadata","nullable":true,"type":{"keyType":"string","type":"map","valueContainsNull":true,"valueType":"string"}},{"metadata":{},"name":"embeddings","nullable":true,"type":{"containsNull":false,"elementType":"float","type":"array"}}],"type":"struct"},"type":"array"}},{"metadata":{"annotatorType":"document"},"name":"sentence","nullable":true,"type":{"containsNull":true,"elementType":{"fields":[{"metadata":{},"name":"annotatorType","nullable":true,"type":"string"},{"metadata":{},"name":"begin","nullable":false,"type":"integer"},{"metadata":{},"name":"end","nullable":false,"type":"integer"},{"metadata":{},"name":"result","nullable":true,"type":"string"},{"metadata":{},"name":"metadata","nullable":true,"type":{"keyType":"string","type":"map","valueContainsNull":true,"valueType":"string"}},{"metadata":{},"name":"embeddings","nullable":true,"type":{"containsNull":false,"elementType":"float","type":"array"}}],"type":"struct"},"type":"array"}},{"metadata":{"annotatorType":"token"},"name":"token","nullable":true,"type":{"containsNull":true,"elementType":{"fields":[{"metadata":{},"name":"annotatorType","nullable":true,"type":"string"},{"metadata":{},"name":"begin","nullable":false,"type":"integer"},{"metadata":{},"name":"end","nullable":false,"type":"integer"},{"metadata":{},"name":"result","nullable":true,"type":"string"},{"metadata":{},"name":"metadata","nullable":true,"type":{"keyType":"string","type":"map","valueContainsNull":true,"valueType":"string"}},{"metadata":{},"name":"embeddings","nullable":true,"type":{"containsNull":false,"elementType":"float","type":"array"}}],"type":"struct"},"type":"array"}},{"metadata":{"annotatorType":"word_embeddings","dimension":200,"ref":"clinical"},"name":"embeddings","nullable":true,"type":{"containsNull":true,"elementType":{"fields":[{"metadata":{},"name":"annotatorType","nullable":true,"type":"string"},{"metadata":{},"name":"begin","nullable":false,"type":"integer"},{"metadata":{},"name":"end","nullable":false,"type":"integer"},{"metadata":{},"name":"result","nullable":true,"type":"string"},{"metadata":{},"name":"metadata","nullable":true,"type":{"keyType":"string","type":"map","valueContainsNull":true,"valueType":"string"}},{"metadata":{},"name":"embeddings","nullable":true,"type":{"containsNull":false,"elementType":"float","type":"array"}}],"type":"struct"},"type":"array"}},{"metadata":{"annotatorType":"named_entity"},"name":"ner","nullable":true,"type":{"containsNull":true,"elementType":{"fields":[{"metadata":{},"name":"annotatorType","nullable":true,"type":"string"},{"metadata":{},"name":"begin","nullable":false,"type":"integer"},{"metadata":{},"name":"end","nullable":false,"type":"integer"},{"metadata":{},"name":"result","nullable":true,"type":"string"},{"metadata":{},"name":"metadata","nullable":true,"type":{"keyType":"string","type":"map","valueContainsNull":true,"valueType":"string"}},{"metadata":{},"name":"embeddings","nullable":true,"type":{"containsNull":false,"elementType":"float","type":"array"}}],"type":"struct"},"type":"array"}},{"metadata":{"annotatorType":"chunk"},"name":"ner_chunk","nullable":true,"type":{"containsNull":true,"elementType":{"fields":[{"metadata":{},"name":"annotatorType","nullable":true,"type":"string"},{"metadata":{},"name":"begin","nullable":false,"type":"integer"},{"metadata":{},"name":"end","nullable":false,"type":"integer"},{"metadata":{},"name":"result","nullable":true,"type":"string"},{"metadata":{},"name":"metadata","nullable":true,"type":{"keyType":"string","type":"map","valueContainsNull":true,"valueType":"string"}},{"metadata":{},"name":"embeddings","nullable":true,"type":{"containsNull":false,"elementType":"float","type":"array"}}],"type":"struct"},"type":"array"}},{"metadata":{"annotatorType":"assertion"},"name":"assertion","nullable":true,"type":{"containsNull":true,"elementType":{"fields":[{"metadata":{},"name":"annotatorType","nullable":true,"type":"string"},{"metadata":{},"name":"begin","nullable":false,"type":"integer"},{"metadata":{},"name":"end","nullable":false,"type":"integer"},{"metadata":{},"name":"result","nullable":true,"type":"string"},{"metadata":{},"name":"metadata","nullable":true,"type":{"keyType":"string","type":"map","valueContainsNull":true,"valueType":"string"}},{"metadata":{},"name":"embeddings","nullable":true,"type":{"containsNull":false,"elementType":"float","type":"array"}}],"type":"struct"},"type":"array"}}],"type":"struct"},"tableIdentifier":null}],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["result.show()"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"7bf3ef67-9448-43e4-9d67-d7a47399911a"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+\n|                text|            document|            sentence|               token|          embeddings|                 ner|           ner_chunk|           assertion|\n+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+\n|The human KCNJ9 (...|[[document, 0, 95...|[[document, 0, 12...|[[token, 0, 2, Th...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 17, 19, ...|[[assertion, 17, ...|\n|BACKGROUND: At pr...|[[document, 0, 14...|[[document, 0, 19...|[[token, 0, 9, BA...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 67, 79, ...|[[assertion, 67, ...|\n|OBJECTIVE: To inv...|[[document, 0, 15...|[[document, 0, 30...|[[token, 0, 8, OB...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 130, 152...|[[assertion, 130,...|\n|Combined EEG/fMRI...|[[document, 0, 16...|[[document, 0, 16...|[[token, 0, 7, Co...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 9, 26, E...|[[assertion, 9, 2...|\n|Kohlschutter synd...|[[document, 0, 25...|[[document, 0, 20...|[[token, 0, 11, K...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 0, 20, K...|[[assertion, 0, 2...|\n|Statistical analy...|[[document, 0, 10...|[[document, 0, 34...|[[token, 0, 10, S...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 0, 34, S...|[[assertion, 0, 3...|\n|The synthetic DOX...|[[document, 0, 57...|[[document, 0, 10...|[[token, 0, 2, Th...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 4, 30, s...|[[assertion, 4, 3...|\n|Our objective was...|[[document, 0, 24...|[[document, 0, 19...|[[token, 0, 2, Ou...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 56, 80, ...|[[assertion, 56, ...|\n|We conducted a ph...|[[document, 0, 14...|[[document, 0, 24...|[[token, 0, 1, We...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 13, 28, ...|[[assertion, 13, ...|\n|&#34;&#34;&#34;Monomeric sarc...|[[document, 0, 14...|[[document, 0, 14...|[[token, 0, 2, &#34;&#34;...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 3, 36, M...|[[assertion, 3, 3...|\n|We presented the ...|[[document, 0, 12...|[[document, 0, 26...|[[token, 0, 1, We...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 135, 178...|[[assertion, 135,...|\n|The literature de...|[[document, 0, 16...|[[document, 0, 13...|[[token, 0, 2, Th...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 63, 83, ...|[[assertion, 63, ...|\n|A novel approach ...|[[document, 0, 64...|[[document, 0, 97...|[[token, 0, 0, A,...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 99, 132,...|[[assertion, 99, ...|\n|An HPLC-ESI-MS-MS...|[[document, 0, 90...|[[document, 0, 24...|[[token, 0, 1, An...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 0, 23, A...|[[assertion, 0, 2...|\n|The localizing an...|[[document, 0, 72...|[[document, 0, 12...|[[token, 0, 2, Th...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 42, 98, ...|[[assertion, 42, ...|\n|OBJECTIVE: To eva...|[[document, 0, 13...|[[document, 0, 43...|[[token, 0, 8, OB...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 62, 80, ...|[[assertion, 62, ...|\n|For the construct...|[[document, 0, 32...|[[document, 0, 20...|[[token, 0, 2, Fo...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 24, 50, ...|[[assertion, 24, ...|\n|We report the res...|[[document, 0, 13...|[[document, 0, 38...|[[token, 0, 1, We...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 25, 32, ...|[[assertion, 25, ...|\n|Intraparenchymal ...|[[document, 0, 10...|[[document, 0, 53...|[[token, 0, 15, I...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 0, 33, I...|[[assertion, 0, 3...|\n|It is known that ...|[[document, 0, 34...|[[document, 0, 14...|[[token, 0, 1, It...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 31, 52, ...|[[assertion, 31, ...|\n+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+\nonly showing top 20 rows\n\n</div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+\n                text|            document|            sentence|               token|          embeddings|                 ner|           ner_chunk|           assertion|\n+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+\nThe human KCNJ9 (...|[[document, 0, 95...|[[document, 0, 12...|[[token, 0, 2, Th...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 17, 19, ...|[[assertion, 17, ...|\nBACKGROUND: At pr...|[[document, 0, 14...|[[document, 0, 19...|[[token, 0, 9, BA...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 67, 79, ...|[[assertion, 67, ...|\nOBJECTIVE: To inv...|[[document, 0, 15...|[[document, 0, 30...|[[token, 0, 8, OB...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 130, 152...|[[assertion, 130,...|\nCombined EEG/fMRI...|[[document, 0, 16...|[[document, 0, 16...|[[token, 0, 7, Co...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 9, 26, E...|[[assertion, 9, 2...|\nKohlschutter synd...|[[document, 0, 25...|[[document, 0, 20...|[[token, 0, 11, K...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 0, 20, K...|[[assertion, 0, 2...|\nStatistical analy...|[[document, 0, 10...|[[document, 0, 34...|[[token, 0, 10, S...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 0, 34, S...|[[assertion, 0, 3...|\nThe synthetic DOX...|[[document, 0, 57...|[[document, 0, 10...|[[token, 0, 2, Th...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 4, 30, s...|[[assertion, 4, 3...|\nOur objective was...|[[document, 0, 24...|[[document, 0, 19...|[[token, 0, 2, Ou...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 56, 80, ...|[[assertion, 56, ...|\nWe conducted a ph...|[[document, 0, 14...|[[document, 0, 24...|[[token, 0, 1, We...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 13, 28, ...|[[assertion, 13, ...|\n&#34;&#34;&#34;Monomeric sarc...|[[document, 0, 14...|[[document, 0, 14...|[[token, 0, 2, &#34;&#34;...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 3, 36, M...|[[assertion, 3, 3...|\nWe presented the ...|[[document, 0, 12...|[[document, 0, 26...|[[token, 0, 1, We...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 135, 178...|[[assertion, 135,...|\nThe literature de...|[[document, 0, 16...|[[document, 0, 13...|[[token, 0, 2, Th...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 63, 83, ...|[[assertion, 63, ...|\nA novel approach ...|[[document, 0, 64...|[[document, 0, 97...|[[token, 0, 0, A,...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 99, 132,...|[[assertion, 99, ...|\nAn HPLC-ESI-MS-MS...|[[document, 0, 90...|[[document, 0, 24...|[[token, 0, 1, An...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 0, 23, A...|[[assertion, 0, 2...|\nThe localizing an...|[[document, 0, 72...|[[document, 0, 12...|[[token, 0, 2, Th...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 42, 98, ...|[[assertion, 42, ...|\nOBJECTIVE: To eva...|[[document, 0, 13...|[[document, 0, 43...|[[token, 0, 8, OB...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 62, 80, ...|[[assertion, 62, ...|\nFor the construct...|[[document, 0, 32...|[[document, 0, 20...|[[token, 0, 2, Fo...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 24, 50, ...|[[assertion, 24, ...|\nWe report the res...|[[document, 0, 13...|[[document, 0, 38...|[[token, 0, 1, We...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 25, 32, ...|[[assertion, 25, ...|\nIntraparenchymal ...|[[document, 0, 10...|[[document, 0, 53...|[[token, 0, 15, I...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 0, 33, I...|[[assertion, 0, 3...|\nIt is known that ...|[[document, 0, 34...|[[document, 0, 14...|[[token, 0, 1, It...|[[word_embeddings...|[[named_entity, 0...|[[chunk, 31, 52, ...|[[assertion, 31, ...|\n+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+\nonly showing top 20 rows\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["result.select('sentence.result').take(1)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"cd49fbf0-b7d7-4523-b08c-d02fc6271ffb"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">Out[10]: [Row(result=[&#39;The human KCNJ9 (Kir 3.3, GIRK3) is a member of the G-protein-activated inwardly rectifying potassium (GIRK) channel family.&#39;, &#39;Here we describe the genomicorganization of the KCNJ9 locus on chromosome 1q21-23 as a candidate gene forType II diabetes mellitus in the Pima Indian population.&#39;, &#39;The gene spansapproximately 7.6 kb and contains one noncoding and two coding exons separated byapproximately 2.2 and approximately 2.6 kb introns, respectively.&#39;, &#39;We identified14 single nucleotide polymorphisms (SNPs), including one that predicts aVal366Ala substitution, and an 8 base-pair&#39;, &#39;(bp) insertion/deletion.&#39;, &#39;Ourexpression studies revealed the presence of the transcript in various humantissues including pancreas, and two major insulin-responsive tissues: fat andskeletal muscle.&#39;, &#39;The characterization of the KCNJ9 gene should facilitate furtherstudies on the function of the KCNJ9 protein and allow evaluation of thepotential role of the locus in Type II diabetes.&#39;])]</div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Out[10]: [Row(result=[&#39;The human KCNJ9 (Kir 3.3, GIRK3) is a member of the G-protein-activated inwardly rectifying potassium (GIRK) channel family.&#39;, &#39;Here we describe the genomicorganization of the KCNJ9 locus on chromosome 1q21-23 as a candidate gene forType II diabetes mellitus in the Pima Indian population.&#39;, &#39;The gene spansapproximately 7.6 kb and contains one noncoding and two coding exons separated byapproximately 2.2 and approximately 2.6 kb introns, respectively.&#39;, &#39;We identified14 single nucleotide polymorphisms (SNPs), including one that predicts aVal366Ala substitution, and an 8 base-pair&#39;, &#39;(bp) insertion/deletion.&#39;, &#39;Ourexpression studies revealed the presence of the transcript in various humantissues including pancreas, and two major insulin-responsive tissues: fat andskeletal muscle.&#39;, &#39;The characterization of the KCNJ9 gene should facilitate furtherstudies on the function of the KCNJ9 protein and allow evaluation of thepotential role of the locus in Type II diabetes.&#39;])]</div>"]}}],"execution_count":0},{"cell_type":"code","source":["result.select(F.explode(F.arrays_zip('ner_chunk.result',  'ner_chunk.begin', 'ner_chunk.end', 'ner_chunk.metadata', 'assertion.result')).alias(\"cols\")) \\\n.select(F.expr(\"cols['0']\").alias(\"chunk\"),\n        F.expr(\"cols['1']\").alias(\"begin\"),\n        F.expr(\"cols['2']\").alias(\"end\"),\n        F.expr(\"cols['3']['entity']\").alias(\"ner_label\"),\n        F.expr(\"cols['3']['sentence']\").alias(\"sent_id\"),\n        F.expr(\"cols['4']\").alias(\"assertion\"),).show(truncate=False)\n"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"5a9f1fd2-875e-4595-bfc1-8b5132cee945"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">+---------------------------------------------+-----+---+---------+-------+---------+\n|chunk                                        |begin|end|ner_label|sent_id|assertion|\n+---------------------------------------------+-----+---+---------+-------+---------+\n|Kir                                          |17   |19 |TEST     |0      |present  |\n|GIRK3                                        |26   |30 |TEST     |0      |present  |\n|chromosome 1q21-23                           |188  |205|TREATMENT|1      |present  |\n|a candidate gene forType II diabetes mellitus|210  |254|PROBLEM  |1      |possible |\n|coding exons                                 |357  |368|TEST     |2      |absent   |\n|byapproximately                              |380  |394|TEST     |2      |present  |\n|introns                                      |425  |431|TEST     |2      |absent   |\n|single nucleotide polymorphisms (SNPs        |464  |500|PROBLEM  |3      |present  |\n|aVal366Ala substitution                      |532  |554|TREATMENT|3      |present  |\n|an 8 base-pair                               |561  |574|PROBLEM  |3      |present  |\n|bp)                                          |577  |579|TEST     |4      |present  |\n|Ourexpression studies                        |601  |621|TEST     |5      |present  |\n|The characterization of the KCNJ9 gene       |773  |810|TEST     |6      |present  |\n|furtherstudies                               |830  |843|TREATMENT|6      |present  |\n|KCNJ9 protein                                |868  |880|TEST     |6      |present  |\n|evaluation                                   |892  |901|TEST     |6      |present  |\n|Type II diabetes                             |940  |955|PROBLEM  |6      |present  |\n|the treatment                                |67   |79 |TREATMENT|0      |present  |\n|breast cancer                                |84   |96 |PROBLEM  |0      |present  |\n|anthracyclines                               |167  |180|TREATMENT|0      |present  |\n+---------------------------------------------+-----+---+---------+-------+---------+\nonly showing top 20 rows\n\n</div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">+---------------------------------------------+-----+---+---------+-------+---------+\nchunk                                        |begin|end|ner_label|sent_id|assertion|\n+---------------------------------------------+-----+---+---------+-------+---------+\nKir                                          |17   |19 |TEST     |0      |present  |\nGIRK3                                        |26   |30 |TEST     |0      |present  |\nchromosome 1q21-23                           |188  |205|TREATMENT|1      |present  |\na candidate gene forType II diabetes mellitus|210  |254|PROBLEM  |1      |possible |\ncoding exons                                 |357  |368|TEST     |2      |absent   |\nbyapproximately                              |380  |394|TEST     |2      |present  |\nintrons                                      |425  |431|TEST     |2      |absent   |\nsingle nucleotide polymorphisms (SNPs        |464  |500|PROBLEM  |3      |present  |\naVal366Ala substitution                      |532  |554|TREATMENT|3      |present  |\nan 8 base-pair                               |561  |574|PROBLEM  |3      |present  |\nbp)                                          |577  |579|TEST     |4      |present  |\nOurexpression studies                        |601  |621|TEST     |5      |present  |\nThe characterization of the KCNJ9 gene       |773  |810|TEST     |6      |present  |\nfurtherstudies                               |830  |843|TREATMENT|6      |present  |\nKCNJ9 protein                                |868  |880|TEST     |6      |present  |\nevaluation                                   |892  |901|TEST     |6      |present  |\nType II diabetes                             |940  |955|PROBLEM  |6      |present  |\nthe treatment                                |67   |79 |TREATMENT|0      |present  |\nbreast cancer                                |84   |96 |PROBLEM  |0      |present  |\nanthracyclines                               |167  |180|TREATMENT|0      |present  |\n+---------------------------------------------+-----+---+---------+-------+---------+\nonly showing top 20 rows\n\n</div>"]}}],"execution_count":0},{"cell_type":"markdown","source":["## Writing a generic Assertion + NER function"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"afef480c-1540-4acc-88c9-5711e5290f3b"}}},{"cell_type":"code","source":["from pyspark.sql.functions import monotonically_increasing_id\n\n\ndef get_base_pipeline (embeddings = 'embeddings_clinical'):\n\n  documentAssembler = DocumentAssembler()\\\n    .setInputCol(\"text\")\\\n    .setOutputCol(\"document\")\n\n  # Sentence Detector annotator, processes various sentences per line\n\n  sentenceDetector = SentenceDetector()\\\n    .setInputCols([\"document\"])\\\n    .setOutputCol(\"sentence\")\n\n  # Tokenizer splits words in a relevant format for NLP\n\n  tokenizer = Tokenizer()\\\n    .setInputCols([\"sentence\"])\\\n    .setOutputCol(\"token\")\n\n  # Clinical word embeddings trained on PubMED dataset\n  word_embeddings = WordEmbeddingsModel.pretrained(embeddings, \"en\", \"clinical/models\")\\\n    .setInputCols([\"sentence\", \"token\"])\\\n    .setOutputCol(\"embeddings\")\n\n  empty_data = spark.createDataFrame([[\"\"]]).toDF(\"text\")\n\n  model = nlpPipeline.fit(empty_data)\n\n  return model\n\n\n\ndef get_clinical_assertion (embeddings, spark_df, nrows = 100, model_name = 'ner_clinical'):\n\n  # NER model trained on i2b2 (sampled from MIMIC) dataset\n  loaded_ner_model = NerDLModel.pretrained(model_name, \"en\", \"clinical/models\") \\\n    .setInputCols([\"sentence\", \"token\", \"embeddings\"]) \\\n    .setOutputCol(\"ner\")\n\n  ner_converter = NerConverter() \\\n  .setInputCols([\"sentence\", \"token\", \"ner\"]) \\\n  .setOutputCol(\"ner_chunk\")\n\n  # Assertion model trained on i2b2 (sampled from MIMIC) dataset\n  # coming from sparknlp_jsl.annotator !!\n  clinical_assertion = AssertionDLModel.pretrained(\"assertion_dl\", \"en\", \"clinical/models\") \\\n      .setInputCols([\"sentence\", \"ner_chunk\", \"embeddings\"]) \\\n      .setOutputCol(\"assertion\")\n      \n\n  base_model = get_base_pipeline (embeddings)\n\n  nlpPipeline = Pipeline(stages=[\n      base_model,\n      loaded_ner_model,\n      ner_converter])\n\n  empty_data = spark.createDataFrame([[\"\"]]).toDF(\"text\")\n\n  model = nlpPipeline.fit(empty_data)\n\n  result = model.transform(spark_df.limit(nrows))\n\n  result = result.withColumn(\"id\", monotonically_increasing_id())\n\n  result_df = result.select(F.explode(F.arrays_zip('ner_chunk.result', 'ner_chunk.metadata', 'assertion.result')).alias(\"cols\")) \\\n  .select(F.expr(\"cols['0']\").alias(\"chunk\"),\n        F.expr(\"cols['1']['entity']\").alias(\"ner_label\"),\n        F.expr(\"cols['2']\").alias(\"assertion\"))\\\n        .filter(\"ner_label!='O'\")\n\n  return result_df"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"2390fc66-0451-475b-9597-e03e2b3cee09"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["embeddings = 'embeddings_clinical'\n\nmodel_name = 'ner_clinical'\n\nnrows = 1000\n\nner_df = get_clinical_assertion (embeddings, pubMedDF, nrows, model_name)\n\nner_df.show()"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"5cc9a0fc-5187-4e7b-9fee-283b5fd343c9"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[{"name":"ner_df","typeStr":"pyspark.sql.dataframe.DataFrame","schema":{"fields":[{"metadata":{},"name":"chunk","nullable":true,"type":"string"},{"metadata":{},"name":"ner_label","nullable":true,"type":"string"},{"metadata":{},"name":"assertion","nullable":true,"type":"string"}],"type":"struct"},"tableIdentifier":null}],"data":"<div class=\"ansiout\">ner_clinical download started this may take some time.\nApproximate size to download 13.8 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n+--------------------+---------+---------+\n|               chunk|ner_label|assertion|\n+--------------------+---------+---------+\n|                 Kir|     TEST|  present|\n|               GIRK3|     TEST|  present|\n|  chromosome 1q21-23|TREATMENT|  present|\n|a candidate gene ...|  PROBLEM| possible|\n|        coding exons|     TEST|   absent|\n|     byapproximately|     TEST|  present|\n|             introns|     TEST|   absent|\n|single nucleotide...|  PROBLEM|  present|\n|aVal366Ala substi...|TREATMENT|  present|\n|      an 8 base-pair|  PROBLEM|  present|\n|                 bp)|     TEST|  present|\n|Ourexpression stu...|     TEST|  present|\n|The characterizat...|     TEST|  present|\n|      furtherstudies|TREATMENT|  present|\n|       KCNJ9 protein|     TEST|  present|\n|          evaluation|     TEST|  present|\n|    Type II diabetes|  PROBLEM|  present|\n|       the treatment|TREATMENT|  present|\n|       breast cancer|  PROBLEM|  present|\n|      anthracyclines|TREATMENT|  present|\n+--------------------+---------+---------+\nonly showing top 20 rows\n\n</div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">ner_clinical download started this may take some time.\nApproximate size to download 13.8 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n+--------------------+---------+---------+\n               chunk|ner_label|assertion|\n+--------------------+---------+---------+\n                 Kir|     TEST|  present|\n               GIRK3|     TEST|  present|\n  chromosome 1q21-23|TREATMENT|  present|\na candidate gene ...|  PROBLEM| possible|\n        coding exons|     TEST|   absent|\n     byapproximately|     TEST|  present|\n             introns|     TEST|   absent|\nsingle nucleotide...|  PROBLEM|  present|\naVal366Ala substi...|TREATMENT|  present|\n      an 8 base-pair|  PROBLEM|  present|\n                 bp)|     TEST|  present|\nOurexpression stu...|     TEST|  present|\nThe characterizat...|     TEST|  present|\n      furtherstudies|TREATMENT|  present|\n       KCNJ9 protein|     TEST|  present|\n          evaluation|     TEST|  present|\n    Type II diabetes|  PROBLEM|  present|\n       the treatment|TREATMENT|  present|\n       breast cancer|  PROBLEM|  present|\n      anthracyclines|TREATMENT|  present|\n+--------------------+---------+---------+\nonly showing top 20 rows\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["embeddings = 'embeddings_clinical'\n\nmodel_name = 'ner_posology'\n\nnrows = 100\n\nner_df = get_clinical_assertion (embeddings, pubMedDF, nrows, model_name)\n\nner_df.show()"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"cb696bfc-6a34-4666-92e9-607683a57506"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[{"name":"ner_df","typeStr":"pyspark.sql.dataframe.DataFrame","schema":{"fields":[{"metadata":{},"name":"chunk","nullable":true,"type":"string"},{"metadata":{},"name":"ner_label","nullable":true,"type":"string"},{"metadata":{},"name":"assertion","nullable":true,"type":"string"}],"type":"struct"},"tableIdentifier":null}],"data":"<div class=\"ansiout\">ner_posology download started this may take some time.\nApproximate size to download 13.7 MB\n\r[ | ]\r[ / ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n</div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">ner_posology download started this may take some time.\nApproximate size to download 13.7 MB\n\r[ | ]\r[ / ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n</div>"]}},{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"type":"ipynbError","data":"","errorSummary":"Cancelled","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>"]}}],"execution_count":0},{"cell_type":"code","source":["embeddings = 'embeddings_clinical'\n\nmodel_name = 'ner_posology'\n\nentry_data = spark.createDataFrame([[\"The patient didn't take a capsule of Advil\"]]).toDF(\"text\")\n\nner_df = get_clinical_assertion (embeddings, entry_data, nrows, model_name)\n\nner_df.show()"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"93fe3c92-4d0e-4e96-99d8-08e41f7aeb6b"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[{"name":"entry_data","typeStr":"pyspark.sql.dataframe.DataFrame","schema":{"fields":[{"metadata":{},"name":"text","nullable":true,"type":"string"}],"type":"struct"},"tableIdentifier":null},{"name":"ner_df","typeStr":"pyspark.sql.dataframe.DataFrame","schema":{"fields":[{"metadata":{},"name":"chunk","nullable":true,"type":"string"},{"metadata":{},"name":"ner_label","nullable":true,"type":"string"},{"metadata":{},"name":"assertion","nullable":true,"type":"string"}],"type":"struct"},"tableIdentifier":null}],"data":"<div class=\"ansiout\">ner_posology download started this may take some time.\nApproximate size to download 13.7 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n+-------+---------+---------+\n|  chunk|ner_label|assertion|\n+-------+---------+---------+\n|capsule|     FORM|  present|\n|  Advil|     DRUG|     null|\n+-------+---------+---------+\n\n</div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">ner_posology download started this may take some time.\nApproximate size to download 13.7 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n+-------+---------+---------+\n  chunk|ner_label|assertion|\n+-------+---------+---------+\ncapsule|     FORM|  present|\n  Advil|     DRUG|     null|\n+-------+---------+---------+\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["import pandas as pd\n\ndef get_clinical_assertion_light (light_model, text):\n\n  light_result = light_model.fullAnnotate(text)[0]\n\n  chunks=[]\n  entities=[]\n  status=[]\n\n  for n,m in zip(light_result['ner_chunk'],light_result['assertion']):\n      \n      chunks.append(n.result)\n      entities.append(n.metadata['entity']) \n      status.append(m.result)\n          \n  df = pd.DataFrame({'chunks':chunks, 'entities':entities, 'assertion':status})\n\n  return df"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"925cdb03-42c3-43d5-ab5f-f6544fd9b5c3"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["clinical_text = \"\"\"\nPatient with severe fever and sore throat. \nHe shows no stomach pain and he maintained on an epidural and PCA for pain control.\nHe also became short of breath with climbing a flight of stairs.\nAfter CT, lung tumor located at the right lower lobe. Father with Alzheimer.\n\"\"\"\n\nlight_model = LightPipeline(model)\n\nget_clinical_assertion_light (light_model, clinical_text)\n"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"2f849a87-7851-4d0b-add0-540c0527d6ba"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chunks</th>\n      <th>entities</th>\n      <th>assertion</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>severe fever</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>sore throat</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>stomach pain</td>\n      <td>PROBLEM</td>\n      <td>absent</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>an epidural</td>\n      <td>TREATMENT</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>PCA</td>\n      <td>TREATMENT</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>pain control</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>short of breath</td>\n      <td>PROBLEM</td>\n      <td>conditional</td>\n    </tr>\n    <tr>\n      <th>7</th>\n      <td>CT</td>\n      <td>TEST</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>8</th>\n      <td>lung tumor</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>9</th>\n      <td>Alzheimer</td>\n      <td>PROBLEM</td>\n      <td>associated_with_someone_else</td>\n    </tr>\n  </tbody>\n</table>\n</div>","textData":"<div class=\"ansiout\">Out[17]: </div>","removedWidgets":[],"addedWidgets":{},"type":"htmlSandbox","arguments":{}}},"output_type":"display_data","data":{"text/html":["<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chunks</th>\n      <th>entities</th>\n      <th>assertion</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>severe fever</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>sore throat</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>stomach pain</td>\n      <td>PROBLEM</td>\n      <td>absent</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>an epidural</td>\n      <td>TREATMENT</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>PCA</td>\n      <td>TREATMENT</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>pain control</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>short of breath</td>\n      <td>PROBLEM</td>\n      <td>conditional</td>\n    </tr>\n    <tr>\n      <th>7</th>\n      <td>CT</td>\n      <td>TEST</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>8</th>\n      <td>lung tumor</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>9</th>\n      <td>Alzheimer</td>\n      <td>PROBLEM</td>\n      <td>associated_with_someone_else</td>\n    </tr>\n  </tbody>\n</table>\n</div>"]}}],"execution_count":0},{"cell_type":"markdown","source":["## Assertion with BioNLP (Cancer Genetics) NER"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"bdd0c509-8006-4c5c-9092-40256f47e34c"}}},{"cell_type":"code","source":["embeddings = 'embeddings_clinical'\n\nmodel_name = 'ner_bionlp'\n\nnrows = 100\n\nner_df = get_clinical_assertion (embeddings, pubMedDF, nrows, model_name)\n\nner_df.show(truncate = False)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"2ed9f618-83f3-4225-8e4c-79984a17c6ac"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[{"name":"ner_df","typeStr":"pyspark.sql.dataframe.DataFrame","schema":{"fields":[{"metadata":{},"name":"chunk","nullable":true,"type":"string"},{"metadata":{},"name":"ner_label","nullable":true,"type":"string"},{"metadata":{},"name":"assertion","nullable":true,"type":"string"}],"type":"struct"},"tableIdentifier":null}],"data":"<div class=\"ansiout\">ner_bionlp download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[ / ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n</div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">ner_bionlp download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[ / ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n</div>"]}},{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"type":"ipynbError","data":"","errorSummary":"Cancelled","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>"]}}],"execution_count":0},{"cell_type":"markdown","source":["# Train a custom Assertion Model"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"647435eb-a456-488f-ade2-f97e4c8de429"}}},{"cell_type":"code","source":["!wget https://s3.amazonaws.com/auxdata.johnsnowlabs.com/public/resources/en/ner/i2b2_assertion_sample.csv"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"3350b59b-948c-45d7-b683-838e030d9283"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["%sh ls -la"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"8c9acf67-9caf-4252-b560-48c8d759747c"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">total 28944\ndrwxr-xr-x 1 root root    4096 Nov  1 13:53 .\ndrwxr-xr-x 1 root root    4096 Nov  1 11:54 ..\n-rw-r--r-- 1 root root  514752 Nov  1 12:56 amazon_alexa.tsv\n-rw-r--r-- 1 root root  514752 Nov  1 13:16 amazon_alexa.tsv.1\ndrwxr-xr-x 2 root root    4096 Jan  1  1970 conf\n-rw-r--r-- 1 root root     717 Nov  1 11:54 derby.log\ndrwxr-xr-x 3 root root    4096 Nov  1 11:54 eventlogs\ndrwxr-xr-x 3 root root    4096 Nov  1 12:03 file:\ndrwxr-xr-x 2 root root    4096 Nov  1 14:00 ganglia\ndrwxr-xr-x 2 root root    4096 Nov  1 13:11 gc_graph\n-rw-r--r-- 1 root root  141585 Apr 14  2020 i2b2_assertion_sample.csv\ndrwxr-xr-x 2 root root    4096 Nov  1 14:00 logs\n-rw-r--r-- 1 root root 4763125 Nov  1 11:57 petfinder-mini.csv\n-rw-r--r-- 1 root root 4763125 Nov  1 12:50 petfinder-mini.csv.1\n-rw-r--r-- 1 root root 4763125 Nov  1 13:05 petfinder-mini.csv.2\n-rw-r--r-- 1 root root 4763125 Nov  1 13:10 petfinder-mini.csv.3\n-rw-r--r-- 1 root root 9363435 Apr  9  2020 pubmed_sample_text_small.csv\ndrwxr-xr-x 2 root root    4096 Nov  1 13:53 tf_graphs\ndrwxr-xr-x 2 root root    4096 Nov  1 13:53 training_logs\n</div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">total 28944\ndrwxr-xr-x 1 root root    4096 Nov  1 13:53 .\ndrwxr-xr-x 1 root root    4096 Nov  1 11:54 ..\n-rw-r--r-- 1 root root  514752 Nov  1 12:56 amazon_alexa.tsv\n-rw-r--r-- 1 root root  514752 Nov  1 13:16 amazon_alexa.tsv.1\ndrwxr-xr-x 2 root root    4096 Jan  1  1970 conf\n-rw-r--r-- 1 root root     717 Nov  1 11:54 derby.log\ndrwxr-xr-x 3 root root    4096 Nov  1 11:54 eventlogs\ndrwxr-xr-x 3 root root    4096 Nov  1 12:03 file:\ndrwxr-xr-x 2 root root    4096 Nov  1 14:00 ganglia\ndrwxr-xr-x 2 root root    4096 Nov  1 13:11 gc_graph\n-rw-r--r-- 1 root root  141585 Apr 14  2020 i2b2_assertion_sample.csv\ndrwxr-xr-x 2 root root    4096 Nov  1 14:00 logs\n-rw-r--r-- 1 root root 4763125 Nov  1 11:57 petfinder-mini.csv\n-rw-r--r-- 1 root root 4763125 Nov  1 12:50 petfinder-mini.csv.1\n-rw-r--r-- 1 root root 4763125 Nov  1 13:05 petfinder-mini.csv.2\n-rw-r--r-- 1 root root 4763125 Nov  1 13:10 petfinder-mini.csv.3\n-rw-r--r-- 1 root root 9363435 Apr  9  2020 pubmed_sample_text_small.csv\ndrwxr-xr-x 2 root root    4096 Nov  1 13:53 tf_graphs\ndrwxr-xr-x 2 root root    4096 Nov  1 13:53 training_logs\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["\nassertion_df = spark.read.option(\"header\", True).option(\"inferSchema\", \"True\").csv(\"i2b2_assertion_sample.csv\")\n"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"711aa642-d5ab-4f57-8063-4451b6ee0d11"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[{"name":"assertion_df","typeStr":"pyspark.sql.dataframe.DataFrame","schema":{"fields":[{"metadata":{},"name":"text","nullable":true,"type":"string"},{"metadata":{},"name":"target","nullable":true,"type":"string"},{"metadata":{},"name":"label","nullable":true,"type":"string"},{"metadata":{},"name":"start","nullable":true,"type":"integer"},{"metadata":{},"name":"end","nullable":true,"type":"integer"}],"type":"struct"},"tableIdentifier":null}],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["assertion_df.show(3, truncate=100)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"d9f89d7b-ce06-43dc-bdff-e53cbb6f8ada"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">+-----------------------------------------------------------------------------+--------------------+-------+-----+---+\n|                                                                         text|              target|  label|start|end|\n+-----------------------------------------------------------------------------+--------------------+-------+-----+---+\n|Due to dye load required to assess the coronary disease were not visualized .|the coronary disease|present|    7|  9|\n|                            She has no history of liver disease , hepatitis .|       liver disease| absent|    5|  6|\n|                                                     1. Undesired fertility .| undesired fertility|present|    1|  2|\n+-----------------------------------------------------------------------------+--------------------+-------+-----+---+\nonly showing top 3 rows\n\n</div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">+-----------------------------------------------------------------------------+--------------------+-------+-----+---+\n                                                                         text|              target|  label|start|end|\n+-----------------------------------------------------------------------------+--------------------+-------+-----+---+\nDue to dye load required to assess the coronary disease were not visualized .|the coronary disease|present|    7|  9|\n                            She has no history of liver disease , hepatitis .|       liver disease| absent|    5|  6|\n                                                     1. Undesired fertility .| undesired fertility|present|    1|  2|\n+-----------------------------------------------------------------------------+--------------------+-------+-----+---+\nonly showing top 3 rows\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["(training_data, test_data) = assertion_df.randomSplit([0.8, 0.2], seed = 100)\nprint(\"Training Dataset Count: \" + str(training_data.count()))\nprint(\"Test Dataset Count: \" + str(test_data.count()))"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"eba6ba43-dbd4-4403-b967-8ecde9e0461a"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[{"name":"training_data","typeStr":"pyspark.sql.dataframe.DataFrame","schema":{"fields":[{"metadata":{},"name":"text","nullable":true,"type":"string"},{"metadata":{},"name":"target","nullable":true,"type":"string"},{"metadata":{},"name":"label","nullable":true,"type":"string"},{"metadata":{},"name":"start","nullable":true,"type":"integer"},{"metadata":{},"name":"end","nullable":true,"type":"integer"}],"type":"struct"},"tableIdentifier":null},{"name":"test_data","typeStr":"pyspark.sql.dataframe.DataFrame","schema":{"fields":[{"metadata":{},"name":"text","nullable":true,"type":"string"},{"metadata":{},"name":"target","nullable":true,"type":"string"},{"metadata":{},"name":"label","nullable":true,"type":"string"},{"metadata":{},"name":"start","nullable":true,"type":"integer"},{"metadata":{},"name":"end","nullable":true,"type":"integer"}],"type":"struct"},"tableIdentifier":null}],"data":"<div class=\"ansiout\">Training Dataset Count: 796\nTest Dataset Count: 204\n</div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Training Dataset Count: 796\nTest Dataset Count: 204\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["training_data.groupBy('label').count().orderBy('count', ascending=False).show(truncate=False)\n"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"208b749b-13fa-4780-93a1-d7cd316e81f6"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">+----------------------------+-----+\n|label                       |count|\n+----------------------------+-----+\n|present                     |529  |\n|absent                      |184  |\n|hypothetical                |37   |\n|possible                    |36   |\n|associated_with_someone_else|6    |\n|conditional                 |4    |\n+----------------------------+-----+\n\n</div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">+----------------------------+-----+\nlabel                       |count|\n+----------------------------+-----+\npresent                     |529  |\nabsent                      |184  |\nhypothetical                |37   |\npossible                    |36   |\nassociated_with_someone_else|6    |\nconditional                 |4    |\n+----------------------------+-----+\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["\ndocument = DocumentAssembler()\\\n    .setInputCol(\"text\")\\\n    .setOutputCol(\"document\")\n\nchunk = Doc2Chunk()\\\n    .setInputCols(\"document\")\\\n    .setOutputCol(\"chunk\")\\\n    .setChunkCol(\"target\")\\\n    .setStartCol(\"start\")\\\n    .setStartColByTokenIndex(True)\\\n    .setFailOnMissing(False)\\\n    .setLowerCase(True)\n\ntoken = Tokenizer()\\\n    .setInputCols(['document'])\\\n    .setOutputCol('token')\n\nembeddings = WordEmbeddingsModel.pretrained(\"embeddings_clinical\", \"en\", \"clinical/models\")\\\n      .setInputCols([\"document\", \"token\"])\\\n      .setOutputCol(\"embeddings\")\n"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"1dd09bfb-5a71-49b0-8e5a-ce40820d154a"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">embeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n</div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">embeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["!mkdir training_logs\n!mkdir tf_graphs\n!wget -q https://raw.githubusercontent.com/JohnSnowLabs/spark-nlp-workshop/master/tutorials/Certification_Trainings/Healthcare/nerdl_graph/blstm_34_32_30_200_6.pb -P /databricks/driver/tf_graphs"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"3b33acf9-6372-4742-b984-748d9c619f21"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["\nassertionStatus = AssertionDLApproach()\\\n    .setLabelCol(\"label\")\\\n    .setInputCols(\"document\", \"chunk\", \"embeddings\")\\\n    .setOutputCol(\"assertion\")\\\n    .setBatchSize(128)\\\n    .setDropout(0.1)\\\n    .setLearningRate(0.001)\\\n    .setEpochs(10)\\\n    .setValidationSplit(0.2)\\\n    .setStartCol(\"start\")\\\n    .setEndCol(\"end\")\\\n    .setMaxSentLen(250)\\\n  .setClasses(6)\\\n.setEnableOutputLogs(True)\\\n.setOutputLogsPath('training_logs/')\\\n.setGraphFolder('file:/databricks/driver/tf_graphs')\n\n# set this path to graphs if you have different nuber of classes other than 6. (use 2.1 Asertion Model Graph generation notebook)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"19b3315c-86e7-4b86-9d0a-18dfd1b1ccc5"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["\nfrom sparknlp_jsl.training import tf_graph\n\nimport sys\n\nsys.path.append('/databricks/driver/tf_graph') \n\ntf_graph.print_model_params(\"assertion_dl\")"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"ac304260-fdb8-432d-8c27-0a096e22047e"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">assertion_dl parameters.\nParameter            Required   Default value        Description\nmax_seq_len          no         250                  Maximum sequence length.\nfeat_size            no         200                  Feature size.\nn_classes            yes        -                    Number of classes.\ndevice               no         /cpu:0               Device for training.\nn_hidden             no         34                   Number of hidden units.\n</div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">assertion_dl parameters.\nParameter            Required   Default value        Description\nmax_seq_len          no         250                  Maximum sequence length.\nfeat_size            no         200                  Feature size.\nn_classes            yes        -                    Number of classes.\ndevice               no         /cpu:0               Device for training.\nn_hidden             no         34                   Number of hidden units.\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["max_seq_len = 250\nfeat_size = 200\nn_classes = 7\n\n\ntf_graph.build(\"assertion_dl\",build_params={\"n_classes\": n_classes, 'max_seq_len':max_seq_len, 'n_classes':n_classes}, model_location=\"dbfs:/databricks/driver/tf_graphs\", model_filename=\"auto\")"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"1e0ad085-48fb-4783-bd49-fcab9a3b8eee"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">assertion_dl graph exported to dbfs:/databricks/driver/tf_graphs/blstm_250_200_34_7.pb\n</div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">assertion_dl graph exported to dbfs:/databricks/driver/tf_graphs/blstm_250_200_34_7.pb\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["'''\nassertionStatus = AssertionLogRegApproach()\\\n    .setLabelCol(\"label\")\\\n    .setInputCols(\"document\", \"chunk\", \"embeddings\")\\\n    .setOutputCol(\"assertion\")\\\n    .setMaxIter(100) # default: 26\n'''"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"a6afa3df-0e11-49e7-b388-553ff51b6837"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"type":"ipynbError","data":"","errorSummary":"Command skipped","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>"]}}],"execution_count":0},{"cell_type":"code","source":["clinical_assertion_pipeline = Pipeline(\n    stages = [\n    document,\n    chunk,\n    token,\n    embeddings,\n    assertionStatus])"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"894628f4-bb2e-448e-bcd1-b29462f9b9cb"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["#%%time\n\nassertion_model = clinical_assertion_pipeline.fit(training_data)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"9327a04d-1037-4261-b766-951d9a902ee0"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["%sh cd training_logs && ls -lt"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"5f35ecae-055a-4c36-a029-039957c159e0"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">total 8\n-rw-r--r-- 1 root root 6714 Nov  1 14:08 AssertionDLApproach_037b304ea9cc.log\n</div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">total 8\n-rw-r--r-- 1 root root 6714 Nov  1 14:08 AssertionDLApproach_037b304ea9cc.log\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["%sh tail -n 26 training_logs/AssertionDLApproach_037b304ea9cc.log"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"f0fa1137-1cab-4731-b92a-70fd5f56aa84"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">present\t 96\t 46\t 11\t 0.6760563\t 0.89719623\t 0.7710843\nconditional\t 0\t 0\t 1\t 0.0\t 0.0\t 0.0\nassociated_with_someone_else\t 0\t 0\t 3\t 0.0\t 0.0\t 0.0\nhypothetical\t 0\t 0\t 8\t 0.0\t 0.0\t 0.0\npossible\t 0\t 0\t 5\t 0.0\t 0.0\t 0.0\ntp: 102 fp: 57 fn: 57 labels: 6\nMacro-average\t prec: 0.1714996, rec: 0.17810415, f1: 0.1747395\nMicro-average\t prec: 0.6415094, rec: 0.6415094, f1: 0.6415094\n\n\nEpoch: 9 started, learning rate: 6.302493E-4, dataset size: 637\nDone, 4.794174729 loss: 4.8547945, batches: 5\nQuality on validation dataset (20.0%), validation examples = 159\ntime to finish evaluation: 1.06s\nlabel\t tp\t fp\t fn\t prec\t rec\t f1\nabsent\t 5\t 11\t 30\t 0.3125\t 0.14285715\t 0.19607843\npresent\t 96\t 47\t 11\t 0.67132866\t 0.89719623\t 0.76800007\nconditional\t 0\t 0\t 1\t 0.0\t 0.0\t 0.0\nassociated_with_someone_else\t 0\t 0\t 3\t 0.0\t 0.0\t 0.0\nhypothetical\t 0\t 0\t 8\t 0.0\t 0.0\t 0.0\npossible\t 0\t 0\t 5\t 0.0\t 0.0\t 0.0\ntp: 101 fp: 58 fn: 58 labels: 6\nMacro-average\t prec: 0.16397144, rec: 0.17334223, f1: 0.16852668\nMicro-average\t prec: 0.6352201, rec: 0.6352201, f1: 0.6352201\n\n\n</div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">present\t 96\t 46\t 11\t 0.6760563\t 0.89719623\t 0.7710843\nconditional\t 0\t 0\t 1\t 0.0\t 0.0\t 0.0\nassociated_with_someone_else\t 0\t 0\t 3\t 0.0\t 0.0\t 0.0\nhypothetical\t 0\t 0\t 8\t 0.0\t 0.0\t 0.0\npossible\t 0\t 0\t 5\t 0.0\t 0.0\t 0.0\ntp: 102 fp: 57 fn: 57 labels: 6\nMacro-average\t prec: 0.1714996, rec: 0.17810415, f1: 0.1747395\nMicro-average\t prec: 0.6415094, rec: 0.6415094, f1: 0.6415094\n\n\nEpoch: 9 started, learning rate: 6.302493E-4, dataset size: 637\nDone, 4.794174729 loss: 4.8547945, batches: 5\nQuality on validation dataset (20.0%), validation examples = 159\ntime to finish evaluation: 1.06s\nlabel\t tp\t fp\t fn\t prec\t rec\t f1\nabsent\t 5\t 11\t 30\t 0.3125\t 0.14285715\t 0.19607843\npresent\t 96\t 47\t 11\t 0.67132866\t 0.89719623\t 0.76800007\nconditional\t 0\t 0\t 1\t 0.0\t 0.0\t 0.0\nassociated_with_someone_else\t 0\t 0\t 3\t 0.0\t 0.0\t 0.0\nhypothetical\t 0\t 0\t 8\t 0.0\t 0.0\t 0.0\npossible\t 0\t 0\t 5\t 0.0\t 0.0\t 0.0\ntp: 101 fp: 58 fn: 58 labels: 6\nMacro-average\t prec: 0.16397144, rec: 0.17334223, f1: 0.16852668\nMicro-average\t prec: 0.6352201, rec: 0.6352201, f1: 0.6352201\n\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["preds = assertion_model.transform(test_data).select('label','assertion.result')\n\npreds.show()"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"0bda53e4-8351-4171-b994-8e9f93b2e93c"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[{"name":"preds","typeStr":"pyspark.sql.dataframe.DataFrame","schema":{"fields":[{"metadata":{},"name":"label","nullable":true,"type":"string"},{"metadata":{},"name":"result","nullable":true,"type":{"containsNull":true,"elementType":"string","type":"array"}}],"type":"struct"},"tableIdentifier":null}],"data":"<div class=\"ansiout\">+------------+---------+\n|       label|   result|\n+------------+---------+\n|     present|[present]|\n|    possible|[present]|\n|     present|[present]|\n|     present|[present]|\n|     present| [absent]|\n|     present|[present]|\n|     present| [absent]|\n|     present|[present]|\n|     present| [absent]|\n|     present|[present]|\n|     present|[present]|\n|hypothetical|[present]|\n|     present|[present]|\n|     present|[present]|\n|     present|[present]|\n|     present| [absent]|\n|     present|[present]|\n|     present|[present]|\n|hypothetical|[present]|\n|     present| [absent]|\n+------------+---------+\nonly showing top 20 rows\n\n</div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">+------------+---------+\n       label|   result|\n+------------+---------+\n     present|[present]|\n    possible|[present]|\n     present|[present]|\n     present|[present]|\n     present| [absent]|\n     present|[present]|\n     present| [absent]|\n     present|[present]|\n     present| [absent]|\n     present|[present]|\n     present|[present]|\nhypothetical|[present]|\n     present|[present]|\n     present|[present]|\n     present|[present]|\n     present| [absent]|\n     present|[present]|\n     present|[present]|\nhypothetical|[present]|\n     present| [absent]|\n+------------+---------+\nonly showing top 20 rows\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["preds_df = preds.toPandas()\n"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"43bcfc6a-4c3a-40b8-8b26-5da14ce8682a"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["preds_df['result'] = preds_df['result'].apply(lambda x : x[0])\n"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"69926665-71df-4ee2-ab7e-bc70099feeba"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["# We are going to use sklearn to evalute the results on test dataset\nfrom sklearn.metrics import classification_report\n\nprint (classification_report(preds_df['result'], preds_df['label']))"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"d5d83d2b-95a4-4a14-a356-f85201bc641d"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">/databricks/python/lib/python3.7/site-packages/sklearn/metrics/classification.py:1145: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples.\n  &#39;recall&#39;, &#39;true&#39;, average, warn_for)\n                              precision    recall  f1-score   support\n\n                      absent       0.25      0.41      0.31        27\nassociated_with_someone_else       0.00      0.00      0.00         0\n                 conditional       0.00      0.00      0.00         0\n                hypothetical       0.00      0.00      0.00         0\n                    possible       0.00      0.00      0.00         0\n                     present       0.88      0.67      0.76       177\n\n                   micro avg       0.64      0.64      0.64       204\n                   macro avg       0.19      0.18      0.18       204\n                weighted avg       0.80      0.64      0.70       204\n\n</div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">/databricks/python/lib/python3.7/site-packages/sklearn/metrics/classification.py:1145: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples.\n  &#39;recall&#39;, &#39;true&#39;, average, warn_for)\n                              precision    recall  f1-score   support\n\n                      absent       0.25      0.41      0.31        27\nassociated_with_someone_else       0.00      0.00      0.00         0\n                 conditional       0.00      0.00      0.00         0\n                hypothetical       0.00      0.00      0.00         0\n                    possible       0.00      0.00      0.00         0\n                     present       0.88      0.67      0.76       177\n\n                   micro avg       0.64      0.64      0.64       204\n                   macro avg       0.19      0.18      0.18       204\n                weighted avg       0.80      0.64      0.70       204\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["assertion_model.stages[-1].write().overwrite().save('/databricks/driver/models/assertion_model_0804')"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"ad145896-0369-40b2-8c05-85351675e7a0"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0}],"metadata":{"application/vnd.databricks.v1+notebook":{"notebookName":"2.Clinical_Assertion_Model_v2.7.0","dashboards":[],"language":"python","widgets":{},"notebookOrigID":4217130038628498}},"nbformat":4,"nbformat_minor":0}
