{"cells":[{"cell_type":"markdown","source":["![JohnSnowLabs](https://nlp.johnsnowlabs.com/assets/images/logo.png)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"aec851fe-e39a-4743-b983-9930d110e2c6"}}},{"cell_type":"markdown","source":["# Training and Reusing Assertion Status Models"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"8e5732f0-beec-44b9-b4e5-42023927f9d4"}}},{"cell_type":"code","source":["import os\nimport json\nimport string\nimport numpy as np\nimport pandas as pd\n\n\nimport sparknlp\nimport sparknlp_jsl\nfrom sparknlp.base import *\nfrom sparknlp.util import *\nfrom sparknlp.annotator import *\nfrom sparknlp_jsl.annotator import *\nfrom sparknlp.pretrained import ResourceDownloader\n\nfrom pyspark.sql import functions as F\nfrom pyspark.ml import Pipeline, PipelineModel\n\npd.set_option('max_colwidth', 100)\npd.set_option('display.max_columns', 100)  \npd.set_option('display.expand_frame_repr', False)\n\nprint('sparknlp.version : ',sparknlp.version())\nprint('sparknlp_jsl.version : ',sparknlp_jsl.version())\n\nspark"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"c0f3e14c-398a-4ecd-864c-3b1c939acafd"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">sparknlp.version :  3.4.0\nsparknlp_jsl.version :  3.4.0\nOut[1]: </div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">sparknlp.version :  3.4.0\nsparknlp_jsl.version :  3.4.0\nOut[1]: </div>"]}},{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"\n            <div>\n                <p><b>SparkSession - hive</b></p>\n                \n        <div>\n            <p><b>SparkContext</b></p>\n\n            <p><a href=\"/?o=7956323724731612#setting/sparkui/0129-132130-cmonouy4/driver-5052836743705468136\">Spark UI</a></p>\n\n            <dl>\n              <dt>Version</dt>\n                <dd><code>v3.2.0</code></dd>\n              <dt>Master</dt>\n                <dd><code>spark://10.139.64.7:7077</code></dd>\n              <dt>AppName</dt>\n                <dd><code>Databricks Shell</code></dd>\n            </dl>\n        </div>\n        \n            </div>\n        ","textData":null,"removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"htmlSandbox","arguments":{}}},"output_type":"display_data","data":{"text/html":["\n            <div>\n                <p><b>SparkSession - hive</b></p>\n                \n        <div>\n            <p><b>SparkContext</b></p>\n\n            <p><a href=\"/?o=7956323724731612#setting/sparkui/0129-132130-cmonouy4/driver-5052836743705468136\">Spark UI</a></p>\n\n            <dl>\n              <dt>Version</dt>\n                <dd><code>v3.2.0</code></dd>\n              <dt>Master</dt>\n                <dd><code>spark://10.139.64.7:7077</code></dd>\n              <dt>AppName</dt>\n                <dd><code>Databricks Shell</code></dd>\n            </dl>\n        </div>\n        \n            </div>\n        "]}}],"execution_count":0},{"cell_type":"markdown","source":["# Clinical Assertion Model (with pretrained models)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"7868685d-d1a9-4896-8527-45da3ec252e7"}}},{"cell_type":"markdown","source":["## Pretrained `assertion_dl` model"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"09425c6d-5c14-4fa3-a719-98e20ab6c21e"}}},{"cell_type":"code","source":["# Annotator that transforms a text column from dataframe into an Annotation ready for NLP\n\nfrom sparknlp_jsl.annotator import *\n\ndocumentAssembler = DocumentAssembler()\\\n  .setInputCol(\"text\")\\\n  .setOutputCol(\"document\")\n\n# Sentence Detector annotator, processes various sentences per line\n\nsentenceDetector = SentenceDetector()\\\n  .setInputCols([\"document\"])\\\n  .setOutputCol(\"sentence\")\n\n# Tokenizer splits words in a relevant format for NLP\n\ntokenizer = Tokenizer()\\\n  .setInputCols([\"sentence\"])\\\n  .setOutputCol(\"token\")\n\n# Clinical word embeddings trained on PubMED dataset\nword_embeddings = WordEmbeddingsModel.pretrained(\"embeddings_clinical\", \"en\", \"clinical/models\")\\\n  .setInputCols([\"sentence\", \"token\"])\\\n  .setOutputCol(\"embeddings\")\n\n# NER model trained on i2b2 (sampled from MIMIC) dataset\nclinical_ner = MedicalNerModel.pretrained(\"ner_clinical\", \"en\", \"clinical/models\") \\\n  .setInputCols([\"sentence\", \"token\", \"embeddings\"]) \\\n  .setOutputCol(\"ner\")\n\nner_converter = NerConverter() \\\n  .setInputCols([\"sentence\", \"token\", \"ner\"]) \\\n  .setOutputCol(\"ner_chunk\")\n\n# Assertion model trained on i2b2 (sampled from MIMIC) dataset\n# coming from sparknlp_jsl.annotator !!\nclinical_assertion = AssertionDLModel.pretrained(\"assertion_dl\", \"en\", \"clinical/models\") \\\n  .setInputCols([\"sentence\", \"ner_chunk\", \"embeddings\"]) \\\n  .setOutputCol(\"assertion\")\n\nnlpPipeline = Pipeline(stages=[\n    documentAssembler, \n    sentenceDetector,\n    tokenizer,\n    word_embeddings,\n    clinical_ner,\n    ner_converter,\n    clinical_assertion\n    ])\n\nempty_data = spark.createDataFrame([[\"\"]]).toDF(\"text\")\nmodel = nlpPipeline.fit(empty_data)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"064d576d-3e0b-497b-810d-c6e8c363a0c5"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">embeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[OK!]\nner_clinical download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[ / ]\r[OK!]\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">embeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[ | ]\r[ / ]\r[OK!]\nner_clinical download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[ / ]\r[ — ]\r[ \\ ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[ / ]\r[OK!]\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["# we also have a LogReg based Assertion Model.\n'''\nclinical_assertion_ml = AssertionLogRegModel.pretrained(\"assertion_ml\", \"en\", \"clinical/models\") \\\n    .setInputCols([\"sentence\", \"ner_chunk\", \"embeddings\"]) \\\n    .setOutputCol(\"assertion\")\n'''"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"0bede1e8-a98e-4774-ad2e-5863b56bd636"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">Out[4]: &#39;\\nclinical_assertion_ml = AssertionLogRegModel.pretrained(&#34;assertion_ml&#34;, &#34;en&#34;, &#34;clinical/models&#34;)     .setInputCols([&#34;sentence&#34;, &#34;ner_chunk&#34;, &#34;embeddings&#34;])     .setOutputCol(&#34;assertion&#34;)\\n&#39;</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Out[4]: &#39;\\nclinical_assertion_ml = AssertionLogRegModel.pretrained(&#34;assertion_ml&#34;, &#34;en&#34;, &#34;clinical/models&#34;)     .setInputCols([&#34;sentence&#34;, &#34;ner_chunk&#34;, &#34;embeddings&#34;])     .setOutputCol(&#34;assertion&#34;)\\n&#39;</div>"]}}],"execution_count":0},{"cell_type":"code","source":["import pandas as pd\n\ntext = 'Patient has a headache for the last 2 weeks and appears anxious when she walks fast. No alopecia noted. She denies pain'\n\nprint (text)\n\nlight_model = LightPipeline(model)\n\nlight_result = light_model.fullAnnotate(text)[0]\n\nchunks=[]\nentities=[]\nstatus=[]\n\nfor n,m in zip(light_result['ner_chunk'],light_result['assertion']):\n    \n    chunks.append(n.result)\n    entities.append(n.metadata['entity']) \n    status.append(m.result)\n        \ndf = pd.DataFrame({'chunks':chunks, 'entities':entities, 'assertion':status})\n\ndf"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"6137f486-8004-4553-bb12-b2f01df9ab93"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">Patient has a headache for the last 2 weeks and appears anxious when she walks fast. No alopecia noted. She denies pain\nOut[4]: </div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Patient has a headache for the last 2 weeks and appears anxious when she walks fast. No alopecia noted. She denies pain\nOut[4]: </div>"]}},{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chunks</th>\n      <th>entities</th>\n      <th>assertion</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>a headache</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>anxious</td>\n      <td>PROBLEM</td>\n      <td>conditional</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>alopecia</td>\n      <td>PROBLEM</td>\n      <td>absent</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>pain</td>\n      <td>PROBLEM</td>\n      <td>absent</td>\n    </tr>\n  </tbody>\n</table>\n</div>","textData":null,"removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"htmlSandbox","arguments":{}}},"output_type":"display_data","data":{"text/html":["<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chunks</th>\n      <th>entities</th>\n      <th>assertion</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>a headache</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>anxious</td>\n      <td>PROBLEM</td>\n      <td>conditional</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>alopecia</td>\n      <td>PROBLEM</td>\n      <td>absent</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>pain</td>\n      <td>PROBLEM</td>\n      <td>absent</td>\n    </tr>\n  </tbody>\n</table>\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["from sparknlp_display import AssertionVisualizer\n\nvisualizer = AssertionVisualizer()\n\nvis = visualizer.display(light_result, 'ner_chunk', 'assertion', return_html=True)\n#visualizer.set_label_colors({'TREATMENT':'#008080', 'PROBLEM':'#800080'})\n\n\ndisplayHTML(vis)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"352bef06-8f6b-4099-983a-2f64bb4c0c55"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"\n<style>\n    @import url('https://fonts.googleapis.com/css2?family=Montserrat:wght@300;400;500;600;700&display=swap');\n    @import url('https://fonts.googleapis.com/css2?family=Vistol Regular:wght@300;400;500;600;700&display=swap');\n    \n    .spark-nlp-display-scroll-entities {\n        border: 1px solid #E7EDF0;\n        border-radius: 3px;\n        text-align: justify;\n        \n    }\n    .spark-nlp-display-scroll-entities span {  \n        font-size: 14px;\n        line-height: 24px;\n        color: #536B76;\n        font-family: 'Montserrat', sans-serif !important;\n    }\n    \n    .spark-nlp-display-entity-wrapper{\n    \n        display: inline-grid;\n        text-align: center;\n        border-radius: 4px;\n        margin: 0 2px 5px 2px;\n        padding: 1px\n    }\n    .spark-nlp-display-entity-name{\n        font-size: 14px;\n        line-height: 24px;\n        font-family: 'Montserrat', sans-serif !important;\n        \n        background: #f1f2f3;\n        border-width: medium;\n        text-align: center;\n        \n        font-weight: 400;\n        \n        border-radius: 5px;\n        padding: 2px 5px;\n        display: block;\n        margin: 3px 2px;\n    \n    }\n    .spark-nlp-display-entity-type{\n        font-size: 14px;\n        line-height: 24px;\n        color: #ffffff;\n        font-family: 'Montserrat', sans-serif !important;\n        \n        text-transform: uppercase;\n        \n        font-weight: 500;\n\n        display: block;\n        padding: 3px 5px;\n    }\n    \n    .spark-nlp-display-entity-resolution{\n        font-size: 14px;\n        line-height: 24px;\n        color: #ffffff;\n        font-family: 'Vistol Regular', sans-serif !important;\n        \n        text-transform: uppercase;\n        \n        font-weight: 500;\n\n        display: block;\n        padding: 3px 5px;\n    }\n    \n    .spark-nlp-display-others{\n        font-size: 14px;\n        line-height: 24px;\n        font-family: 'Montserrat', sans-serif !important;\n        \n        font-weight: 400;\n    }\n\n</style>\n <span class=\"spark-nlp-display-others\" style=\"background-color: white\">Patient has </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #800080B3\"><span class=\"spark-nlp-display-entity-name\">a headache </span><span class=\"spark-nlp-display-entity-type\">PROBLEM</span><span class=\"spark-nlp-display-entity-resolution\" style=\"background-color: #800080FF\">present </span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> for the last 2 weeks and appears </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #800080B3\"><span class=\"spark-nlp-display-entity-name\">anxious </span><span class=\"spark-nlp-display-entity-type\">PROBLEM</span><span class=\"spark-nlp-display-entity-resolution\" style=\"background-color: #800080FF\">conditional </span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> when she walks fast. No </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #800080B3\"><span class=\"spark-nlp-display-entity-name\">alopecia </span><span class=\"spark-nlp-display-entity-type\">PROBLEM</span><span class=\"spark-nlp-display-entity-resolution\" style=\"background-color: #800080FF\">absent </span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> noted. She denies </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #800080B3\"><span class=\"spark-nlp-display-entity-name\">pain </span><span class=\"spark-nlp-display-entity-type\">PROBLEM</span><span class=\"spark-nlp-display-entity-resolution\" style=\"background-color: #800080FF\">absent </span></span></div>","textData":null,"removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"htmlSandbox","arguments":{}}},"output_type":"display_data","data":{"text/html":["\n<style>\n    @import url('https://fonts.googleapis.com/css2?family=Montserrat:wght@300;400;500;600;700&display=swap');\n    @import url('https://fonts.googleapis.com/css2?family=Vistol Regular:wght@300;400;500;600;700&display=swap');\n    \n    .spark-nlp-display-scroll-entities {\n        border: 1px solid #E7EDF0;\n        border-radius: 3px;\n        text-align: justify;\n        \n    }\n    .spark-nlp-display-scroll-entities span {  \n        font-size: 14px;\n        line-height: 24px;\n        color: #536B76;\n        font-family: 'Montserrat', sans-serif !important;\n    }\n    \n    .spark-nlp-display-entity-wrapper{\n    \n        display: inline-grid;\n        text-align: center;\n        border-radius: 4px;\n        margin: 0 2px 5px 2px;\n        padding: 1px\n    }\n    .spark-nlp-display-entity-name{\n        font-size: 14px;\n        line-height: 24px;\n        font-family: 'Montserrat', sans-serif !important;\n        \n        background: #f1f2f3;\n        border-width: medium;\n        text-align: center;\n        \n        font-weight: 400;\n        \n        border-radius: 5px;\n        padding: 2px 5px;\n        display: block;\n        margin: 3px 2px;\n    \n    }\n    .spark-nlp-display-entity-type{\n        font-size: 14px;\n        line-height: 24px;\n        color: #ffffff;\n        font-family: 'Montserrat', sans-serif !important;\n        \n        text-transform: uppercase;\n        \n        font-weight: 500;\n\n        display: block;\n        padding: 3px 5px;\n    }\n    \n    .spark-nlp-display-entity-resolution{\n        font-size: 14px;\n        line-height: 24px;\n        color: #ffffff;\n        font-family: 'Vistol Regular', sans-serif !important;\n        \n        text-transform: uppercase;\n        \n        font-weight: 500;\n\n        display: block;\n        padding: 3px 5px;\n    }\n    \n    .spark-nlp-display-others{\n        font-size: 14px;\n        line-height: 24px;\n        font-family: 'Montserrat', sans-serif !important;\n        \n        font-weight: 400;\n    }\n\n</style>\n <span class=\"spark-nlp-display-others\" style=\"background-color: white\">Patient has </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #800080B3\"><span class=\"spark-nlp-display-entity-name\">a headache </span><span class=\"spark-nlp-display-entity-type\">PROBLEM</span><span class=\"spark-nlp-display-entity-resolution\" style=\"background-color: #800080FF\">present </span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> for the last 2 weeks and appears </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #800080B3\"><span class=\"spark-nlp-display-entity-name\">anxious </span><span class=\"spark-nlp-display-entity-type\">PROBLEM</span><span class=\"spark-nlp-display-entity-resolution\" style=\"background-color: #800080FF\">conditional </span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> when she walks fast. No </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #800080B3\"><span class=\"spark-nlp-display-entity-name\">alopecia </span><span class=\"spark-nlp-display-entity-type\">PROBLEM</span><span class=\"spark-nlp-display-entity-resolution\" style=\"background-color: #800080FF\">absent </span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> noted. She denies </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #800080B3\"><span class=\"spark-nlp-display-entity-name\">pain </span><span class=\"spark-nlp-display-entity-type\">PROBLEM</span><span class=\"spark-nlp-display-entity-resolution\" style=\"background-color: #800080FF\">absent </span></span></div>"]}}],"execution_count":0},{"cell_type":"code","source":["! wget -q\thttps://s3.amazonaws.com/auxdata.johnsnowlabs.com/public/resources/en/pubmed/pubmed_sample_text_small.csv -P /dbfs/"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"f1e1c581-7c0e-4b56-be6e-d604278e3d18"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["import pyspark.sql.functions as F\n\npubMedDF = spark.read\\\n                .option(\"header\", \"true\")\\\n                .csv(\"/pubmed_sample_text_small.csv\")\\\n                \npubMedDF.show(truncate=50)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"1321721d-d896-4179-87c9-6716a68ccba4"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">+--------------------------------------------------+\n|                                              text|\n+--------------------------------------------------+\n|The human KCNJ9 (Kir 3.3, GIRK3) is a member of...|\n|BACKGROUND: At present, it is one of the most i...|\n|OBJECTIVE: To investigate the relationship betw...|\n|Combined EEG/fMRI recording has been used to lo...|\n|Kohlschutter syndrome is a rare neurodegenerati...|\n|Statistical analysis of neuroimages is commonly...|\n|The synthetic DOX-LNA conjugate was characteriz...|\n|Our objective was to compare three different me...|\n|We conducted a phase II study to assess the eff...|\n|&#34;&#34;&#34;Monomeric sarcosine oxidase (MSOX) is a flav...|\n|We presented the tachinid fly Exorista japonica...|\n|The literature dealing with the water conductin...|\n|A novel approach to synthesize chitosan-O-isopr...|\n|An HPLC-ESI-MS-MS method has been developed for...|\n|The localizing and lateralizing values of eye a...|\n|OBJECTIVE: To evaluate the effectiveness and ac...|\n|For the construction of new combinatorial libra...|\n|We report the results of a screen for genetic a...|\n|Intraparenchymal pericatheter cyst is rarely re...|\n|It is known that patients with Klinefelter&#39;s sy...|\n+--------------------------------------------------+\nonly showing top 20 rows\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">+--------------------------------------------------+\n                                              text|\n+--------------------------------------------------+\nThe human KCNJ9 (Kir 3.3, GIRK3) is a member of...|\nBACKGROUND: At present, it is one of the most i...|\nOBJECTIVE: To investigate the relationship betw...|\nCombined EEG/fMRI recording has been used to lo...|\nKohlschutter syndrome is a rare neurodegenerati...|\nStatistical analysis of neuroimages is commonly...|\nThe synthetic DOX-LNA conjugate was characteriz...|\nOur objective was to compare three different me...|\nWe conducted a phase II study to assess the eff...|\n&#34;&#34;&#34;Monomeric sarcosine oxidase (MSOX) is a flav...|\nWe presented the tachinid fly Exorista japonica...|\nThe literature dealing with the water conductin...|\nA novel approach to synthesize chitosan-O-isopr...|\nAn HPLC-ESI-MS-MS method has been developed for...|\nThe localizing and lateralizing values of eye a...|\nOBJECTIVE: To evaluate the effectiveness and ac...|\nFor the construction of new combinatorial libra...|\nWe report the results of a screen for genetic a...|\nIntraparenchymal pericatheter cyst is rarely re...|\nIt is known that patients with Klinefelter&#39;s sy...|\n+--------------------------------------------------+\nonly showing top 20 rows\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["result = model.transform(pubMedDF.limit(100))"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"dd660faf-5874-42eb-99ef-a083952d91ab"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["result.show()"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"7bf3ef67-9448-43e4-9d67-d7a47399911a"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+\n|                text|            document|            sentence|               token|          embeddings|                 ner|           ner_chunk|           assertion|\n+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+\n|A numerical metho...|[{document, 0, 88...|[{document, 0, 15...|[{token, 0, 0, A,...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 69, 95, ...|[{assertion, 69, ...|\n|Recording of the ...|[{document, 0, 44...|[{document, 0, 10...|[{token, 0, 8, Re...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 13, 30, ...|[{assertion, 13, ...|\n|Colicin E5 cleave...|[{document, 0, 12...|[{document, 0, 11...|[{token, 0, 6, Co...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 0, 9, Co...|[{assertion, 0, 9...|\n|Recombinant adeno...|[{document, 0, 13...|[{document, 0, 20...|[{token, 0, 10, R...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 0, 28, R...|[{assertion, 0, 2...|\n|To understand its...|[{document, 0, 29...|[{document, 0, 19...|[{token, 0, 1, To...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 37, 52, ...|[{assertion, 37, ...|\n|OBJECTIVE: To rev...|[{document, 0, 20...|[{document, 0, 20...|[{token, 0, 8, OB...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 63, 91, ...|[{assertion, 63, ...|\n|AIM: Understandin...|[{document, 0, 15...|[{document, 0, 16...|[{token, 0, 2, AI...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 46, 74, ...|[{assertion, 46, ...|\n|In the present st...|[{document, 0, 92...|[{document, 0, 21...|[{token, 0, 1, In...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 3, 19, t...|[{assertion, 3, 1...|\n|AIM: To describe ...|[{document, 0, 14...|[{document, 0, 20...|[{token, 0, 2, AI...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 17, 32, ...|[{assertion, 17, ...|\n|Although the risk...|[{document, 0, 12...|[{document, 0, 11...|[{token, 0, 7, Al...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 118, 163...|[{assertion, 118,...|\n|Successful long-t...|[{document, 0, 16...|[{document, 0, 12...|[{token, 0, 9, Su...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 0, 43, S...|[{assertion, 0, 4...|\n|Pulmonary alveola...|[{document, 0, 16...|[{document, 0, 15...|[{token, 0, 8, Pu...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 68, 94, ...|[{assertion, 68, ...|\n|Prenatal care is ...|[{document, 0, 26...|[{document, 0, 68...|[{token, 0, 7, Pr...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 70, 86, ...|[{assertion, 70, ...|\n|Controversy exist...|[{document, 0, 12...|[{document, 0, 18...|[{token, 0, 10, C...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 43, 66, ...|[{assertion, 43, ...|\n|Children with a f...|[{document, 0, 34...|[{document, 0, 13...|[{token, 0, 7, Ch...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 34, 46, ...|[{assertion, 34, ...|\n|The mucosal immun...|[{document, 0, 19...|[{document, 0, 12...|[{token, 0, 2, Th...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 51, 57, ...|[{assertion, 51, ...|\n|Urinary concretio...|[{document, 0, 86...|[{document, 0, 15...|[{token, 0, 6, Ur...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 0, 18, U...|[{assertion, 0, 1...|\n|In this study, th...|[{document, 0, 14...|[{document, 0, 17...|[{token, 0, 1, In...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 3, 12, t...|[{assertion, 3, 1...|\n|The presence of c...|[{document, 0, 44...|[{document, 0, 12...|[{token, 0, 2, Th...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 16, 32, ...|[{assertion, 16, ...|\n|Phosphatidylethan...|[{document, 0, 17...|[{document, 0, 26...|[{token, 0, 23, P...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 0, 48, P...|[{assertion, 0, 4...|\n+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+\nonly showing top 20 rows\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+\n                text|            document|            sentence|               token|          embeddings|                 ner|           ner_chunk|           assertion|\n+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+\nA numerical metho...|[{document, 0, 88...|[{document, 0, 15...|[{token, 0, 0, A,...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 69, 95, ...|[{assertion, 69, ...|\nRecording of the ...|[{document, 0, 44...|[{document, 0, 10...|[{token, 0, 8, Re...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 13, 30, ...|[{assertion, 13, ...|\nColicin E5 cleave...|[{document, 0, 12...|[{document, 0, 11...|[{token, 0, 6, Co...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 0, 9, Co...|[{assertion, 0, 9...|\nRecombinant adeno...|[{document, 0, 13...|[{document, 0, 20...|[{token, 0, 10, R...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 0, 28, R...|[{assertion, 0, 2...|\nTo understand its...|[{document, 0, 29...|[{document, 0, 19...|[{token, 0, 1, To...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 37, 52, ...|[{assertion, 37, ...|\nOBJECTIVE: To rev...|[{document, 0, 20...|[{document, 0, 20...|[{token, 0, 8, OB...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 63, 91, ...|[{assertion, 63, ...|\nAIM: Understandin...|[{document, 0, 15...|[{document, 0, 16...|[{token, 0, 2, AI...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 46, 74, ...|[{assertion, 46, ...|\nIn the present st...|[{document, 0, 92...|[{document, 0, 21...|[{token, 0, 1, In...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 3, 19, t...|[{assertion, 3, 1...|\nAIM: To describe ...|[{document, 0, 14...|[{document, 0, 20...|[{token, 0, 2, AI...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 17, 32, ...|[{assertion, 17, ...|\nAlthough the risk...|[{document, 0, 12...|[{document, 0, 11...|[{token, 0, 7, Al...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 118, 163...|[{assertion, 118,...|\nSuccessful long-t...|[{document, 0, 16...|[{document, 0, 12...|[{token, 0, 9, Su...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 0, 43, S...|[{assertion, 0, 4...|\nPulmonary alveola...|[{document, 0, 16...|[{document, 0, 15...|[{token, 0, 8, Pu...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 68, 94, ...|[{assertion, 68, ...|\nPrenatal care is ...|[{document, 0, 26...|[{document, 0, 68...|[{token, 0, 7, Pr...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 70, 86, ...|[{assertion, 70, ...|\nControversy exist...|[{document, 0, 12...|[{document, 0, 18...|[{token, 0, 10, C...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 43, 66, ...|[{assertion, 43, ...|\nChildren with a f...|[{document, 0, 34...|[{document, 0, 13...|[{token, 0, 7, Ch...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 34, 46, ...|[{assertion, 34, ...|\nThe mucosal immun...|[{document, 0, 19...|[{document, 0, 12...|[{token, 0, 2, Th...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 51, 57, ...|[{assertion, 51, ...|\nUrinary concretio...|[{document, 0, 86...|[{document, 0, 15...|[{token, 0, 6, Ur...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 0, 18, U...|[{assertion, 0, 1...|\nIn this study, th...|[{document, 0, 14...|[{document, 0, 17...|[{token, 0, 1, In...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 3, 12, t...|[{assertion, 3, 1...|\nThe presence of c...|[{document, 0, 44...|[{document, 0, 12...|[{token, 0, 2, Th...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 16, 32, ...|[{assertion, 16, ...|\nPhosphatidylethan...|[{document, 0, 17...|[{document, 0, 26...|[{token, 0, 23, P...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 0, 48, P...|[{assertion, 0, 4...|\n+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+\nonly showing top 20 rows\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["result.select('sentence.result').take(1)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"cd49fbf0-b7d7-4523-b08c-d02fc6271ffb"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">Out[14]: [Row(result=[&#39;A numerical method was developed using geometrical optics to predict far-fieldoptical scattering from particles that are symmetric about the optic axis.&#39;, &#39;Thediffractive component of scattering is calculated and combined with thereflective and refractive components to give the total scattering pattern.&#39;, &#39;Thephase terms of the scattered light are calculated as well.&#39;, &#39;Verification of themethod was achieved by assuming a spherical particle and comparing the results toMie scattering theory.&#39;, &#39;Agreement with the Mie theory was excellent in theforward-scattering direction.&#39;, &#39;However, small-amplitude oscillations near therainbow regions were not observed using the numerical method.&#39;, &#39;Numerical data fromspheroidal particles and hemispherical particles are also presented.&#39;, &#39;The use ofhemispherical particles as a calibration standard for intensity-type opticalparticle-sizing instruments is discussed.&#39;])]</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Out[14]: [Row(result=[&#39;A numerical method was developed using geometrical optics to predict far-fieldoptical scattering from particles that are symmetric about the optic axis.&#39;, &#39;Thediffractive component of scattering is calculated and combined with thereflective and refractive components to give the total scattering pattern.&#39;, &#39;Thephase terms of the scattered light are calculated as well.&#39;, &#39;Verification of themethod was achieved by assuming a spherical particle and comparing the results toMie scattering theory.&#39;, &#39;Agreement with the Mie theory was excellent in theforward-scattering direction.&#39;, &#39;However, small-amplitude oscillations near therainbow regions were not observed using the numerical method.&#39;, &#39;Numerical data fromspheroidal particles and hemispherical particles are also presented.&#39;, &#39;The use ofhemispherical particles as a calibration standard for intensity-type opticalparticle-sizing instruments is discussed.&#39;])]</div>"]}}],"execution_count":0},{"cell_type":"code","source":["result.select(F.explode(F.arrays_zip(result.ner_chunk.result,  result.ner_chunk.begin, result.ner_chunk.end, result.ner_chunk.metadata, \n                                     result.assertion.result)).alias(\"cols\")) \\\n      .select(F.expr(\"cols['0']\").alias(\"chunk\"),\n              F.expr(\"cols['1']\").alias(\"begin\"),\n              F.expr(\"cols['2']\").alias(\"end\"),\n              F.expr(\"cols['3']['entity']\").alias(\"ner_label\"),\n              F.expr(\"cols['3']['sentence']\").alias(\"sent_id\"),\n              F.expr(\"cols['4']\").alias(\"assertion\"),).show(truncate=False)\n\n"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"6bbadeea-6168-42ca-b500-bdaa548126d2"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">+-----------------------------------------------------------+-----+---+---------+-------+-----------+\n|chunk                                                      |begin|end|ner_label|sent_id|assertion  |\n+-----------------------------------------------------------+-----+---+---------+-------+-----------+\n|the G-protein-activated inwardly rectifying potassium (GIRK|48   |106|TREATMENT|0      |conditional|\n|the genomicorganization                                    |142  |164|TREATMENT|1      |present    |\n|a candidate gene forType II diabetes mellitus              |210  |254|PROBLEM  |1      |present    |\n|byapproximately                                            |380  |394|TREATMENT|2      |present    |\n|single nucleotide polymorphisms                            |464  |494|TREATMENT|3      |present    |\n|aVal366Ala substitution                                    |532  |554|PROBLEM  |3      |present    |\n|an 8 base-pair                                             |561  |574|PROBLEM  |3      |present    |\n|insertion/deletion                                         |581  |598|PROBLEM  |4      |present    |\n|Ourexpression studies                                      |601  |621|TEST     |5      |present    |\n|the transcript in various humantissues                     |648  |685|PROBLEM  |5      |present    |\n|fat andskeletal muscle                                     |749  |770|PROBLEM  |5      |possible   |\n|furtherstudies                                             |830  |843|PROBLEM  |6      |present    |\n|the KCNJ9 protein                                          |864  |880|TREATMENT|6      |present    |\n|evaluation                                                 |892  |901|TEST     |6      |possible   |\n|Type II diabetes                                           |940  |955|PROBLEM  |6      |present    |\n|the treatment                                              |67   |79 |TREATMENT|0      |present    |\n|breast cancer                                              |84   |96 |PROBLEM  |0      |present    |\n|the standard therapy                                       |109  |128|TREATMENT|0      |present    |\n|anthracyclines                                             |167  |180|TREATMENT|0      |present    |\n|taxanes                                                    |186  |192|TREATMENT|0      |present    |\n+-----------------------------------------------------------+-----+---+---------+-------+-----------+\nonly showing top 20 rows\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">+-----------------------------------------------------------+-----+---+---------+-------+-----------+\nchunk                                                      |begin|end|ner_label|sent_id|assertion  |\n+-----------------------------------------------------------+-----+---+---------+-------+-----------+\nthe G-protein-activated inwardly rectifying potassium (GIRK|48   |106|TREATMENT|0      |conditional|\nthe genomicorganization                                    |142  |164|TREATMENT|1      |present    |\na candidate gene forType II diabetes mellitus              |210  |254|PROBLEM  |1      |present    |\nbyapproximately                                            |380  |394|TREATMENT|2      |present    |\nsingle nucleotide polymorphisms                            |464  |494|TREATMENT|3      |present    |\naVal366Ala substitution                                    |532  |554|PROBLEM  |3      |present    |\nan 8 base-pair                                             |561  |574|PROBLEM  |3      |present    |\ninsertion/deletion                                         |581  |598|PROBLEM  |4      |present    |\nOurexpression studies                                      |601  |621|TEST     |5      |present    |\nthe transcript in various humantissues                     |648  |685|PROBLEM  |5      |present    |\nfat andskeletal muscle                                     |749  |770|PROBLEM  |5      |possible   |\nfurtherstudies                                             |830  |843|PROBLEM  |6      |present    |\nthe KCNJ9 protein                                          |864  |880|TREATMENT|6      |present    |\nevaluation                                                 |892  |901|TEST     |6      |possible   |\nType II diabetes                                           |940  |955|PROBLEM  |6      |present    |\nthe treatment                                              |67   |79 |TREATMENT|0      |present    |\nbreast cancer                                              |84   |96 |PROBLEM  |0      |present    |\nthe standard therapy                                       |109  |128|TREATMENT|0      |present    |\nanthracyclines                                             |167  |180|TREATMENT|0      |present    |\ntaxanes                                                    |186  |192|TREATMENT|0      |present    |\n+-----------------------------------------------------------+-----+---+---------+-------+-----------+\nonly showing top 20 rows\n\n</div>"]}}],"execution_count":0},{"cell_type":"markdown","source":["## Pretrained `assertion_dl_radiology` model"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"f80c498d-3329-409b-b33e-f1b020314d58"}}},{"cell_type":"code","source":["from sparknlp_jsl.annotator import *\n\ndocumentAssembler = DocumentAssembler()\\\n    .setInputCol(\"text\")\\\n    .setOutputCol(\"document\")\n\n# Sentence Detector annotator, processes various sentences per line\nsentenceDetector = SentenceDetectorDLModel\\\n    .pretrained(\"sentence_detector_dl_healthcare\",\"en\",\"clinical/models\") \\\n    .setInputCols([\"document\"]) \\\n    .setOutputCol(\"sentence\")\n\n# Tokenizer splits words in a relevant format for NLP\ntokenizer = Tokenizer()\\\n    .setInputCols([\"sentence\"])\\\n    .setOutputCol(\"token\")\n\n# Clinical word embeddings trained on PubMED dataset\nword_embeddings = WordEmbeddingsModel.pretrained(\"embeddings_clinical\", \"en\", \"clinical/models\")\\\n    .setInputCols([\"sentence\", \"token\"])\\\n    .setOutputCol(\"embeddings\")\n\n# NER model for radiology\nradiology_ner = MedicalNerModel.pretrained(\"ner_radiology\", \"en\", \"clinical/models\") \\\n    .setInputCols([\"sentence\", \"token\", \"embeddings\"]) \\\n    .setOutputCol(\"ner\")\n\nner_converter = NerConverter() \\\n    .setInputCols([\"sentence\", \"token\", \"ner\"]) \\\n    .setOutputCol(\"ner_chunk\")\\\n    .setWhiteList([\"ImagingFindings\"])\n\n# Assertion model trained on radiology dataset\n# coming from sparknlp_jsl.annotator !!\n\nradiology_assertion = AssertionDLModel.pretrained(\"assertion_dl_radiology\", \"en\", \"clinical/models\") \\\n    .setInputCols([\"sentence\", \"ner_chunk\", \"embeddings\"]) \\\n    .setOutputCol(\"assertion\")\n\nnlpPipeline = Pipeline(stages=[\n    documentAssembler, \n    sentenceDetector,\n    tokenizer,\n    word_embeddings,\n    radiology_ner,\n    ner_converter,\n    radiology_assertion\n    ])\n\nempty_data = spark.createDataFrame([[\"\"]]).toDF(\"text\")\nradiologyAssertion_model = nlpPipeline.fit(empty_data)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"565efb2d-2d61-4965-8b3c-81d75dd3c93e"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">sentence_detector_dl_healthcare download started this may take some time.\nApproximate size to download 367.3 KB\n\r[ | ]\r[ / ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\nner_radiology download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[ / ]\r[ — ]\r[OK!]\nassertion_dl_radiology download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[ / ]\r[ — ]\r[OK!]\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">sentence_detector_dl_healthcare download started this may take some time.\nApproximate size to download 367.3 KB\n\r[ | ]\r[ / ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\nner_radiology download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[ / ]\r[ — ]\r[OK!]\nassertion_dl_radiology download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[ / ]\r[ — ]\r[OK!]\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["# A sample text from a radiology report\n\ntext = \"\"\"No right-sided pleural effusion or pneumothorax is definitively seen and there are mildly displaced fractures of the left lateral 8th and likely 9th ribs.\"\"\""],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"5910a402-1986-4513-9d70-65d67aefa0a6"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["data = spark.createDataFrame([[text]]).toDF(\"text\")"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"9aa1ee2f-efc2-441a-8836-0377e5f1eb41"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["result = radiologyAssertion_model.transform(data)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"04cd2f55-bfb5-4063-83f7-b16622ff1491"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["result.select(F.explode(F.arrays_zip(result.ner_chunk.result, result.ner_chunk.metadata, result.assertion.result)).alias(\"cols\")) \\\n      .select(F.expr(\"cols['0']\").alias(\"chunk\"),\n              F.expr(\"cols['1']['entity']\").alias(\"ner_label\"),\n              F.expr(\"cols['1']['sentence']\").alias(\"sent_id\"),\n              F.expr(\"cols['2']\").alias(\"assertion\")).show(truncate=False)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"9de0b7bb-0516-4b76-9241-ffae0acd5080"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">+-------------------+---------------+-------+---------+\n|chunk              |ner_label      |sent_id|assertion|\n+-------------------+---------------+-------+---------+\n|effusion           |ImagingFindings|0      |Negative |\n|pneumothorax       |ImagingFindings|0      |Negative |\n|displaced fractures|ImagingFindings|0      |Confirmed|\n+-------------------+---------------+-------+---------+\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">+-------------------+---------------+-------+---------+\nchunk              |ner_label      |sent_id|assertion|\n+-------------------+---------------+-------+---------+\neffusion           |ImagingFindings|0      |Negative |\npneumothorax       |ImagingFindings|0      |Negative |\ndisplaced fractures|ImagingFindings|0      |Confirmed|\n+-------------------+---------------+-------+---------+\n\n</div>"]}}],"execution_count":0},{"cell_type":"markdown","source":["## Writing a generic Assertion + NER function"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"afef480c-1540-4acc-88c9-5711e5290f3b"}}},{"cell_type":"code","source":["from pyspark.sql.functions import monotonically_increasing_id\n\n\ndef get_base_pipeline (embeddings = 'embeddings_clinical'):\n\n    documentAssembler = DocumentAssembler()\\\n        .setInputCol(\"text\")\\\n        .setOutputCol(\"document\")\n\n  # Sentence Detector annotator, processes various sentences per line\n\n    sentenceDetector = SentenceDetector()\\\n        .setInputCols([\"document\"])\\\n        .setOutputCol(\"sentence\")\n\n  # Tokenizer splits words in a relevant format for NLP\n\n    tokenizer = Tokenizer()\\\n        .setInputCols([\"sentence\"])\\\n        .setOutputCol(\"token\")\n\n  # Clinical word embeddings trained on PubMED dataset\n    word_embeddings = WordEmbeddingsModel.pretrained(embeddings, \"en\", \"clinical/models\")\\\n        .setInputCols([\"sentence\", \"token\"])\\\n        .setOutputCol(\"embeddings\")\n\n    base_pipeline = Pipeline(stages=[\n                        documentAssembler,\n                        sentenceDetector,\n                        tokenizer,\n                        word_embeddings])\n\n    return base_pipeline\n\n\n\ndef get_clinical_assertion (embeddings, spark_df, nrows = 100, model_name = 'ner_clinical'):\n\n  # NER model trained on i2b2 (sampled from MIMIC) dataset\n    loaded_ner_model = MedicalNerModel.pretrained(model_name, \"en\", \"clinical/models\") \\\n        .setInputCols([\"sentence\", \"token\", \"embeddings\"]) \\\n        .setOutputCol(\"ner\")\n\n    ner_converter = NerConverter() \\\n        .setInputCols([\"sentence\", \"token\", \"ner\"]) \\\n        .setOutputCol(\"ner_chunk\")\n\n  # Assertion model trained on i2b2 (sampled from MIMIC) dataset\n  # coming from sparknlp_jsl.annotator !!\n    clinical_assertion = AssertionDLModel.pretrained(\"assertion_dl\", \"en\", \"clinical/models\") \\\n        .setInputCols([\"sentence\", \"ner_chunk\", \"embeddings\"]) \\\n        .setOutputCol(\"assertion\")\n      \n\n    base_model = get_base_pipeline (embeddings)\n\n    nlpPipeline = Pipeline(stages=[\n        base_model,\n        loaded_ner_model,\n        ner_converter,\n        clinical_assertion])\n\n    empty_data = spark.createDataFrame([[\"\"]]).toDF(\"text\")\n\n    model = nlpPipeline.fit(empty_data)\n\n    result = model.transform(spark_df.limit(nrows))\n\n    result = result.withColumn(\"id\", monotonically_increasing_id())\n\n    result_df = result.select(F.explode(F.arrays_zip(result.ner_chunk.result, result.ner_chunk.metadata, result.assertion.result)).alias(\"cols\")) \\\n                      .select(F.expr(\"cols['0']\").alias(\"chunk\"),\n                              F.expr(\"cols['1']['entity']\").alias(\"ner_label\"),\n                              F.expr(\"cols['2']\").alias(\"assertion\"))\\\n                              .filter(\"ner_label!='O'\")\n\n    return result_df"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"ac7bc5e4-45ed-4fb6-91b8-4581865fdb5b"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["embeddings = 'embeddings_clinical'\n\nmodel_name = 'ner_clinical'\n\nnrows = 1000\n\nner_df = get_clinical_assertion (embeddings, pubMedDF, nrows, model_name)\n\nner_df.show()"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"5cc9a0fc-5187-4e7b-9fee-283b5fd343c9"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">ner_clinical download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n+--------------------+---------+-----------+\n|               chunk|ner_label|  assertion|\n+--------------------+---------+-----------+\n|the G-protein-act...|TREATMENT|conditional|\n|the genomicorgani...|TREATMENT|    present|\n|a candidate gene ...|  PROBLEM|    present|\n|     byapproximately|TREATMENT|    present|\n|single nucleotide...|TREATMENT|    present|\n|aVal366Ala substi...|  PROBLEM|    present|\n|      an 8 base-pair|  PROBLEM|    present|\n|  insertion/deletion|  PROBLEM|    present|\n|Ourexpression stu...|     TEST|    present|\n|the transcript in...|  PROBLEM|    present|\n|fat andskeletal m...|  PROBLEM|   possible|\n|      furtherstudies|  PROBLEM|    present|\n|   the KCNJ9 protein|TREATMENT|    present|\n|          evaluation|     TEST|   possible|\n|    Type II diabetes|  PROBLEM|    present|\n|       the treatment|TREATMENT|    present|\n|       breast cancer|  PROBLEM|    present|\n|the standard therapy|TREATMENT|    present|\n|      anthracyclines|TREATMENT|    present|\n|             taxanes|TREATMENT|    present|\n+--------------------+---------+-----------+\nonly showing top 20 rows\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">ner_clinical download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n+--------------------+---------+-----------+\n               chunk|ner_label|  assertion|\n+--------------------+---------+-----------+\nthe G-protein-act...|TREATMENT|conditional|\nthe genomicorgani...|TREATMENT|    present|\na candidate gene ...|  PROBLEM|    present|\n     byapproximately|TREATMENT|    present|\nsingle nucleotide...|TREATMENT|    present|\naVal366Ala substi...|  PROBLEM|    present|\n      an 8 base-pair|  PROBLEM|    present|\n  insertion/deletion|  PROBLEM|    present|\nOurexpression stu...|     TEST|    present|\nthe transcript in...|  PROBLEM|    present|\nfat andskeletal m...|  PROBLEM|   possible|\n      furtherstudies|  PROBLEM|    present|\n   the KCNJ9 protein|TREATMENT|    present|\n          evaluation|     TEST|   possible|\n    Type II diabetes|  PROBLEM|    present|\n       the treatment|TREATMENT|    present|\n       breast cancer|  PROBLEM|    present|\nthe standard therapy|TREATMENT|    present|\n      anthracyclines|TREATMENT|    present|\n             taxanes|TREATMENT|    present|\n+--------------------+---------+-----------+\nonly showing top 20 rows\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["embeddings = 'embeddings_clinical'\n\nmodel_name = 'ner_posology'\n\nnrows = 100\n\nner_df = get_clinical_assertion (embeddings, pubMedDF, nrows, model_name)\n\nner_df.show()"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"cb696bfc-6a34-4666-92e9-607683a57506"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">ner_posology download started this may take some time.\nApproximate size to download 13.8 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n+----------------+---------+---------+\n|           chunk|ner_label|assertion|\n+----------------+---------+---------+\n|  anthracyclines|     DRUG|  present|\n|         taxanes|     DRUG|  present|\n|     vinorelbine|     DRUG|  present|\n|     vinorelbine|     DRUG|  present|\n|  anthracyclines|     DRUG|  present|\n|         taxanes|     DRUG|  present|\n|  Vinorelbinewas|     DRUG|   absent|\n|       25 mg/m(2| STRENGTH|  present|\n|   intravenously|    ROUTE|   absent|\n|         on days|FREQUENCY|  present|\n| thatvinorelbine|     DRUG|  present|\n|  anthracyclines|     DRUG|  present|\n|         taxanes|     DRUG|  present|\n|             DOX|     DRUG|   absent|\n|    trandolapril|     DRUG| possible|\n|        losartan|     DRUG|  present|\n|          3-week| DURATION|   absent|\n|         placebo|     DRUG|  present|\n|2 mgtrandolapril| STRENGTH|   absent|\n|           50 mg| STRENGTH| possible|\n+----------------+---------+---------+\nonly showing top 20 rows\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">ner_posology download started this may take some time.\nApproximate size to download 13.8 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n+----------------+---------+---------+\n           chunk|ner_label|assertion|\n+----------------+---------+---------+\n  anthracyclines|     DRUG|  present|\n         taxanes|     DRUG|  present|\n     vinorelbine|     DRUG|  present|\n     vinorelbine|     DRUG|  present|\n  anthracyclines|     DRUG|  present|\n         taxanes|     DRUG|  present|\n  Vinorelbinewas|     DRUG|   absent|\n       25 mg/m(2| STRENGTH|  present|\n   intravenously|    ROUTE|   absent|\n         on days|FREQUENCY|  present|\n thatvinorelbine|     DRUG|  present|\n  anthracyclines|     DRUG|  present|\n         taxanes|     DRUG|  present|\n             DOX|     DRUG|   absent|\n    trandolapril|     DRUG| possible|\n        losartan|     DRUG|  present|\n          3-week| DURATION|   absent|\n         placebo|     DRUG|  present|\n2 mgtrandolapril| STRENGTH|   absent|\n           50 mg| STRENGTH| possible|\n+----------------+---------+---------+\nonly showing top 20 rows\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["embeddings = 'embeddings_clinical'\n\nmodel_name = 'ner_posology_greedy'\n\nentry_data = spark.createDataFrame([[\"The patient did not take a capsule of Advil\"]]).toDF(\"text\")\n\nner_df = get_clinical_assertion (embeddings, entry_data, nrows, model_name)\n\nner_df.show()"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"93fe3c92-4d0e-4e96-99d8-08e41f7aeb6b"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">ner_posology_greedy download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n+----------------+---------+---------+\n|           chunk|ner_label|assertion|\n+----------------+---------+---------+\n|capsule of Advil|     DRUG|   absent|\n+----------------+---------+---------+\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">ner_posology_greedy download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n+----------------+---------+---------+\n           chunk|ner_label|assertion|\n+----------------+---------+---------+\ncapsule of Advil|     DRUG|   absent|\n+----------------+---------+---------+\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["import pandas as pd\n\ndef get_clinical_assertion_light (light_model, text):\n\n  light_result = light_model.fullAnnotate(text)[0]\n\n  chunks=[]\n  entities=[]\n  status=[]\n\n  for n,m in zip(light_result['ner_chunk'],light_result['assertion']):\n      \n      chunks.append(n.result)\n      entities.append(n.metadata['entity']) \n      status.append(m.result)\n          \n  df = pd.DataFrame({'chunks':chunks, 'entities':entities, 'assertion':status})\n\n  return df"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"925cdb03-42c3-43d5-ab5f-f6544fd9b5c3"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["clinical_text = \"\"\"\nPatient with severe fever and sore throat. \nHe shows no stomach pain and he maintained on an epidural and PCA for pain control.\nHe also became short of breath with climbing a flight of stairs.\nAfter CT, lung tumor located at the right lower lobe. Father with Alzheimer.\n\"\"\"\n\nlight_model = LightPipeline(model)\n\nget_clinical_assertion_light (light_model, clinical_text)\n"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"2f849a87-7851-4d0b-add0-540c0527d6ba"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">Out[31]: </div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Out[31]: </div>"]}},{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chunks</th>\n      <th>entities</th>\n      <th>assertion</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>severe fever</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>sore throat</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>stomach pain</td>\n      <td>PROBLEM</td>\n      <td>absent</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>an epidural</td>\n      <td>TREATMENT</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>PCA</td>\n      <td>TREATMENT</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>pain control</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>short of breath</td>\n      <td>PROBLEM</td>\n      <td>conditional</td>\n    </tr>\n    <tr>\n      <th>7</th>\n      <td>CT</td>\n      <td>TEST</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>8</th>\n      <td>lung tumor</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>9</th>\n      <td>Alzheimer</td>\n      <td>PROBLEM</td>\n      <td>associated_with_someone_else</td>\n    </tr>\n  </tbody>\n</table>\n</div>","textData":null,"removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"htmlSandbox","arguments":{}}},"output_type":"display_data","data":{"text/html":["<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chunks</th>\n      <th>entities</th>\n      <th>assertion</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>severe fever</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>sore throat</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>stomach pain</td>\n      <td>PROBLEM</td>\n      <td>absent</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>an epidural</td>\n      <td>TREATMENT</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>PCA</td>\n      <td>TREATMENT</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>pain control</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>short of breath</td>\n      <td>PROBLEM</td>\n      <td>conditional</td>\n    </tr>\n    <tr>\n      <th>7</th>\n      <td>CT</td>\n      <td>TEST</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>8</th>\n      <td>lung tumor</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>9</th>\n      <td>Alzheimer</td>\n      <td>PROBLEM</td>\n      <td>associated_with_someone_else</td>\n    </tr>\n  </tbody>\n</table>\n</div>"]}}],"execution_count":0},{"cell_type":"markdown","source":["## Assertion with BioNLP (Cancer Genetics) NER"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"bdd0c509-8006-4c5c-9092-40256f47e34c"}}},{"cell_type":"code","source":["embeddings = 'embeddings_clinical'\n\nmodel_name = 'ner_bionlp'\n\nnrows = 100\n\nner_df = get_clinical_assertion (embeddings, pubMedDF, nrows, model_name)\n\nner_df.show(truncate = False)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"2ed9f618-83f3-4225-8e4c-79984a17c6ac"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">ner_bionlp download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[ / ]\r[ — ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n+----------------+----------------------+---------+\n|chunk           |ner_label             |assertion|\n+----------------+----------------------+---------+\n|optic axis      |Multi-tissue_structure|possible |\n|esophageal ECG  |Multi-tissue_structure|present  |\n|nose            |Organ                 |present  |\n|esophageal ECG  |Multi-tissue_structure|present  |\n|E5              |Gene_or_gene_product  |present  |\n|Tyr             |Amino_acid            |present  |\n|Asn             |Amino_acid            |present  |\n|Asp             |Amino_acid            |present  |\n|Escherichia coli|Organism              |present  |\n|RNases Aand T1  |Gene_or_gene_product  |present  |\n|E5-CRD          |Gene_or_gene_product  |absent   |\n|usualRNases     |Gene_or_gene_product  |absent   |\n|RNase           |Gene_or_gene_product  |present  |\n|Lys25           |Simple_chemical       |present  |\n|Gln29           |Simple_chemical       |present  |\n|Lys60           |Simple_chemical       |present  |\n|Ile94           |Gene_or_gene_product  |present  |\n|Arg33           |Gene_or_gene_product  |possible |\n|Arg33           |Gene_or_gene_product  |possible |\n|Lys60           |Simple_chemical       |present  |\n+----------------+----------------------+---------+\nonly showing top 20 rows\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">ner_bionlp download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[ / ]\r[ — ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n+----------------+----------------------+---------+\nchunk           |ner_label             |assertion|\n+----------------+----------------------+---------+\noptic axis      |Multi-tissue_structure|possible |\nesophageal ECG  |Multi-tissue_structure|present  |\nnose            |Organ                 |present  |\nesophageal ECG  |Multi-tissue_structure|present  |\nE5              |Gene_or_gene_product  |present  |\nTyr             |Amino_acid            |present  |\nAsn             |Amino_acid            |present  |\nAsp             |Amino_acid            |present  |\nEscherichia coli|Organism              |present  |\nRNases Aand T1  |Gene_or_gene_product  |present  |\nE5-CRD          |Gene_or_gene_product  |absent   |\nusualRNases     |Gene_or_gene_product  |absent   |\nRNase           |Gene_or_gene_product  |present  |\nLys25           |Simple_chemical       |present  |\nGln29           |Simple_chemical       |present  |\nLys60           |Simple_chemical       |present  |\nIle94           |Gene_or_gene_product  |present  |\nArg33           |Gene_or_gene_product  |possible |\nArg33           |Gene_or_gene_product  |possible |\nLys60           |Simple_chemical       |present  |\n+----------------+----------------------+---------+\nonly showing top 20 rows\n\n</div>"]}}],"execution_count":0},{"cell_type":"markdown","source":["# Assertion Filterer\nAssertionFilterer will allow you to filter out the named entities by the list of acceptable assertion statuses. This annotator would be quite handy if you want to set a white list for the acceptable assertion statuses like present or conditional; and do not want absent conditions get out of your pipeline."],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"5b6f9442-9a54-49cc-ad1d-186fc5ea326e"}}},{"cell_type":"code","source":["clinical_ner = MedicalNerModel.pretrained(\"ner_clinical\", \"en\", \"clinical/models\") \\\n      .setInputCols([\"sentence\", \"token\", \"embeddings\"]) \\\n      .setOutputCol(\"ner\")\n\nner_converter = NerConverter() \\\n      .setInputCols([\"sentence\", \"token\", \"ner\"]) \\\n      .setOutputCol(\"ner_chunk\")\n\nclinical_assertion = AssertionDLModel.pretrained(\"assertion_dl\", \"en\", \"clinical/models\") \\\n      .setInputCols([\"sentence\", \"ner_chunk\", \"embeddings\"]) \\\n      .setOutputCol(\"assertion\")\n\nassertion_filterer = AssertionFilterer()\\\n      .setInputCols(\"sentence\",\"ner_chunk\",\"assertion\")\\\n      .setOutputCol(\"assertion_filtered\")\\\n      .setWhiteList([\"present\"])\n\nnlpPipeline = Pipeline(stages=[\n      documentAssembler, \n      sentenceDetector,\n      tokenizer,\n      word_embeddings,\n      clinical_ner,\n      ner_converter,\n      clinical_assertion,\n      assertion_filterer\n    ])\n\nempty_data = spark.createDataFrame([[\"\"]]).toDF(\"text\")\nassertionFilter_model = nlpPipeline.fit(empty_data)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"f50d1b8e-edff-4d4c-839f-d0ba3d0b8923"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">ner_clinical download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">ner_clinical download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["text = 'Patient has a headache for the last 2 weeks, needs to get a head CT, and appears anxious when she walks fast. Alopecia noted. She denies pain.'\n\nlight_model = LightPipeline(assertionFilter_model)\nlight_result = light_model.annotate(text)\n\nlight_result.keys()"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"f3800714-30a3-4515-9932-5a8c0c2021a6"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">Out[34]: dict_keys([&#39;assertion_filtered&#39;, &#39;document&#39;, &#39;ner_chunk&#39;, &#39;assertion&#39;, &#39;token&#39;, &#39;ner&#39;, &#39;embeddings&#39;, &#39;sentence&#39;])</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Out[34]: dict_keys([&#39;assertion_filtered&#39;, &#39;document&#39;, &#39;ner_chunk&#39;, &#39;assertion&#39;, &#39;token&#39;, &#39;ner&#39;, &#39;embeddings&#39;, &#39;sentence&#39;])</div>"]}}],"execution_count":0},{"cell_type":"code","source":["list(zip(light_result['ner_chunk'], light_result['assertion']))"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"b8986c91-53c4-4a08-a36b-1fea93c1c762"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">Out[35]: [(&#39;a headache&#39;, &#39;present&#39;),\n (&#39;a head CT&#39;, &#39;present&#39;),\n (&#39;anxious&#39;, &#39;present&#39;),\n (&#39;fast&#39;, &#39;present&#39;),\n (&#39;Alopecia&#39;, &#39;present&#39;),\n (&#39;pain&#39;, &#39;absent&#39;)]</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Out[35]: [(&#39;a headache&#39;, &#39;present&#39;),\n (&#39;a head CT&#39;, &#39;present&#39;),\n (&#39;anxious&#39;, &#39;present&#39;),\n (&#39;fast&#39;, &#39;present&#39;),\n (&#39;Alopecia&#39;, &#39;present&#39;),\n (&#39;pain&#39;, &#39;absent&#39;)]</div>"]}}],"execution_count":0},{"cell_type":"code","source":["assertion_filterer.getWhiteList()"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"44113efd-939b-4fd1-95f2-110e957a6635"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">Out[36]: [&#39;present&#39;]</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Out[36]: [&#39;present&#39;]</div>"]}}],"execution_count":0},{"cell_type":"code","source":["light_result['assertion_filtered']"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"aa81ca8c-7573-42cb-a895-3981c4ff1bdc"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">Out[37]: [&#39;a headache&#39;, &#39;a head CT&#39;, &#39;anxious&#39;, &#39;fast&#39;, &#39;Alopecia&#39;]</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Out[37]: [&#39;a headache&#39;, &#39;a head CT&#39;, &#39;anxious&#39;, &#39;fast&#39;, &#39;Alopecia&#39;]</div>"]}}],"execution_count":0},{"cell_type":"markdown","source":["# Train a custom Assertion Model"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"647435eb-a456-488f-ade2-f97e4c8de429"}}},{"cell_type":"markdown","source":["**WARNING:** For training an Assertion model, please use TensorFlow version 2.3"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"9aa4951b-abd4-44a7-bc32-51719afb01a6"}}},{"cell_type":"code","source":["!wget -q https://s3.amazonaws.com/auxdata.johnsnowlabs.com/public/resources/en/ner/i2b2_assertion_sample.csv -P /dbfs/"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"2d72802c-1c83-4572-a52e-c7e1074e69cb"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["assertion_df = spark.read.option(\"header\", True).option(\"inferSchema\", \"True\").csv(\"/i2b2_assertion_sample.csv\")\n"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"711aa642-d5ab-4f57-8063-4451b6ee0d11"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["assertion_df.show(3, truncate=100)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"d9f89d7b-ce06-43dc-bdff-e53cbb6f8ada"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">+-----------------------------------------------------------------------------+--------------------+-------+-----+---+\n|                                                                         text|              target|  label|start|end|\n+-----------------------------------------------------------------------------+--------------------+-------+-----+---+\n|Due to dye load required to assess the coronary disease were not visualized .|the coronary disease|present|    7|  9|\n|                            She has no history of liver disease , hepatitis .|       liver disease| absent|    5|  6|\n|                                                     1. Undesired fertility .| undesired fertility|present|    1|  2|\n+-----------------------------------------------------------------------------+--------------------+-------+-----+---+\nonly showing top 3 rows\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">+-----------------------------------------------------------------------------+--------------------+-------+-----+---+\n                                                                         text|              target|  label|start|end|\n+-----------------------------------------------------------------------------+--------------------+-------+-----+---+\nDue to dye load required to assess the coronary disease were not visualized .|the coronary disease|present|    7|  9|\n                            She has no history of liver disease , hepatitis .|       liver disease| absent|    5|  6|\n                                                     1. Undesired fertility .| undesired fertility|present|    1|  2|\n+-----------------------------------------------------------------------------+--------------------+-------+-----+---+\nonly showing top 3 rows\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["(training_data, test_data) = assertion_df.randomSplit([0.8, 0.2], seed = 100)\nprint(\"Training Dataset Count: \" + str(training_data.count()))\nprint(\"Test Dataset Count: \" + str(test_data.count()))"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"eba6ba43-dbd4-4403-b967-8ecde9e0461a"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">Training Dataset Count: 814\nTest Dataset Count: 186\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Training Dataset Count: 814\nTest Dataset Count: 186\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["training_data.groupBy('label').count().orderBy('count', ascending=False).show(truncate=False)\n"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"208b749b-13fa-4780-93a1-d7cd316e81f6"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">+----------------------------+-----+\n|label                       |count|\n+----------------------------+-----+\n|present                     |546  |\n|absent                      |182  |\n|hypothetical                |41   |\n|possible                    |34   |\n|associated_with_someone_else|6    |\n|conditional                 |5    |\n+----------------------------+-----+\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">+----------------------------+-----+\nlabel                       |count|\n+----------------------------+-----+\npresent                     |546  |\nabsent                      |182  |\nhypothetical                |41   |\npossible                    |34   |\nassociated_with_someone_else|6    |\nconditional                 |5    |\n+----------------------------+-----+\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["document = DocumentAssembler()\\\n    .setInputCol(\"text\")\\\n    .setOutputCol(\"document\")\n\nchunk = Doc2Chunk()\\\n    .setInputCols(\"document\")\\\n    .setOutputCol(\"chunk\")\\\n    .setChunkCol(\"target\")\\\n    .setStartCol(\"start\")\\\n    .setStartColByTokenIndex(True)\\\n    .setFailOnMissing(False)\\\n    .setLowerCase(True)\n\ntoken = Tokenizer()\\\n    .setInputCols(['document'])\\\n    .setOutputCol('token')\n\nembeddings = WordEmbeddingsModel.pretrained(\"embeddings_clinical\", \"en\", \"clinical/models\")\\\n      .setInputCols([\"document\", \"token\"])\\\n      .setOutputCol(\"embeddings\")\n"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"1dd09bfb-5a71-49b0-8e5a-ce40820d154a"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">embeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">embeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n</div>"]}}],"execution_count":0},{"cell_type":"markdown","source":["#### With a custom tf graph"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"90a7fafa-35e2-4437-a473-7f7c482ba351"}}},{"cell_type":"code","source":["import tensorflow\nfrom sparknlp_jsl.training import tf_graph"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"11bead5c-9e4b-4e5b-b1cf-164e59decbfb"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["%fs mkdirs file:/dbfs/assertion_tf_graphs"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"26545bcf-6cf6-4d85-a521-b70df503c635"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">res2: Boolean = true\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{"isDbfsCommandResult":false},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">res2: Boolean = true\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["%fs mkdirs file:/dbfs/assertion_logs"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"febac50c-d17f-466f-85ee-87b24b54449d"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">res3: Boolean = true\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{"isDbfsCommandResult":false},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">res3: Boolean = true\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["%fs ls file:/dbfs"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"1b48fc09-9f8e-4f79-80bd-eb6b7b13dc9a"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"overflow":false,"datasetInfos":[],"data":[["file:/dbfs/AntBNC_lemmas_ver_001.txt","AntBNC_lemmas_ver_001.txt",1348552],["file:/dbfs/AskAPatient.fold-0.test.txt","AskAPatient.fold-0.test.txt",33834],["file:/dbfs/AskAPatient.fold-0.train.txt","AskAPatient.fold-0.train.txt",594794],["file:/dbfs/AskAPatient.fold-0.validation.txt","AskAPatient.fold-0.validation.txt",32162],["file:/dbfs/FileStore/","FileStore/",4096],["file:/dbfs/JohnSnowLabs/","JohnSnowLabs/",4096],["file:/dbfs/MT_00.pdf","MT_00.pdf",528022],["file:/dbfs/MT_01.pdf","MT_01.pdf",165397],["file:/dbfs/MT_02.pdf","MT_02.pdf",309334],["file:/dbfs/MT_03.pdf","MT_03.pdf",248334],["file:/dbfs/NCBI_disease_official_test.conll","NCBI_disease_official_test.conll",328172],["file:/dbfs/NCBI_disease_official_test.conll.1","NCBI_disease_official_test.conll.1",328172],["file:/dbfs/NCBI_disease_official_test.conll.2","NCBI_disease_official_test.conll.2",328172],["file:/dbfs/NCBI_disease_official_train_dev.conll","NCBI_disease_official_train_dev.conll",2136193],["file:/dbfs/NCBI_disease_official_train_dev.conll.1","NCBI_disease_official_train_dev.conll.1",2136193],["file:/dbfs/NCBI_disease_official_train_dev.conll.2","NCBI_disease_official_train_dev.conll.2",2136193],["file:/dbfs/NER_NCBIconlltest.parquet/","NER_NCBIconlltest.parquet/",4096],["file:/dbfs/NER_NCBIconlltest.txt","NER_NCBIconlltest.txt",244358],["file:/dbfs/NER_NCBIconlltrain.txt","NER_NCBIconlltrain.txt",1091488],["file:/dbfs/Users/","Users/",4096],["file:/dbfs/Workspace/","Workspace/",4096],["file:/dbfs/aclimdb_test.csv","aclimdb_test.csv",32715164],["file:/dbfs/aclimdb_train.csv","aclimdb_train.csv",33497180],["file:/dbfs/ade_data/","ade_data/",4096],["file:/dbfs/age.json","age.json",780],["file:/dbfs/amazon_alexa.tsv","amazon_alexa.tsv",514752],["file:/dbfs/annotators.csv","annotators.csv",3136],["file:/dbfs/assertion_logs/","assertion_logs/",4096],["file:/dbfs/assertion_tf_graphs/","assertion_tf_graphs/",4096],["file:/dbfs/benchmark/","benchmark/",4096],["file:/dbfs/cities.json","cities.json",101],["file:/dbfs/cities.tsv","cities.tsv",48],["file:/dbfs/classifierdl_tensorflow","classifierdl_tensorflow",22419981],["file:/dbfs/classifierdl_transform/","classifierdl_transform/",4096],["file:/dbfs/classifierdl_transform_new/","classifierdl_transform_new/",4096],["file:/dbfs/clf_bert_logs/","clf_bert_logs/",4096],["file:/dbfs/clf_dl_logs/","clf_dl_logs/",4096],["file:/dbfs/clf_glove_logs/","clf_glove_logs/",4096],["file:/dbfs/clf_use_logs/","clf_use_logs/",4096],["file:/dbfs/custom_deid_masker_model/","custom_deid_masker_model/",4096],["file:/dbfs/custom_obfuscator_model/","custom_obfuscator_model/",4096],["file:/dbfs/databricks/","databricks/",4096],["file:/dbfs/databricks-datasets/","databricks-datasets/",4096],["file:/dbfs/databricks-results/","databricks-results/",4096],["file:/dbfs/datasets","datasets",24032125],["file:/dbfs/date.json","date.json",205],["file:/dbfs/dbfs/","dbfs/",4096],["file:/dbfs/deneme/","deneme/",4096],["file:/dbfs/dicom_1.dcm","dicom_1.dcm",3836670],["file:/dbfs/dicom_2.dcm","dicom_2.dcm",640512],["file:/dbfs/dicom_3.dcm","dicom_3.dcm",900810],["file:/dbfs/fields/","fields/",4096],["file:/dbfs/financial_entities.txt","financial_entities.txt",27],["file:/dbfs/gc_graph/","gc_graph/",4096],["file:/dbfs/gender.csv","gender.csv",105],["file:/dbfs/gender.json","gender.json",98],["file:/dbfs/generic_logs/","generic_logs/",4096],["file:/dbfs/hipaa-table-001.txt","hipaa-table-001.txt",2497],["file:/dbfs/hls","hls",103558],["file:/dbfs/holmes.txt","holmes.txt",559722],["file:/dbfs/home/","home/",4096],["file:/dbfs/i2b2_assertion_sample.csv","i2b2_assertion_sample.csv",141585],["file:/dbfs/i2b2_assertion_sample.csv.1","i2b2_assertion_sample.csv.1",141585],["file:/dbfs/i2b2_assertion_sample.csv.2","i2b2_assertion_sample.csv.2",141585],["file:/dbfs/i2b2_assertion_sample.csv.3","i2b2_assertion_sample.csv.3",141585],["file:/dbfs/i2b2_assertion_sample.csv.4","i2b2_assertion_sample.csv.4",141585],["file:/dbfs/i2b2_assertion_sample.csv.5","i2b2_assertion_sample.csv.5",141585],["file:/dbfs/i2b2_assertion_sample.csv.6","i2b2_assertion_sample.csv.6",141585],["file:/dbfs/i2b2_assertion_sample.csv.7","i2b2_assertion_sample.csv.7",141585],["file:/dbfs/i2b2_assertion_sample.csv.8","i2b2_assertion_sample.csv.8",141585],["file:/dbfs/i2b2_assertion_sample.csv.9","i2b2_assertion_sample.csv.9",141585],["file:/dbfs/i2b2_clinical_rel_dataset.csv","i2b2_clinical_rel_dataset.csv",285456],["file:/dbfs/icd10_askapatient_25_1000/","icd10_askapatient_25_1000/",4096],["file:/dbfs/inference.parquet/","inference.parquet/",4096],["file:/dbfs/local_disk0/","local_disk0/",4096],["file:/dbfs/medical_ner_graph/","medical_ner_graph/",4096],["file:/dbfs/medical_ner_graphs/","medical_ner_graphs/",4096],["file:/dbfs/medical_ner_graphs_tf2/","medical_ner_graphs_tf2/",4096],["file:/dbfs/metadata/","metadata/",4096],["file:/dbfs/mnt/","mnt/",4096],["file:/dbfs/model/","model/",4096],["file:/dbfs/models/","models/",4096],["file:/dbfs/natural_scene.jpeg","natural_scene.jpeg",133271],["file:/dbfs/ner/","ner/",4096],["file:/dbfs/ner_logs/","ner_logs/",4096],["file:/dbfs/ner_logs_best/","ner_logs_best/",4096],["file:/dbfs/ner_logs_tf2/","ner_logs_tf2/",4096],["file:/dbfs/nerdl_test.parquet/","nerdl_test.parquet/",4096],["file:/dbfs/news_category_test.csv","news_category_test.csv",1504408],["file:/dbfs/news_category_train.csv","news_category_train.csv",24032125],["file:/dbfs/news_use/","news_use/",4096],["file:/dbfs/noised.pdf","noised.pdf",2115939],["file:/dbfs/obfuscator_unique_ref_test.txt","obfuscator_unique_ref_test.txt",337],["file:/dbfs/ocr/","ocr/",4096],["file:/dbfs/oncology_notes","oncology_notes",111682],["file:/dbfs/oncology_study","oncology_study",111468],["file:/dbfs/pet.in1202D.out2.pb","pet.in1202D.out2.pb",122582],["file:/dbfs/petfinder-mini.csv","petfinder-mini.csv",4763125],["file:/dbfs/pubmed-sample.csv","pubmed-sample.csv",10484510],["file:/dbfs/pubmed_sample_text_small.csv","pubmed_sample_text_small.csv",9363435],["file:/dbfs/pubmed_sample_text_small.csv.1","pubmed_sample_text_small.csv.1",9363435],["file:/dbfs/pubmed_sample_text_small.csv.2","pubmed_sample_text_small.csv.2",9363435],["file:/dbfs/pubmed_sample_text_small.csv.3","pubmed_sample_text_small.csv.3",9363435],["file:/dbfs/pypi/","pypi/",4096],["file:/dbfs/re_graphs/","re_graphs/",4096],["file:/dbfs/re_logs/","re_logs/",4096],["file:/dbfs/regex_rules.txt","regex_rules.txt",195],["file:/dbfs/results/","results/",4096],["file:/dbfs/root/","root/",4096],["file:/dbfs/sample-sentences-en.txt","sample-sentences-en.txt",284],["file:/dbfs/sample_ADE_dataset.csv","sample_ADE_dataset.csv",9434],["file:/dbfs/sample_doc.pdf","sample_doc.pdf",212973],["file:/dbfs/snomed_temp/","snomed_temp/",4096],["file:/dbfs/spam_ham_dataset.csv","spam_ham_dataset.csv",5502589],["file:/dbfs/sport_entities.txt","sport_entities.txt",36],["file:/dbfs/temp/","temp/",4096],["file:/dbfs/test_1.parquet/","test_1.parquet/",4096],["file:/dbfs/test_2.parquet/","test_2.parquet/",4096],["file:/dbfs/test_document.pdf","test_document.pdf",693743],["file:/dbfs/text_with_noise.png","text_with_noise.png",100544],["file:/dbfs/title_conference.csv","title_conference.csv",188402],["file:/dbfs/tmp/","tmp/",4096],["file:/dbfs/tmp2/","tmp2/",4096],["file:/dbfs/toxic_test.snappy.parquet","toxic_test.snappy.parquet",296833],["file:/dbfs/toxic_train.snappy.parquet","toxic_train.snappy.parquet",2767307],["file:/dbfs/user/","user/",4096],["file:/dbfs/~/","~/",4096]],"plotOptions":{"displayType":"table","customPlotOptions":{},"pivotColumns":null,"pivotAggregation":null,"xColumns":null,"yColumns":null},"columnCustomDisplayInfos":{},"aggType":"","isJsonSchema":true,"removedWidgets":[],"aggSchema":[],"schema":[{"name":"path","type":"\"string\"","metadata":"{}"},{"name":"name","type":"\"string\"","metadata":"{}"},{"name":"size","type":"\"long\"","metadata":"{}"}],"aggError":"","aggData":[],"addedWidgets":{},"metadata":{"isDbfsCommandResult":false},"dbfsResultPath":null,"type":"table","aggOverflow":false,"aggSeriesLimitReached":false,"arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .table-result-container {\n    max-height: 300px;\n    overflow: auto;\n  }\n  table, th, td {\n    border: 1px solid black;\n    border-collapse: collapse;\n  }\n  th, td {\n    padding: 5px;\n  }\n  th {\n    text-align: left;\n  }\n</style><div class='table-result-container'><table class='table-result'><thead style='background-color: white'><tr><th>path</th><th>name</th><th>size</th></tr></thead><tbody><tr><td>file:/dbfs/AntBNC_lemmas_ver_001.txt</td><td>AntBNC_lemmas_ver_001.txt</td><td>1348552</td></tr><tr><td>file:/dbfs/AskAPatient.fold-0.test.txt</td><td>AskAPatient.fold-0.test.txt</td><td>33834</td></tr><tr><td>file:/dbfs/AskAPatient.fold-0.train.txt</td><td>AskAPatient.fold-0.train.txt</td><td>594794</td></tr><tr><td>file:/dbfs/AskAPatient.fold-0.validation.txt</td><td>AskAPatient.fold-0.validation.txt</td><td>32162</td></tr><tr><td>file:/dbfs/FileStore/</td><td>FileStore/</td><td>4096</td></tr><tr><td>file:/dbfs/JohnSnowLabs/</td><td>JohnSnowLabs/</td><td>4096</td></tr><tr><td>file:/dbfs/MT_00.pdf</td><td>MT_00.pdf</td><td>528022</td></tr><tr><td>file:/dbfs/MT_01.pdf</td><td>MT_01.pdf</td><td>165397</td></tr><tr><td>file:/dbfs/MT_02.pdf</td><td>MT_02.pdf</td><td>309334</td></tr><tr><td>file:/dbfs/MT_03.pdf</td><td>MT_03.pdf</td><td>248334</td></tr><tr><td>file:/dbfs/NCBI_disease_official_test.conll</td><td>NCBI_disease_official_test.conll</td><td>328172</td></tr><tr><td>file:/dbfs/NCBI_disease_official_test.conll.1</td><td>NCBI_disease_official_test.conll.1</td><td>328172</td></tr><tr><td>file:/dbfs/NCBI_disease_official_test.conll.2</td><td>NCBI_disease_official_test.conll.2</td><td>328172</td></tr><tr><td>file:/dbfs/NCBI_disease_official_train_dev.conll</td><td>NCBI_disease_official_train_dev.conll</td><td>2136193</td></tr><tr><td>file:/dbfs/NCBI_disease_official_train_dev.conll.1</td><td>NCBI_disease_official_train_dev.conll.1</td><td>2136193</td></tr><tr><td>file:/dbfs/NCBI_disease_official_train_dev.conll.2</td><td>NCBI_disease_official_train_dev.conll.2</td><td>2136193</td></tr><tr><td>file:/dbfs/NER_NCBIconlltest.parquet/</td><td>NER_NCBIconlltest.parquet/</td><td>4096</td></tr><tr><td>file:/dbfs/NER_NCBIconlltest.txt</td><td>NER_NCBIconlltest.txt</td><td>244358</td></tr><tr><td>file:/dbfs/NER_NCBIconlltrain.txt</td><td>NER_NCBIconlltrain.txt</td><td>1091488</td></tr><tr><td>file:/dbfs/Users/</td><td>Users/</td><td>4096</td></tr><tr><td>file:/dbfs/Workspace/</td><td>Workspace/</td><td>4096</td></tr><tr><td>file:/dbfs/aclimdb_test.csv</td><td>aclimdb_test.csv</td><td>32715164</td></tr><tr><td>file:/dbfs/aclimdb_train.csv</td><td>aclimdb_train.csv</td><td>33497180</td></tr><tr><td>file:/dbfs/ade_data/</td><td>ade_data/</td><td>4096</td></tr><tr><td>file:/dbfs/age.json</td><td>age.json</td><td>780</td></tr><tr><td>file:/dbfs/amazon_alexa.tsv</td><td>amazon_alexa.tsv</td><td>514752</td></tr><tr><td>file:/dbfs/annotators.csv</td><td>annotators.csv</td><td>3136</td></tr><tr><td>file:/dbfs/assertion_logs/</td><td>assertion_logs/</td><td>4096</td></tr><tr><td>file:/dbfs/assertion_tf_graphs/</td><td>assertion_tf_graphs/</td><td>4096</td></tr><tr><td>file:/dbfs/benchmark/</td><td>benchmark/</td><td>4096</td></tr><tr><td>file:/dbfs/cities.json</td><td>cities.json</td><td>101</td></tr><tr><td>file:/dbfs/cities.tsv</td><td>cities.tsv</td><td>48</td></tr><tr><td>file:/dbfs/classifierdl_tensorflow</td><td>classifierdl_tensorflow</td><td>22419981</td></tr><tr><td>file:/dbfs/classifierdl_transform/</td><td>classifierdl_transform/</td><td>4096</td></tr><tr><td>file:/dbfs/classifierdl_transform_new/</td><td>classifierdl_transform_new/</td><td>4096</td></tr><tr><td>file:/dbfs/clf_bert_logs/</td><td>clf_bert_logs/</td><td>4096</td></tr><tr><td>file:/dbfs/clf_dl_logs/</td><td>clf_dl_logs/</td><td>4096</td></tr><tr><td>file:/dbfs/clf_glove_logs/</td><td>clf_glove_logs/</td><td>4096</td></tr><tr><td>file:/dbfs/clf_use_logs/</td><td>clf_use_logs/</td><td>4096</td></tr><tr><td>file:/dbfs/custom_deid_masker_model/</td><td>custom_deid_masker_model/</td><td>4096</td></tr><tr><td>file:/dbfs/custom_obfuscator_model/</td><td>custom_obfuscator_model/</td><td>4096</td></tr><tr><td>file:/dbfs/databricks/</td><td>databricks/</td><td>4096</td></tr><tr><td>file:/dbfs/databricks-datasets/</td><td>databricks-datasets/</td><td>4096</td></tr><tr><td>file:/dbfs/databricks-results/</td><td>databricks-results/</td><td>4096</td></tr><tr><td>file:/dbfs/datasets</td><td>datasets</td><td>24032125</td></tr><tr><td>file:/dbfs/date.json</td><td>date.json</td><td>205</td></tr><tr><td>file:/dbfs/dbfs/</td><td>dbfs/</td><td>4096</td></tr><tr><td>file:/dbfs/deneme/</td><td>deneme/</td><td>4096</td></tr><tr><td>file:/dbfs/dicom_1.dcm</td><td>dicom_1.dcm</td><td>3836670</td></tr><tr><td>file:/dbfs/dicom_2.dcm</td><td>dicom_2.dcm</td><td>640512</td></tr><tr><td>file:/dbfs/dicom_3.dcm</td><td>dicom_3.dcm</td><td>900810</td></tr><tr><td>file:/dbfs/fields/</td><td>fields/</td><td>4096</td></tr><tr><td>file:/dbfs/financial_entities.txt</td><td>financial_entities.txt</td><td>27</td></tr><tr><td>file:/dbfs/gc_graph/</td><td>gc_graph/</td><td>4096</td></tr><tr><td>file:/dbfs/gender.csv</td><td>gender.csv</td><td>105</td></tr><tr><td>file:/dbfs/gender.json</td><td>gender.json</td><td>98</td></tr><tr><td>file:/dbfs/generic_logs/</td><td>generic_logs/</td><td>4096</td></tr><tr><td>file:/dbfs/hipaa-table-001.txt</td><td>hipaa-table-001.txt</td><td>2497</td></tr><tr><td>file:/dbfs/hls</td><td>hls</td><td>103558</td></tr><tr><td>file:/dbfs/holmes.txt</td><td>holmes.txt</td><td>559722</td></tr><tr><td>file:/dbfs/home/</td><td>home/</td><td>4096</td></tr><tr><td>file:/dbfs/i2b2_assertion_sample.csv</td><td>i2b2_assertion_sample.csv</td><td>141585</td></tr><tr><td>file:/dbfs/i2b2_assertion_sample.csv.1</td><td>i2b2_assertion_sample.csv.1</td><td>141585</td></tr><tr><td>file:/dbfs/i2b2_assertion_sample.csv.2</td><td>i2b2_assertion_sample.csv.2</td><td>141585</td></tr><tr><td>file:/dbfs/i2b2_assertion_sample.csv.3</td><td>i2b2_assertion_sample.csv.3</td><td>141585</td></tr><tr><td>file:/dbfs/i2b2_assertion_sample.csv.4</td><td>i2b2_assertion_sample.csv.4</td><td>141585</td></tr><tr><td>file:/dbfs/i2b2_assertion_sample.csv.5</td><td>i2b2_assertion_sample.csv.5</td><td>141585</td></tr><tr><td>file:/dbfs/i2b2_assertion_sample.csv.6</td><td>i2b2_assertion_sample.csv.6</td><td>141585</td></tr><tr><td>file:/dbfs/i2b2_assertion_sample.csv.7</td><td>i2b2_assertion_sample.csv.7</td><td>141585</td></tr><tr><td>file:/dbfs/i2b2_assertion_sample.csv.8</td><td>i2b2_assertion_sample.csv.8</td><td>141585</td></tr><tr><td>file:/dbfs/i2b2_assertion_sample.csv.9</td><td>i2b2_assertion_sample.csv.9</td><td>141585</td></tr><tr><td>file:/dbfs/i2b2_clinical_rel_dataset.csv</td><td>i2b2_clinical_rel_dataset.csv</td><td>285456</td></tr><tr><td>file:/dbfs/icd10_askapatient_25_1000/</td><td>icd10_askapatient_25_1000/</td><td>4096</td></tr><tr><td>file:/dbfs/inference.parquet/</td><td>inference.parquet/</td><td>4096</td></tr><tr><td>file:/dbfs/local_disk0/</td><td>local_disk0/</td><td>4096</td></tr><tr><td>file:/dbfs/medical_ner_graph/</td><td>medical_ner_graph/</td><td>4096</td></tr><tr><td>file:/dbfs/medical_ner_graphs/</td><td>medical_ner_graphs/</td><td>4096</td></tr><tr><td>file:/dbfs/medical_ner_graphs_tf2/</td><td>medical_ner_graphs_tf2/</td><td>4096</td></tr><tr><td>file:/dbfs/metadata/</td><td>metadata/</td><td>4096</td></tr><tr><td>file:/dbfs/mnt/</td><td>mnt/</td><td>4096</td></tr><tr><td>file:/dbfs/model/</td><td>model/</td><td>4096</td></tr><tr><td>file:/dbfs/models/</td><td>models/</td><td>4096</td></tr><tr><td>file:/dbfs/natural_scene.jpeg</td><td>natural_scene.jpeg</td><td>133271</td></tr><tr><td>file:/dbfs/ner/</td><td>ner/</td><td>4096</td></tr><tr><td>file:/dbfs/ner_logs/</td><td>ner_logs/</td><td>4096</td></tr><tr><td>file:/dbfs/ner_logs_best/</td><td>ner_logs_best/</td><td>4096</td></tr><tr><td>file:/dbfs/ner_logs_tf2/</td><td>ner_logs_tf2/</td><td>4096</td></tr><tr><td>file:/dbfs/nerdl_test.parquet/</td><td>nerdl_test.parquet/</td><td>4096</td></tr><tr><td>file:/dbfs/news_category_test.csv</td><td>news_category_test.csv</td><td>1504408</td></tr><tr><td>file:/dbfs/news_category_train.csv</td><td>news_category_train.csv</td><td>24032125</td></tr><tr><td>file:/dbfs/news_use/</td><td>news_use/</td><td>4096</td></tr><tr><td>file:/dbfs/noised.pdf</td><td>noised.pdf</td><td>2115939</td></tr><tr><td>file:/dbfs/obfuscator_unique_ref_test.txt</td><td>obfuscator_unique_ref_test.txt</td><td>337</td></tr><tr><td>file:/dbfs/ocr/</td><td>ocr/</td><td>4096</td></tr><tr><td>file:/dbfs/oncology_notes</td><td>oncology_notes</td><td>111682</td></tr><tr><td>file:/dbfs/oncology_study</td><td>oncology_study</td><td>111468</td></tr><tr><td>file:/dbfs/pet.in1202D.out2.pb</td><td>pet.in1202D.out2.pb</td><td>122582</td></tr><tr><td>file:/dbfs/petfinder-mini.csv</td><td>petfinder-mini.csv</td><td>4763125</td></tr><tr><td>file:/dbfs/pubmed-sample.csv</td><td>pubmed-sample.csv</td><td>10484510</td></tr><tr><td>file:/dbfs/pubmed_sample_text_small.csv</td><td>pubmed_sample_text_small.csv</td><td>9363435</td></tr><tr><td>file:/dbfs/pubmed_sample_text_small.csv.1</td><td>pubmed_sample_text_small.csv.1</td><td>9363435</td></tr><tr><td>file:/dbfs/pubmed_sample_text_small.csv.2</td><td>pubmed_sample_text_small.csv.2</td><td>9363435</td></tr><tr><td>file:/dbfs/pubmed_sample_text_small.csv.3</td><td>pubmed_sample_text_small.csv.3</td><td>9363435</td></tr><tr><td>file:/dbfs/pypi/</td><td>pypi/</td><td>4096</td></tr><tr><td>file:/dbfs/re_graphs/</td><td>re_graphs/</td><td>4096</td></tr><tr><td>file:/dbfs/re_logs/</td><td>re_logs/</td><td>4096</td></tr><tr><td>file:/dbfs/regex_rules.txt</td><td>regex_rules.txt</td><td>195</td></tr><tr><td>file:/dbfs/results/</td><td>results/</td><td>4096</td></tr><tr><td>file:/dbfs/root/</td><td>root/</td><td>4096</td></tr><tr><td>file:/dbfs/sample-sentences-en.txt</td><td>sample-sentences-en.txt</td><td>284</td></tr><tr><td>file:/dbfs/sample_ADE_dataset.csv</td><td>sample_ADE_dataset.csv</td><td>9434</td></tr><tr><td>file:/dbfs/sample_doc.pdf</td><td>sample_doc.pdf</td><td>212973</td></tr><tr><td>file:/dbfs/snomed_temp/</td><td>snomed_temp/</td><td>4096</td></tr><tr><td>file:/dbfs/spam_ham_dataset.csv</td><td>spam_ham_dataset.csv</td><td>5502589</td></tr><tr><td>file:/dbfs/sport_entities.txt</td><td>sport_entities.txt</td><td>36</td></tr><tr><td>file:/dbfs/temp/</td><td>temp/</td><td>4096</td></tr><tr><td>file:/dbfs/test_1.parquet/</td><td>test_1.parquet/</td><td>4096</td></tr><tr><td>file:/dbfs/test_2.parquet/</td><td>test_2.parquet/</td><td>4096</td></tr><tr><td>file:/dbfs/test_document.pdf</td><td>test_document.pdf</td><td>693743</td></tr><tr><td>file:/dbfs/text_with_noise.png</td><td>text_with_noise.png</td><td>100544</td></tr><tr><td>file:/dbfs/title_conference.csv</td><td>title_conference.csv</td><td>188402</td></tr><tr><td>file:/dbfs/tmp/</td><td>tmp/</td><td>4096</td></tr><tr><td>file:/dbfs/tmp2/</td><td>tmp2/</td><td>4096</td></tr><tr><td>file:/dbfs/toxic_test.snappy.parquet</td><td>toxic_test.snappy.parquet</td><td>296833</td></tr><tr><td>file:/dbfs/toxic_train.snappy.parquet</td><td>toxic_train.snappy.parquet</td><td>2767307</td></tr><tr><td>file:/dbfs/user/</td><td>user/</td><td>4096</td></tr><tr><td>file:/dbfs/~/</td><td>~/</td><td>4096</td></tr></tbody></table></div>"]}}],"execution_count":0},{"cell_type":"code","source":["tf_graph.build(\"assertion_dl\",build_params={\"n_classes\": 6}, model_location=\"/dbfs/assertion_tf_graphs\", model_filename=\"auto\")"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"41b053f9-0b3a-4e7d-85c3-461d15dc18d7"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">/databricks/python/lib/python3.8/site-packages/tensorflow_addons/utils/ensure_tf_install.py:53: UserWarning: Tensorflow Addons supports using Python ops for all Tensorflow versions above or equal to 2.6.0 and strictly below 2.9.0 (nightly versions are not supported). \n The versions of TensorFlow you are currently using is 2.3.0 and is not supported. \nSome things might work, some things might not.\nIf you were to encounter a bug, do not file an issue.\nIf you want to make sure you&#39;re using a tested and supported configuration, either change the TensorFlow version or the TensorFlow Addons&#39;s version. \nYou can find the compatibility matrix in TensorFlow Addon&#39;s readme:\nhttps://github.com/tensorflow/addons\n  warnings.warn(\nWARNING:tensorflow:From /databricks/python/lib/python3.8/site-packages/tensorflow/python/compat/v2_compat.py:96: disable_resource_variables (from tensorflow.python.ops.variable_scope) is deprecated and will be removed in a future version.\nInstructions for updating:\nnon-resource variables are not supported in the long term\nDevice mapping:\n/job:localhost/replica:0/task:0/device:XLA_CPU:0 -&gt; device: XLA_CPU device\n\nWARNING:tensorflow:From /databricks/python/lib/python3.8/site-packages/sparknlp_jsl/_tf_graph_builders/assertion_dl/assertion_model.py:48: BasicLSTMCell.__init__ (from tensorflow.python.keras.layers.legacy_rnn.rnn_cell_impl) is deprecated and will be removed in a future version.\nInstructions for updating:\nThis class is equivalent as tf.keras.layers.LSTMCell, and will be replaced by that in Tensorflow 2.0.\nWARNING:tensorflow:From /databricks/python/lib/python3.8/site-packages/sparknlp_jsl/_tf_graph_builders/tf2contrib/rnn.py:229: bidirectional_dynamic_rnn (from tensorflow.python.ops.rnn) is deprecated and will be removed in a future version.\nInstructions for updating:\nPlease use `keras.layers.Bidirectional(keras.layers.RNN(cell))`, which is equivalent to this API\nWARNING:tensorflow:From /databricks/python/lib/python3.8/site-packages/tensorflow/python/ops/rnn.py:438: dynamic_rnn (from tensorflow.python.ops.rnn) is deprecated and will be removed in a future version.\nInstructions for updating:\nPlease use `keras.layers.RNN(cell)`, which is equivalent to this API\nWARNING:tensorflow:From /databricks/python/lib/python3.8/site-packages/tensorflow/python/keras/layers/legacy_rnn/rnn_cell_impl.py:738: Layer.add_variable (from tensorflow.python.keras.engine.base_layer_v1) is deprecated and will be removed in a future version.\nInstructions for updating:\nPlease use `layer.add_weight` method instead.\nWARNING:tensorflow:From /databricks/python/lib/python3.8/site-packages/tensorflow/python/keras/layers/legacy_rnn/rnn_cell_impl.py:744: calling Zeros.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.\nInstructions for updating:\nCall initializer instance with the dtype argument instead of passing it to the constructor\nassertion_dl graph exported to /dbfs/assertion_tf_graphs/blstm_34_32_30_200_6.pb\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">/databricks/python/lib/python3.8/site-packages/tensorflow_addons/utils/ensure_tf_install.py:53: UserWarning: Tensorflow Addons supports using Python ops for all Tensorflow versions above or equal to 2.6.0 and strictly below 2.9.0 (nightly versions are not supported). \n The versions of TensorFlow you are currently using is 2.3.0 and is not supported. \nSome things might work, some things might not.\nIf you were to encounter a bug, do not file an issue.\nIf you want to make sure you&#39;re using a tested and supported configuration, either change the TensorFlow version or the TensorFlow Addons&#39;s version. \nYou can find the compatibility matrix in TensorFlow Addon&#39;s readme:\nhttps://github.com/tensorflow/addons\n  warnings.warn(\nWARNING:tensorflow:From /databricks/python/lib/python3.8/site-packages/tensorflow/python/compat/v2_compat.py:96: disable_resource_variables (from tensorflow.python.ops.variable_scope) is deprecated and will be removed in a future version.\nInstructions for updating:\nnon-resource variables are not supported in the long term\nDevice mapping:\n/job:localhost/replica:0/task:0/device:XLA_CPU:0 -&gt; device: XLA_CPU device\n\nWARNING:tensorflow:From /databricks/python/lib/python3.8/site-packages/sparknlp_jsl/_tf_graph_builders/assertion_dl/assertion_model.py:48: BasicLSTMCell.__init__ (from tensorflow.python.keras.layers.legacy_rnn.rnn_cell_impl) is deprecated and will be removed in a future version.\nInstructions for updating:\nThis class is equivalent as tf.keras.layers.LSTMCell, and will be replaced by that in Tensorflow 2.0.\nWARNING:tensorflow:From /databricks/python/lib/python3.8/site-packages/sparknlp_jsl/_tf_graph_builders/tf2contrib/rnn.py:229: bidirectional_dynamic_rnn (from tensorflow.python.ops.rnn) is deprecated and will be removed in a future version.\nInstructions for updating:\nPlease use `keras.layers.Bidirectional(keras.layers.RNN(cell))`, which is equivalent to this API\nWARNING:tensorflow:From /databricks/python/lib/python3.8/site-packages/tensorflow/python/ops/rnn.py:438: dynamic_rnn (from tensorflow.python.ops.rnn) is deprecated and will be removed in a future version.\nInstructions for updating:\nPlease use `keras.layers.RNN(cell)`, which is equivalent to this API\nWARNING:tensorflow:From /databricks/python/lib/python3.8/site-packages/tensorflow/python/keras/layers/legacy_rnn/rnn_cell_impl.py:738: Layer.add_variable (from tensorflow.python.keras.engine.base_layer_v1) is deprecated and will be removed in a future version.\nInstructions for updating:\nPlease use `layer.add_weight` method instead.\nWARNING:tensorflow:From /databricks/python/lib/python3.8/site-packages/tensorflow/python/keras/layers/legacy_rnn/rnn_cell_impl.py:744: calling Zeros.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.\nInstructions for updating:\nCall initializer instance with the dtype argument instead of passing it to the constructor\nassertion_dl graph exported to /dbfs/assertion_tf_graphs/blstm_34_32_30_200_6.pb\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["%fs ls file:/dbfs/assertion_tf_graphs/"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"35b456e8-2b12-4054-847e-c56041ee6a46"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"overflow":false,"datasetInfos":[],"data":[["file:/dbfs/assertion_tf_graphs/blstm_34_32_30_200_6.pb","blstm_34_32_30_200_6.pb",1333335]],"plotOptions":{"displayType":"table","customPlotOptions":{},"pivotColumns":null,"pivotAggregation":null,"xColumns":null,"yColumns":null},"columnCustomDisplayInfos":{},"aggType":"","isJsonSchema":true,"removedWidgets":[],"aggSchema":[],"schema":[{"name":"path","type":"\"string\"","metadata":"{}"},{"name":"name","type":"\"string\"","metadata":"{}"},{"name":"size","type":"\"long\"","metadata":"{}"}],"aggError":"","aggData":[],"addedWidgets":{},"metadata":{"isDbfsCommandResult":false},"dbfsResultPath":null,"type":"table","aggOverflow":false,"aggSeriesLimitReached":false,"arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .table-result-container {\n    max-height: 300px;\n    overflow: auto;\n  }\n  table, th, td {\n    border: 1px solid black;\n    border-collapse: collapse;\n  }\n  th, td {\n    padding: 5px;\n  }\n  th {\n    text-align: left;\n  }\n</style><div class='table-result-container'><table class='table-result'><thead style='background-color: white'><tr><th>path</th><th>name</th><th>size</th></tr></thead><tbody><tr><td>file:/dbfs/assertion_tf_graphs/blstm_34_32_30_200_6.pb</td><td>blstm_34_32_30_200_6.pb</td><td>1333335</td></tr></tbody></table></div>"]}}],"execution_count":0},{"cell_type":"code","source":["assertionStatus = AssertionDLApproach()\\\n    .setLabelCol(\"label\")\\\n    .setInputCols(\"document\", \"chunk\", \"embeddings\")\\\n    .setOutputCol(\"assertion\")\\\n    .setBatchSize(128)\\\n    .setDropout(0.1)\\\n    .setLearningRate(0.001)\\\n    .setEpochs(50)\\\n    .setValidationSplit(0.2)\\\n    .setStartCol(\"start\")\\\n    .setEndCol(\"end\")\\\n    .setMaxSentLen(250)\\\n    .setEnableOutputLogs(True)\\\n    .setOutputLogsPath('dbfs:/assertion_logs')\\\n    .setGraphFolder('dbfs:/assertion_tf_graphs')\n\n\nclinical_assertion_pipeline = Pipeline(\n    stages = [\n    document,\n    chunk,\n    token,\n    embeddings,\n    assertionStatus])"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"19b3315c-86e7-4b86-9d0a-18dfd1b1ccc5"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["assertion_model = clinical_assertion_pipeline.fit(training_data)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"9327a04d-1037-4261-b766-951d9a902ee0"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["assertion_model.stages"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"88d0ad2e-b6ef-4141-9387-1bd7692eb125"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">Out[27]: [DocumentAssembler_3912fb43d16f,\n Doc2Chunk_7659725cd03d,\n REGEX_TOKENIZER_d68f222cda3d,\n WORD_EMBEDDINGS_MODEL_9004b1d00302,\n ASSERTION_DL_f46aadd1b4c0]</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Out[27]: [DocumentAssembler_3912fb43d16f,\n Doc2Chunk_7659725cd03d,\n REGEX_TOKENIZER_d68f222cda3d,\n WORD_EMBEDDINGS_MODEL_9004b1d00302,\n ASSERTION_DL_f46aadd1b4c0]</div>"]}}],"execution_count":0},{"cell_type":"code","source":["preds = assertion_model.transform(test_data).select('label','assertion.result')\n\npreds.show()"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"0bda53e4-8351-4171-b994-8e9f93b2e93c"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">+------------+--------------+\n|       label|        result|\n+------------+--------------+\n|     present|     [present]|\n|     present|     [present]|\n|     present|     [present]|\n|     present|     [present]|\n|     present|     [present]|\n|     present|     [present]|\n|     present|     [present]|\n|     present|     [present]|\n|     present|     [present]|\n|     present|     [present]|\n|hypothetical|[hypothetical]|\n|     present|     [present]|\n|     present|     [present]|\n|     present|     [present]|\n|     present|     [present]|\n|     present|     [present]|\n|     present|     [present]|\n|     present|     [present]|\n|     present|     [present]|\n|     present|     [present]|\n+------------+--------------+\nonly showing top 20 rows\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">+------------+--------------+\n       label|        result|\n+------------+--------------+\n     present|     [present]|\n     present|     [present]|\n     present|     [present]|\n     present|     [present]|\n     present|     [present]|\n     present|     [present]|\n     present|     [present]|\n     present|     [present]|\n     present|     [present]|\n     present|     [present]|\nhypothetical|[hypothetical]|\n     present|     [present]|\n     present|     [present]|\n     present|     [present]|\n     present|     [present]|\n     present|     [present]|\n     present|     [present]|\n     present|     [present]|\n     present|     [present]|\n     present|     [present]|\n+------------+--------------+\nonly showing top 20 rows\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["preds_df = preds.toPandas()"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"43bcfc6a-4c3a-40b8-8b26-5da14ce8682a"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["preds_df['result'] = preds_df['result'].apply(lambda x : x[0])\n"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"69926665-71df-4ee2-ab7e-bc70099feeba"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["# We are going to use sklearn to evalute the results on test dataset\nfrom sklearn.metrics import classification_report\n\nprint (classification_report(preds_df['result'], preds_df['label']))"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"d5d83d2b-95a4-4a14-a356-f85201bc641d"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">/databricks/python/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.\n  _warn_prf(average, modifier, msg_start, len(result))\n/databricks/python/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.\n  _warn_prf(average, modifier, msg_start, len(result))\n/databricks/python/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.\n  _warn_prf(average, modifier, msg_start, len(result))\n                              precision    recall  f1-score   support\n\n                      absent       0.61      0.74      0.67        38\nassociated_with_someone_else       0.00      0.00      0.00         0\n                 conditional       0.00      0.00      0.00         0\n                hypothetical       0.56      0.62      0.59         8\n                    possible       0.00      0.00      0.00         0\n                     present       0.92      0.78      0.84       140\n\n                    accuracy                           0.76       186\n                   macro avg       0.35      0.36      0.35       186\n                weighted avg       0.84      0.76      0.80       186\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">/databricks/python/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.\n  _warn_prf(average, modifier, msg_start, len(result))\n/databricks/python/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.\n  _warn_prf(average, modifier, msg_start, len(result))\n/databricks/python/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.\n  _warn_prf(average, modifier, msg_start, len(result))\n                              precision    recall  f1-score   support\n\n                      absent       0.61      0.74      0.67        38\nassociated_with_someone_else       0.00      0.00      0.00         0\n                 conditional       0.00      0.00      0.00         0\n                hypothetical       0.56      0.62      0.59         8\n                    possible       0.00      0.00      0.00         0\n                     present       0.92      0.78      0.84       140\n\n                    accuracy                           0.76       186\n                   macro avg       0.35      0.36      0.35       186\n                weighted avg       0.84      0.76      0.80       186\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["#saving the model that we've trained\nassertion_model.stages[-1].write().overwrite().save('/databricks/driver/models/custom_assertion_model')"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"ad145896-0369-40b2-8c05-85351675e7a0"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"mimetype":"text/x-python","name":"python","pygments_lexer":"ipython3","codemirror_mode":{"name":"ipython","version":3},"version":"3.8.3","nbconvert_exporter":"python","file_extension":".py"},"application/vnd.databricks.v1+notebook":{"notebookName":"5. Training and Reusing Assertion Status Models","dashboards":[],"notebookMetadata":{"pythonIndentUnit":2},"language":"python","widgets":{},"notebookOrigID":3018298012742877}},"nbformat":4,"nbformat_minor":0}
