{"cells":[{"cell_type":"markdown","source":["![JohnSnowLabs](https://nlp.johnsnowlabs.com/assets/images/logo.png)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"aec851fe-e39a-4743-b983-9930d110e2c6"}}},{"cell_type":"markdown","source":["# Training and Reusing Assertion Status Models"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"8e5732f0-beec-44b9-b4e5-42023927f9d4"}}},{"cell_type":"code","source":["import os\nimport json\nimport string\nimport numpy as np\nimport pandas as pd\n\n\nimport sparknlp\nimport sparknlp_jsl\nfrom sparknlp.base import *\nfrom sparknlp.util import *\nfrom sparknlp.annotator import *\nfrom sparknlp_jsl.annotator import *\nfrom sparknlp.pretrained import ResourceDownloader\n\nfrom pyspark.sql import functions as F\nfrom pyspark.ml import Pipeline, PipelineModel\n\npd.set_option('max_colwidth', 100)\npd.set_option('display.max_columns', 100)  \npd.set_option('display.expand_frame_repr', False)\n\nprint('sparknlp.version : ',sparknlp.version())\nprint('sparknlp_jsl.version : ',sparknlp_jsl.version())\n\nspark"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"c0f3e14c-398a-4ecd-864c-3b1c939acafd"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">sparknlp.version :  3.4.2\nsparknlp_jsl.version :  3.5.1\nOut[1]: </div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">sparknlp.version :  3.4.2\nsparknlp_jsl.version :  3.5.1\nOut[1]: </div>"]}},{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"\n            <div>\n                <p><b>SparkSession - hive</b></p>\n                \n        <div>\n            <p><b>SparkContext</b></p>\n\n            <p><a href=\"/?o=7956323724731612#setting/sparkui/0425-135237-t532foh4/driver-7456520259161290056\">Spark UI</a></p>\n\n            <dl>\n              <dt>Version</dt>\n                <dd><code>v3.1.2</code></dd>\n              <dt>Master</dt>\n                <dd><code>spark://10.139.64.6:7077</code></dd>\n              <dt>AppName</dt>\n                <dd><code>Databricks Shell</code></dd>\n            </dl>\n        </div>\n        \n            </div>\n        ","textData":null,"removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"htmlSandbox","arguments":{}}},"output_type":"display_data","data":{"text/html":["\n            <div>\n                <p><b>SparkSession - hive</b></p>\n                \n        <div>\n            <p><b>SparkContext</b></p>\n\n            <p><a href=\"/?o=7956323724731612#setting/sparkui/0425-135237-t532foh4/driver-7456520259161290056\">Spark UI</a></p>\n\n            <dl>\n              <dt>Version</dt>\n                <dd><code>v3.1.2</code></dd>\n              <dt>Master</dt>\n                <dd><code>spark://10.139.64.6:7077</code></dd>\n              <dt>AppName</dt>\n                <dd><code>Databricks Shell</code></dd>\n            </dl>\n        </div>\n        \n            </div>\n        "]}}],"execution_count":0},{"cell_type":"markdown","source":["# Clinical Assertion Model (with pretrained models)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"7868685d-d1a9-4896-8527-45da3ec252e7"}}},{"cell_type":"markdown","source":["The deep neural network architecture for assertion status detection in Spark NLP is based on a Bi-LSTM framework, and is a modified version of the architecture proposed by Federico Fancellu, Adam Lopez and Bonnie Webber ([Neural Networks For Negation Scope Detection](https://aclanthology.org/P16-1047.pdf)). Its goal is to classify the assertions made on given medical concepts as being present, absent, or possible in the patient, conditionally present in the patient under certain circumstances,\nhypothetically present in the patient at some future point, and\nmentioned in the patient report but associated with someoneelse.\nIn the proposed implementation, input units depend on the\ntarget tokens (a named entity) and the neighboring words that\nare explicitly encoded as a sequence using word embeddings.\nSimilar to paper mentioned above,  it is observed that that 95% of the scope tokens (neighboring words) fall in a window of 9 tokens to the left and 15\nto the right of the target tokens in the same dataset. Therefore, the same window size was implemented and it following parameters were used: learning\nrate 0.0012, dropout 0.05, batch size 64 and a maximum sentence length 250. The model has been implemented within\nSpark NLP as an annotator called AssertionDLModel. After\ntraining 20 epoch and measuring accuracy on the official test\nset, this implementation exceeds the latest state-of-the-art\naccuracy benchmarks as summarized as following table:\n\n|Assertion Label|Spark NLP|Latest Best|\n|-|-|-|\n|Absent       |0.944 |0.937|\n|Someone-else |0.904|0.869|\n|Conditional  |0.441|0.422|\n|Hypothetical |0.862|0.890|\n|Possible     |0.680|0.630|\n|Present      |0.953|0.957|\n|micro F1     |0.939|0.934|"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"ba94a7d4-f6d8-4e9e-a64d-ce0b08082a20"}}},{"cell_type":"markdown","source":["|    | model_name              |Predicted Entities|\n|---:|:------------------------|-|\n|  1 | assertion_dl            |Present, Absent, Possible, Planned, Someoneelse, Past, Family, None, Hypotetical|\n|  2 | assertion_dl_biobert    |absent, present, conditional, associated_with_someone_else, hypothetical, possible|\n|  3 | assertion_dl_healthcare |absent, present, conditional, associated_with_someone_else, hypothetical, possible|\n|  4 | assertion_dl_large      |hypothetical, present, absent, possible, conditional, associated_with_someone_else|\n|  5 | assertion_dl_radiology  |Confirmed, Suspected, Negative|\n|  6 | assertion_jsl           |Present, Absent, Possible, Planned, Someoneelse, Past, Family, None, Hypotetical|\n|  7 | assertion_jsl_large     |present, absent, possible, planned, someoneelse, past|\n|  8 | assertion_ml            |Hypothetical, Present, Absent, Possible, Conditional, Associated_with_someone_else|"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"bbe964c7-540f-48d3-ab59-b18b0b543941"}}},{"cell_type":"markdown","source":["## Pretrained `assertion_dl` model"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"09425c6d-5c14-4fa3-a719-98e20ab6c21e"}}},{"cell_type":"code","source":["# Annotator that transforms a text column from dataframe into an Annotation ready for NLP\n\nfrom sparknlp_jsl.annotator import *\n\ndocumentAssembler = DocumentAssembler()\\\n  .setInputCol(\"text\")\\\n  .setOutputCol(\"document\")\n\n# Sentence Detector annotator, processes various sentences per line\n\nsentenceDetector = SentenceDetector()\\\n  .setInputCols([\"document\"])\\\n  .setOutputCol(\"sentence\")\n\n# Tokenizer splits words in a relevant format for NLP\n\ntokenizer = Tokenizer()\\\n  .setInputCols([\"sentence\"])\\\n  .setOutputCol(\"token\")\n\n# Clinical word embeddings trained on PubMED dataset\nword_embeddings = WordEmbeddingsModel.pretrained(\"embeddings_clinical\", \"en\", \"clinical/models\")\\\n  .setInputCols([\"sentence\", \"token\"])\\\n  .setOutputCol(\"embeddings\")\n\n# NER model trained on i2b2 (sampled from MIMIC) dataset\nclinical_ner = MedicalNerModel.pretrained(\"ner_clinical\", \"en\", \"clinical/models\") \\\n  .setInputCols([\"sentence\", \"token\", \"embeddings\"]) \\\n  .setOutputCol(\"ner\")\n\nner_converter = NerConverter() \\\n  .setInputCols([\"sentence\", \"token\", \"ner\"]) \\\n  .setOutputCol(\"ner_chunk\")\n\n# Assertion model trained on i2b2 (sampled from MIMIC) dataset\n# coming from sparknlp_jsl.annotator !!\nclinical_assertion = AssertionDLModel.pretrained(\"assertion_dl\", \"en\", \"clinical/models\") \\\n  .setInputCols([\"sentence\", \"ner_chunk\", \"embeddings\"]) \\\n  .setOutputCol(\"assertion\")\n\nnlpPipeline = Pipeline(stages=[\n    documentAssembler, \n    sentenceDetector,\n    tokenizer,\n    word_embeddings,\n    clinical_ner,\n    ner_converter,\n    clinical_assertion\n    ])\n\nempty_data = spark.createDataFrame([[\"\"]]).toDF(\"text\")\nmodel = nlpPipeline.fit(empty_data)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"064d576d-3e0b-497b-810d-c6e8c363a0c5"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">embeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\nner_clinical download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">embeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\nner_clinical download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["# we also have a LogReg based Assertion Model.\n'''\nclinical_assertion_ml = AssertionLogRegModel.pretrained(\"assertion_ml\", \"en\", \"clinical/models\") \\\n    .setInputCols([\"sentence\", \"ner_chunk\", \"embeddings\"]) \\\n    .setOutputCol(\"assertion\")\n'''"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"0bede1e8-a98e-4774-ad2e-5863b56bd636"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">Out[3]: &#39;\\nclinical_assertion_ml = AssertionLogRegModel.pretrained(&#34;assertion_ml&#34;, &#34;en&#34;, &#34;clinical/models&#34;)     .setInputCols([&#34;sentence&#34;, &#34;ner_chunk&#34;, &#34;embeddings&#34;])     .setOutputCol(&#34;assertion&#34;)\\n&#39;</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Out[3]: &#39;\\nclinical_assertion_ml = AssertionLogRegModel.pretrained(&#34;assertion_ml&#34;, &#34;en&#34;, &#34;clinical/models&#34;)     .setInputCols([&#34;sentence&#34;, &#34;ner_chunk&#34;, &#34;embeddings&#34;])     .setOutputCol(&#34;assertion&#34;)\\n&#39;</div>"]}}],"execution_count":0},{"cell_type":"code","source":["import pandas as pd\n\ntext = 'Patient has a headache for the last 2 weeks, needs to get a head CT, and appears anxious when she walks fast. No alopecia noted. She denies pain'\n\nprint (text)\n\nlight_model = LightPipeline(model)\n\nlight_result = light_model.fullAnnotate(text)[0]\n\nchunks=[]\nentities=[]\nstatus=[]\n\nfor n,m in zip(light_result['ner_chunk'],light_result['assertion']):\n    \n    chunks.append(n.result)\n    entities.append(n.metadata['entity']) \n    status.append(m.result)\n        \ndf = pd.DataFrame({'chunks':chunks, 'entities':entities, 'assertion':status})\n\ndf"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"6137f486-8004-4553-bb12-b2f01df9ab93"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">Patient has a headache for the last 2 weeks, needs to get a head CT, and appears anxious when she walks fast. No alopecia noted. She denies pain\nOut[4]: </div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Patient has a headache for the last 2 weeks, needs to get a head CT, and appears anxious when she walks fast. No alopecia noted. She denies pain\nOut[4]: </div>"]}},{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chunks</th>\n      <th>entities</th>\n      <th>assertion</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>a headache</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>a head CT</td>\n      <td>TEST</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>anxious</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>alopecia</td>\n      <td>PROBLEM</td>\n      <td>absent</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>pain</td>\n      <td>PROBLEM</td>\n      <td>absent</td>\n    </tr>\n  </tbody>\n</table>\n</div>","textData":null,"removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"htmlSandbox","arguments":{}}},"output_type":"display_data","data":{"text/html":["<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chunks</th>\n      <th>entities</th>\n      <th>assertion</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>a headache</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>a head CT</td>\n      <td>TEST</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>anxious</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>alopecia</td>\n      <td>PROBLEM</td>\n      <td>absent</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>pain</td>\n      <td>PROBLEM</td>\n      <td>absent</td>\n    </tr>\n  </tbody>\n</table>\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["from sparknlp_display import AssertionVisualizer\n\nvisualizer = AssertionVisualizer()\n\nvis = visualizer.display(light_result, 'ner_chunk', 'assertion', return_html=True)\n#visualizer.set_label_colors({'TREATMENT':'#008080', 'PROBLEM':'#800080'})\n\n\ndisplayHTML(vis)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"352bef06-8f6b-4099-983a-2f64bb4c0c55"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"\n<style>\n    @import url('https://fonts.googleapis.com/css2?family=Montserrat:wght@300;400;500;600;700&display=swap');\n    @import url('https://fonts.googleapis.com/css2?family=Vistol Regular:wght@300;400;500;600;700&display=swap');\n    \n    .spark-nlp-display-scroll-entities {\n        border: 1px solid #E7EDF0;\n        border-radius: 3px;\n        text-align: justify;\n        \n    }\n    .spark-nlp-display-scroll-entities span {  \n        font-size: 14px;\n        line-height: 24px;\n        color: #536B76;\n        font-family: 'Montserrat', sans-serif !important;\n    }\n    \n    .spark-nlp-display-entity-wrapper{\n    \n        display: inline-grid;\n        text-align: center;\n        border-radius: 4px;\n        margin: 0 2px 5px 2px;\n        padding: 1px\n    }\n    .spark-nlp-display-entity-name{\n        font-size: 14px;\n        line-height: 24px;\n        font-family: 'Montserrat', sans-serif !important;\n        \n        background: #f1f2f3;\n        border-width: medium;\n        text-align: center;\n        \n        font-weight: 400;\n        \n        border-radius: 5px;\n        padding: 2px 5px;\n        display: block;\n        margin: 3px 2px;\n    \n    }\n    .spark-nlp-display-entity-type{\n        font-size: 14px;\n        line-height: 24px;\n        color: #ffffff;\n        font-family: 'Montserrat', sans-serif !important;\n        \n        text-transform: uppercase;\n        \n        font-weight: 500;\n\n        display: block;\n        padding: 3px 5px;\n    }\n    \n    .spark-nlp-display-entity-resolution{\n        font-size: 14px;\n        line-height: 24px;\n        color: #ffffff;\n        font-family: 'Vistol Regular', sans-serif !important;\n        \n        text-transform: uppercase;\n        \n        font-weight: 500;\n\n        display: block;\n        padding: 3px 5px;\n    }\n    \n    .spark-nlp-display-others{\n        font-size: 14px;\n        line-height: 24px;\n        font-family: 'Montserrat', sans-serif !important;\n        \n        font-weight: 400;\n    }\n\n</style>\n <span class=\"spark-nlp-display-others\" style=\"background-color: white\">Patient has </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #800080B3\"><span class=\"spark-nlp-display-entity-name\">a headache </span><span class=\"spark-nlp-display-entity-type\">PROBLEM</span><span class=\"spark-nlp-display-entity-resolution\" style=\"background-color: #800080FF\">present </span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> for the last 2 weeks, needs to get </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #77b5feB3\"><span class=\"spark-nlp-display-entity-name\">a head CT </span><span class=\"spark-nlp-display-entity-type\">TEST</span><span class=\"spark-nlp-display-entity-resolution\" style=\"background-color: #77b5feFF\">present </span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\">, and appears </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #800080B3\"><span class=\"spark-nlp-display-entity-name\">anxious </span><span class=\"spark-nlp-display-entity-type\">PROBLEM</span><span class=\"spark-nlp-display-entity-resolution\" style=\"background-color: #800080FF\">present </span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> when she walks fast. No </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #800080B3\"><span class=\"spark-nlp-display-entity-name\">alopecia </span><span class=\"spark-nlp-display-entity-type\">PROBLEM</span><span class=\"spark-nlp-display-entity-resolution\" style=\"background-color: #800080FF\">absent </span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> noted. She denies </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #800080B3\"><span class=\"spark-nlp-display-entity-name\">pain </span><span class=\"spark-nlp-display-entity-type\">PROBLEM</span><span class=\"spark-nlp-display-entity-resolution\" style=\"background-color: #800080FF\">absent </span></span></div>","textData":null,"removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"htmlSandbox","arguments":{}}},"output_type":"display_data","data":{"text/html":["\n<style>\n    @import url('https://fonts.googleapis.com/css2?family=Montserrat:wght@300;400;500;600;700&display=swap');\n    @import url('https://fonts.googleapis.com/css2?family=Vistol Regular:wght@300;400;500;600;700&display=swap');\n    \n    .spark-nlp-display-scroll-entities {\n        border: 1px solid #E7EDF0;\n        border-radius: 3px;\n        text-align: justify;\n        \n    }\n    .spark-nlp-display-scroll-entities span {  \n        font-size: 14px;\n        line-height: 24px;\n        color: #536B76;\n        font-family: 'Montserrat', sans-serif !important;\n    }\n    \n    .spark-nlp-display-entity-wrapper{\n    \n        display: inline-grid;\n        text-align: center;\n        border-radius: 4px;\n        margin: 0 2px 5px 2px;\n        padding: 1px\n    }\n    .spark-nlp-display-entity-name{\n        font-size: 14px;\n        line-height: 24px;\n        font-family: 'Montserrat', sans-serif !important;\n        \n        background: #f1f2f3;\n        border-width: medium;\n        text-align: center;\n        \n        font-weight: 400;\n        \n        border-radius: 5px;\n        padding: 2px 5px;\n        display: block;\n        margin: 3px 2px;\n    \n    }\n    .spark-nlp-display-entity-type{\n        font-size: 14px;\n        line-height: 24px;\n        color: #ffffff;\n        font-family: 'Montserrat', sans-serif !important;\n        \n        text-transform: uppercase;\n        \n        font-weight: 500;\n\n        display: block;\n        padding: 3px 5px;\n    }\n    \n    .spark-nlp-display-entity-resolution{\n        font-size: 14px;\n        line-height: 24px;\n        color: #ffffff;\n        font-family: 'Vistol Regular', sans-serif !important;\n        \n        text-transform: uppercase;\n        \n        font-weight: 500;\n\n        display: block;\n        padding: 3px 5px;\n    }\n    \n    .spark-nlp-display-others{\n        font-size: 14px;\n        line-height: 24px;\n        font-family: 'Montserrat', sans-serif !important;\n        \n        font-weight: 400;\n    }\n\n</style>\n <span class=\"spark-nlp-display-others\" style=\"background-color: white\">Patient has </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #800080B3\"><span class=\"spark-nlp-display-entity-name\">a headache </span><span class=\"spark-nlp-display-entity-type\">PROBLEM</span><span class=\"spark-nlp-display-entity-resolution\" style=\"background-color: #800080FF\">present </span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> for the last 2 weeks, needs to get </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #77b5feB3\"><span class=\"spark-nlp-display-entity-name\">a head CT </span><span class=\"spark-nlp-display-entity-type\">TEST</span><span class=\"spark-nlp-display-entity-resolution\" style=\"background-color: #77b5feFF\">present </span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\">, and appears </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #800080B3\"><span class=\"spark-nlp-display-entity-name\">anxious </span><span class=\"spark-nlp-display-entity-type\">PROBLEM</span><span class=\"spark-nlp-display-entity-resolution\" style=\"background-color: #800080FF\">present </span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> when she walks fast. No </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #800080B3\"><span class=\"spark-nlp-display-entity-name\">alopecia </span><span class=\"spark-nlp-display-entity-type\">PROBLEM</span><span class=\"spark-nlp-display-entity-resolution\" style=\"background-color: #800080FF\">absent </span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> noted. She denies </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #800080B3\"><span class=\"spark-nlp-display-entity-name\">pain </span><span class=\"spark-nlp-display-entity-type\">PROBLEM</span><span class=\"spark-nlp-display-entity-resolution\" style=\"background-color: #800080FF\">absent </span></span></div>"]}}],"execution_count":0},{"cell_type":"code","source":["! wget -q\thttps://s3.amazonaws.com/auxdata.johnsnowlabs.com/public/resources/en/pubmed/pubmed_sample_text_small.csv -P /dbfs/"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"f1e1c581-7c0e-4b56-be6e-d604278e3d18"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["import pyspark.sql.functions as F\n\npubMedDF = spark.read\\\n                .option(\"header\", \"true\")\\\n                .csv(\"/pubmed_sample_text_small.csv\")\\\n                \npubMedDF.show(truncate=50)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"1321721d-d896-4179-87c9-6716a68ccba4"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">+--------------------------------------------------+\n|                                              text|\n+--------------------------------------------------+\n|The human KCNJ9 (Kir 3.3, GIRK3) is a member of...|\n|BACKGROUND: At present, it is one of the most i...|\n|OBJECTIVE: To investigate the relationship betw...|\n|Combined EEG/fMRI recording has been used to lo...|\n|Kohlschutter syndrome is a rare neurodegenerati...|\n|Statistical analysis of neuroimages is commonly...|\n|The synthetic DOX-LNA conjugate was characteriz...|\n|Our objective was to compare three different me...|\n|We conducted a phase II study to assess the eff...|\n|&#34;&#34;&#34;Monomeric sarcosine oxidase (MSOX) is a flav...|\n|We presented the tachinid fly Exorista japonica...|\n|The literature dealing with the water conductin...|\n|A novel approach to synthesize chitosan-O-isopr...|\n|An HPLC-ESI-MS-MS method has been developed for...|\n|The localizing and lateralizing values of eye a...|\n|OBJECTIVE: To evaluate the effectiveness and ac...|\n|For the construction of new combinatorial libra...|\n|We report the results of a screen for genetic a...|\n|Intraparenchymal pericatheter cyst is rarely re...|\n|It is known that patients with Klinefelter&#39;s sy...|\n+--------------------------------------------------+\nonly showing top 20 rows\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">+--------------------------------------------------+\n                                              text|\n+--------------------------------------------------+\nThe human KCNJ9 (Kir 3.3, GIRK3) is a member of...|\nBACKGROUND: At present, it is one of the most i...|\nOBJECTIVE: To investigate the relationship betw...|\nCombined EEG/fMRI recording has been used to lo...|\nKohlschutter syndrome is a rare neurodegenerati...|\nStatistical analysis of neuroimages is commonly...|\nThe synthetic DOX-LNA conjugate was characteriz...|\nOur objective was to compare three different me...|\nWe conducted a phase II study to assess the eff...|\n&#34;&#34;&#34;Monomeric sarcosine oxidase (MSOX) is a flav...|\nWe presented the tachinid fly Exorista japonica...|\nThe literature dealing with the water conductin...|\nA novel approach to synthesize chitosan-O-isopr...|\nAn HPLC-ESI-MS-MS method has been developed for...|\nThe localizing and lateralizing values of eye a...|\nOBJECTIVE: To evaluate the effectiveness and ac...|\nFor the construction of new combinatorial libra...|\nWe report the results of a screen for genetic a...|\nIntraparenchymal pericatheter cyst is rarely re...|\nIt is known that patients with Klinefelter&#39;s sy...|\n+--------------------------------------------------+\nonly showing top 20 rows\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["result = model.transform(pubMedDF.limit(100))"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"dd660faf-5874-42eb-99ef-a083952d91ab"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["result.show()"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"7bf3ef67-9448-43e4-9d67-d7a47399911a"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+\n|                text|            document|            sentence|               token|          embeddings|                 ner|           ner_chunk|           assertion|\n+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+\n|The human KCNJ9 (...|[{document, 0, 95...|[{document, 0, 12...|[{token, 0, 2, Th...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 48, 106,...|[{assertion, 48, ...|\n|BACKGROUND: At pr...|[{document, 0, 14...|[{document, 0, 19...|[{token, 0, 9, BA...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 67, 79, ...|[{assertion, 67, ...|\n|OBJECTIVE: To inv...|[{document, 0, 15...|[{document, 0, 30...|[{token, 0, 8, OB...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 188, 231...|[{assertion, 188,...|\n|Combined EEG/fMRI...|[{document, 0, 16...|[{document, 0, 16...|[{token, 0, 7, Co...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 0, 26, C...|[{assertion, 0, 2...|\n|Kohlschutter synd...|[{document, 0, 25...|[{document, 0, 20...|[{token, 0, 11, K...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 0, 20, K...|[{assertion, 0, 2...|\n|Statistical analy...|[{document, 0, 10...|[{document, 0, 34...|[{token, 0, 10, S...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 0, 34, S...|[{assertion, 0, 3...|\n|The synthetic DOX...|[{document, 0, 57...|[{document, 0, 10...|[{token, 0, 2, Th...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 0, 30, T...|[{assertion, 0, 3...|\n|Our objective was...|[{document, 0, 24...|[{document, 0, 19...|[{token, 0, 2, Ou...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 56, 80, ...|[{assertion, 56, ...|\n|We conducted a ph...|[{document, 0, 14...|[{document, 0, 24...|[{token, 0, 1, We...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 13, 28, ...|[{assertion, 13, ...|\n|&#34;&#34;&#34;Monomeric sarc...|[{document, 0, 14...|[{document, 0, 14...|[{token, 0, 2, &#34;&#34;...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 3, 36, M...|[{assertion, 3, 3...|\n|We presented the ...|[{document, 0, 12...|[{document, 0, 26...|[{token, 0, 1, We...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 13, 46, ...|[{assertion, 13, ...|\n|The literature de...|[{document, 0, 16...|[{document, 0, 13...|[{token, 0, 2, Th...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 154, 166...|[{assertion, 154,...|\n|A novel approach ...|[{document, 0, 64...|[{document, 0, 97...|[{token, 0, 0, A,...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 99, 124,...|[{assertion, 99, ...|\n|An HPLC-ESI-MS-MS...|[{document, 0, 90...|[{document, 0, 24...|[{token, 0, 1, An...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 0, 23, A...|[{assertion, 0, 2...|\n|The localizing an...|[{document, 0, 72...|[{document, 0, 12...|[{token, 0, 2, Th...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 50, 70, ...|[{assertion, 50, ...|\n|OBJECTIVE: To eva...|[{document, 0, 13...|[{document, 0, 43...|[{token, 0, 8, OB...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 154, 186...|[{assertion, 154,...|\n|For the construct...|[{document, 0, 32...|[{document, 0, 20...|[{token, 0, 2, Fo...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 24, 50, ...|[{assertion, 24, ...|\n|We report the res...|[{document, 0, 13...|[{document, 0, 38...|[{token, 0, 1, We...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 25, 32, ...|[{assertion, 25, ...|\n|Intraparenchymal ...|[{document, 0, 10...|[{document, 0, 53...|[{token, 0, 15, I...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 0, 33, I...|[{assertion, 0, 3...|\n|It is known that ...|[{document, 0, 34...|[{document, 0, 14...|[{token, 0, 1, It...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 31, 52, ...|[{assertion, 31, ...|\n+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+\nonly showing top 20 rows\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+\n                text|            document|            sentence|               token|          embeddings|                 ner|           ner_chunk|           assertion|\n+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+\nThe human KCNJ9 (...|[{document, 0, 95...|[{document, 0, 12...|[{token, 0, 2, Th...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 48, 106,...|[{assertion, 48, ...|\nBACKGROUND: At pr...|[{document, 0, 14...|[{document, 0, 19...|[{token, 0, 9, BA...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 67, 79, ...|[{assertion, 67, ...|\nOBJECTIVE: To inv...|[{document, 0, 15...|[{document, 0, 30...|[{token, 0, 8, OB...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 188, 231...|[{assertion, 188,...|\nCombined EEG/fMRI...|[{document, 0, 16...|[{document, 0, 16...|[{token, 0, 7, Co...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 0, 26, C...|[{assertion, 0, 2...|\nKohlschutter synd...|[{document, 0, 25...|[{document, 0, 20...|[{token, 0, 11, K...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 0, 20, K...|[{assertion, 0, 2...|\nStatistical analy...|[{document, 0, 10...|[{document, 0, 34...|[{token, 0, 10, S...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 0, 34, S...|[{assertion, 0, 3...|\nThe synthetic DOX...|[{document, 0, 57...|[{document, 0, 10...|[{token, 0, 2, Th...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 0, 30, T...|[{assertion, 0, 3...|\nOur objective was...|[{document, 0, 24...|[{document, 0, 19...|[{token, 0, 2, Ou...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 56, 80, ...|[{assertion, 56, ...|\nWe conducted a ph...|[{document, 0, 14...|[{document, 0, 24...|[{token, 0, 1, We...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 13, 28, ...|[{assertion, 13, ...|\n&#34;&#34;&#34;Monomeric sarc...|[{document, 0, 14...|[{document, 0, 14...|[{token, 0, 2, &#34;&#34;...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 3, 36, M...|[{assertion, 3, 3...|\nWe presented the ...|[{document, 0, 12...|[{document, 0, 26...|[{token, 0, 1, We...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 13, 46, ...|[{assertion, 13, ...|\nThe literature de...|[{document, 0, 16...|[{document, 0, 13...|[{token, 0, 2, Th...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 154, 166...|[{assertion, 154,...|\nA novel approach ...|[{document, 0, 64...|[{document, 0, 97...|[{token, 0, 0, A,...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 99, 124,...|[{assertion, 99, ...|\nAn HPLC-ESI-MS-MS...|[{document, 0, 90...|[{document, 0, 24...|[{token, 0, 1, An...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 0, 23, A...|[{assertion, 0, 2...|\nThe localizing an...|[{document, 0, 72...|[{document, 0, 12...|[{token, 0, 2, Th...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 50, 70, ...|[{assertion, 50, ...|\nOBJECTIVE: To eva...|[{document, 0, 13...|[{document, 0, 43...|[{token, 0, 8, OB...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 154, 186...|[{assertion, 154,...|\nFor the construct...|[{document, 0, 32...|[{document, 0, 20...|[{token, 0, 2, Fo...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 24, 50, ...|[{assertion, 24, ...|\nWe report the res...|[{document, 0, 13...|[{document, 0, 38...|[{token, 0, 1, We...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 25, 32, ...|[{assertion, 25, ...|\nIntraparenchymal ...|[{document, 0, 10...|[{document, 0, 53...|[{token, 0, 15, I...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 0, 33, I...|[{assertion, 0, 3...|\nIt is known that ...|[{document, 0, 34...|[{document, 0, 14...|[{token, 0, 1, It...|[{word_embeddings...|[{named_entity, 0...|[{chunk, 31, 52, ...|[{assertion, 31, ...|\n+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+\nonly showing top 20 rows\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["result.select('sentence.result').take(1)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"cd49fbf0-b7d7-4523-b08c-d02fc6271ffb"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">Out[10]: [Row(result=[&#39;The human KCNJ9 (Kir 3.3, GIRK3) is a member of the G-protein-activated inwardly rectifying potassium (GIRK) channel family.&#39;, &#39;Here we describe the genomicorganization of the KCNJ9 locus on chromosome 1q21-23 as a candidate gene forType II diabetes mellitus in the Pima Indian population.&#39;, &#39;The gene spansapproximately 7.6 kb and contains one noncoding and two coding exons separated byapproximately 2.2 and approximately 2.6 kb introns, respectively.&#39;, &#39;We identified14 single nucleotide polymorphisms (SNPs), including one that predicts aVal366Ala substitution, and an 8 base-pair&#39;, &#39;(bp) insertion/deletion.&#39;, &#39;Ourexpression studies revealed the presence of the transcript in various humantissues including pancreas, and two major insulin-responsive tissues: fat andskeletal muscle.&#39;, &#39;The characterization of the KCNJ9 gene should facilitate furtherstudies on the function of the KCNJ9 protein and allow evaluation of thepotential role of the locus in Type II diabetes.&#39;])]</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Out[10]: [Row(result=[&#39;The human KCNJ9 (Kir 3.3, GIRK3) is a member of the G-protein-activated inwardly rectifying potassium (GIRK) channel family.&#39;, &#39;Here we describe the genomicorganization of the KCNJ9 locus on chromosome 1q21-23 as a candidate gene forType II diabetes mellitus in the Pima Indian population.&#39;, &#39;The gene spansapproximately 7.6 kb and contains one noncoding and two coding exons separated byapproximately 2.2 and approximately 2.6 kb introns, respectively.&#39;, &#39;We identified14 single nucleotide polymorphisms (SNPs), including one that predicts aVal366Ala substitution, and an 8 base-pair&#39;, &#39;(bp) insertion/deletion.&#39;, &#39;Ourexpression studies revealed the presence of the transcript in various humantissues including pancreas, and two major insulin-responsive tissues: fat andskeletal muscle.&#39;, &#39;The characterization of the KCNJ9 gene should facilitate furtherstudies on the function of the KCNJ9 protein and allow evaluation of thepotential role of the locus in Type II diabetes.&#39;])]</div>"]}}],"execution_count":0},{"cell_type":"code","source":["result.select(F.explode(F.arrays_zip(result.ner_chunk.result,\n                                     result.ner_chunk.begin, \n                                     result.ner_chunk.end,\n                                     result.ner_chunk.metadata,\n                                     result.assertion.result)).alias(\"cols\")) \\\n      .select(F.expr(\"cols['0']\").alias(\"chunk\"),\n              F.expr(\"cols['1']\").alias(\"begin\"),\n              F.expr(\"cols['2']\").alias(\"end\"),\n              F.expr(\"cols['3']['entity']\").alias(\"ner_label\"),\n              F.expr(\"cols['3']['sentence']\").alias(\"sent_id\"),\n              F.expr(\"cols['4']\").alias(\"assertion\"),).show(truncate=False)\n"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"6bbadeea-6168-42ca-b500-bdaa548126d2"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">+-----------------------------------------------------------+-----+---+---------+-------+-----------+\n|chunk                                                      |begin|end|ner_label|sent_id|assertion  |\n+-----------------------------------------------------------+-----+---+---------+-------+-----------+\n|the G-protein-activated inwardly rectifying potassium (GIRK|48   |106|TREATMENT|0      |conditional|\n|the genomicorganization                                    |142  |164|TREATMENT|1      |present    |\n|a candidate gene forType II diabetes mellitus              |210  |254|PROBLEM  |1      |present    |\n|byapproximately                                            |380  |394|TREATMENT|2      |present    |\n|single nucleotide polymorphisms                            |464  |494|TREATMENT|3      |present    |\n|aVal366Ala substitution                                    |532  |554|PROBLEM  |3      |present    |\n|an 8 base-pair                                             |561  |574|PROBLEM  |3      |present    |\n|insertion/deletion                                         |581  |598|PROBLEM  |4      |present    |\n|Ourexpression studies                                      |601  |621|TEST     |5      |present    |\n|the transcript in various humantissues                     |648  |685|PROBLEM  |5      |present    |\n|fat andskeletal muscle                                     |749  |770|PROBLEM  |5      |possible   |\n|furtherstudies                                             |830  |843|PROBLEM  |6      |present    |\n|the KCNJ9 protein                                          |864  |880|TREATMENT|6      |present    |\n|evaluation                                                 |892  |901|TEST     |6      |possible   |\n|Type II diabetes                                           |940  |955|PROBLEM  |6      |present    |\n|the treatment                                              |67   |79 |TREATMENT|0      |present    |\n|breast cancer                                              |84   |96 |PROBLEM  |0      |present    |\n|the standard therapy                                       |109  |128|TREATMENT|0      |present    |\n|anthracyclines                                             |167  |180|TREATMENT|0      |present    |\n|taxanes                                                    |186  |192|TREATMENT|0      |present    |\n+-----------------------------------------------------------+-----+---+---------+-------+-----------+\nonly showing top 20 rows\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">+-----------------------------------------------------------+-----+---+---------+-------+-----------+\nchunk                                                      |begin|end|ner_label|sent_id|assertion  |\n+-----------------------------------------------------------+-----+---+---------+-------+-----------+\nthe G-protein-activated inwardly rectifying potassium (GIRK|48   |106|TREATMENT|0      |conditional|\nthe genomicorganization                                    |142  |164|TREATMENT|1      |present    |\na candidate gene forType II diabetes mellitus              |210  |254|PROBLEM  |1      |present    |\nbyapproximately                                            |380  |394|TREATMENT|2      |present    |\nsingle nucleotide polymorphisms                            |464  |494|TREATMENT|3      |present    |\naVal366Ala substitution                                    |532  |554|PROBLEM  |3      |present    |\nan 8 base-pair                                             |561  |574|PROBLEM  |3      |present    |\ninsertion/deletion                                         |581  |598|PROBLEM  |4      |present    |\nOurexpression studies                                      |601  |621|TEST     |5      |present    |\nthe transcript in various humantissues                     |648  |685|PROBLEM  |5      |present    |\nfat andskeletal muscle                                     |749  |770|PROBLEM  |5      |possible   |\nfurtherstudies                                             |830  |843|PROBLEM  |6      |present    |\nthe KCNJ9 protein                                          |864  |880|TREATMENT|6      |present    |\nevaluation                                                 |892  |901|TEST     |6      |possible   |\nType II diabetes                                           |940  |955|PROBLEM  |6      |present    |\nthe treatment                                              |67   |79 |TREATMENT|0      |present    |\nbreast cancer                                              |84   |96 |PROBLEM  |0      |present    |\nthe standard therapy                                       |109  |128|TREATMENT|0      |present    |\nanthracyclines                                             |167  |180|TREATMENT|0      |present    |\ntaxanes                                                    |186  |192|TREATMENT|0      |present    |\n+-----------------------------------------------------------+-----+---+---------+-------+-----------+\nonly showing top 20 rows\n\n</div>"]}}],"execution_count":0},{"cell_type":"markdown","source":["## Pretrained `assertion_dl_radiology` model"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"f80c498d-3329-409b-b33e-f1b020314d58"}}},{"cell_type":"code","source":["from sparknlp_jsl.annotator import *\n\ndocumentAssembler = DocumentAssembler()\\\n    .setInputCol(\"text\")\\\n    .setOutputCol(\"document\")\n\n# Sentence Detector annotator, processes various sentences per line\nsentenceDetector = SentenceDetectorDLModel\\\n    .pretrained(\"sentence_detector_dl_healthcare\",\"en\",\"clinical/models\") \\\n    .setInputCols([\"document\"]) \\\n    .setOutputCol(\"sentence\")\n\n# Tokenizer splits words in a relevant format for NLP\ntokenizer = Tokenizer()\\\n    .setInputCols([\"sentence\"])\\\n    .setOutputCol(\"token\")\n\n# Clinical word embeddings trained on PubMED dataset\nword_embeddings = WordEmbeddingsModel.pretrained(\"embeddings_clinical\", \"en\", \"clinical/models\")\\\n    .setInputCols([\"sentence\", \"token\"])\\\n    .setOutputCol(\"embeddings\")\n\n# NER model for radiology\nradiology_ner = MedicalNerModel.pretrained(\"ner_radiology\", \"en\", \"clinical/models\") \\\n    .setInputCols([\"sentence\", \"token\", \"embeddings\"]) \\\n    .setOutputCol(\"ner\")\n\nner_converter = NerConverter() \\\n    .setInputCols([\"sentence\", \"token\", \"ner\"]) \\\n    .setOutputCol(\"ner_chunk\")\\\n    .setWhiteList([\"ImagingFindings\"])\n\n# Assertion model trained on radiology dataset\n# coming from sparknlp_jsl.annotator !!\n\nradiology_assertion = AssertionDLModel.pretrained(\"assertion_dl_radiology\", \"en\", \"clinical/models\") \\\n    .setInputCols([\"sentence\", \"ner_chunk\", \"embeddings\"]) \\\n    .setOutputCol(\"assertion\")\n\nnlpPipeline = Pipeline(stages=[\n    documentAssembler, \n    sentenceDetector,\n    tokenizer,\n    word_embeddings,\n    radiology_ner,\n    ner_converter,\n    radiology_assertion\n    ])\n\nempty_data = spark.createDataFrame([[\"\"]]).toDF(\"text\")\nradiologyAssertion_model = nlpPipeline.fit(empty_data)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"565efb2d-2d61-4965-8b3c-81d75dd3c93e"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">sentence_detector_dl_healthcare download started this may take some time.\nApproximate size to download 367.3 KB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\nner_radiology download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[OK!]\nassertion_dl_radiology download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">sentence_detector_dl_healthcare download started this may take some time.\nApproximate size to download 367.3 KB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\nner_radiology download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[OK!]\nassertion_dl_radiology download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["# A sample text from a radiology report\n\ntext = \"\"\"No right-sided pleural effusion or pneumothorax is definitively seen and there are mildly displaced fractures of the left lateral 8th and likely 9th ribs.\"\"\""],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"5910a402-1986-4513-9d70-65d67aefa0a6"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["data = spark.createDataFrame([[text]]).toDF(\"text\")"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"9aa1ee2f-efc2-441a-8836-0377e5f1eb41"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["result = radiologyAssertion_model.transform(data)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"04cd2f55-bfb5-4063-83f7-b16622ff1491"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["result.select(F.explode(F.arrays_zip(result.ner_chunk.result, \n                                     result.ner_chunk.metadata, \n                                     result.assertion.result)).alias(\"cols\")) \\\n      .select(F.expr(\"cols['0']\").alias(\"chunk\"),\n              F.expr(\"cols['1']['entity']\").alias(\"ner_label\"),\n              F.expr(\"cols['1']['sentence']\").alias(\"sent_id\"),\n              F.expr(\"cols['2']\").alias(\"assertion\")).show(truncate=False)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"9de0b7bb-0516-4b76-9241-ffae0acd5080"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">+-------------------+---------------+-------+---------+\n|chunk              |ner_label      |sent_id|assertion|\n+-------------------+---------------+-------+---------+\n|effusion           |ImagingFindings|0      |Negative |\n|pneumothorax       |ImagingFindings|0      |Negative |\n|displaced fractures|ImagingFindings|0      |Confirmed|\n+-------------------+---------------+-------+---------+\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">+-------------------+---------------+-------+---------+\nchunk              |ner_label      |sent_id|assertion|\n+-------------------+---------------+-------+---------+\neffusion           |ImagingFindings|0      |Negative |\npneumothorax       |ImagingFindings|0      |Negative |\ndisplaced fractures|ImagingFindings|0      |Confirmed|\n+-------------------+---------------+-------+---------+\n\n</div>"]}}],"execution_count":0},{"cell_type":"markdown","source":["## Writing a generic Assertion + NER function"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"afef480c-1540-4acc-88c9-5711e5290f3b"}}},{"cell_type":"code","source":["from pyspark.sql.functions import monotonically_increasing_id\n\n\ndef get_base_pipeline (embeddings = 'embeddings_clinical'):\n\n    documentAssembler = DocumentAssembler()\\\n        .setInputCol(\"text\")\\\n        .setOutputCol(\"document\")\n\n  # Sentence Detector annotator, processes various sentences per line\n    sentenceDetector = SentenceDetector()\\\n        .setInputCols([\"document\"])\\\n        .setOutputCol(\"sentence\")\n\n  # Tokenizer splits words in a relevant format for NLP\n    tokenizer = Tokenizer()\\\n        .setInputCols([\"sentence\"])\\\n        .setOutputCol(\"token\")\n\n  # Clinical word embeddings trained on PubMED dataset\n    word_embeddings = WordEmbeddingsModel.pretrained(embeddings, \"en\", \"clinical/models\")\\\n        .setInputCols([\"sentence\", \"token\"])\\\n        .setOutputCol(\"embeddings\")\n\n    base_pipeline = Pipeline(stages=[\n                        documentAssembler,\n                        sentenceDetector,\n                        tokenizer,\n                        word_embeddings])\n\n    return base_pipeline\n\n\n\ndef get_clinical_assertion (embeddings, spark_df, nrows = 100, model_name = 'ner_clinical'):\n\n  # NER model trained on i2b2 (sampled from MIMIC) dataset\n    loaded_ner_model = MedicalNerModel.pretrained(model_name, \"en\", \"clinical/models\") \\\n        .setInputCols([\"sentence\", \"token\", \"embeddings\"]) \\\n        .setOutputCol(\"ner\")\n\n    ner_converter = NerConverter() \\\n        .setInputCols([\"sentence\", \"token\", \"ner\"]) \\\n        .setOutputCol(\"ner_chunk\")\n\n  # Assertion model trained on i2b2 (sampled from MIMIC) dataset\n  # coming from sparknlp_jsl.annotator !!\n    clinical_assertion = AssertionDLModel.pretrained(\"assertion_dl\", \"en\", \"clinical/models\") \\\n        .setInputCols([\"sentence\", \"ner_chunk\", \"embeddings\"]) \\\n        .setOutputCol(\"assertion\")\n      \n\n    base_model = get_base_pipeline (embeddings)\n\n    nlpPipeline = Pipeline(stages=[\n        base_model,\n        loaded_ner_model,\n        ner_converter,\n        clinical_assertion])\n\n    empty_data = spark.createDataFrame([[\"\"]]).toDF(\"text\")\n\n    model = nlpPipeline.fit(empty_data)\n\n    result = model.transform(spark_df.limit(nrows))\n\n    result = result.withColumn(\"id\", monotonically_increasing_id())\n\n    result_df = result.select(F.explode(F.arrays_zip(result.ner_chunk.result, result.ner_chunk.metadata, result.assertion.result)).alias(\"cols\")) \\\n                      .select(F.expr(\"cols['0']\").alias(\"chunk\"),\n                              F.expr(\"cols['1']['entity']\").alias(\"ner_label\"),\n                              F.expr(\"cols['2']\").alias(\"assertion\"))\\\n                              .filter(\"ner_label!='O'\")\n\n    return result_df"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"ac7bc5e4-45ed-4fb6-91b8-4581865fdb5b"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["embeddings = 'embeddings_clinical'\n\nmodel_name = 'ner_clinical_large'\n\nnrows = 1000\n\nner_df = get_clinical_assertion (embeddings, pubMedDF, nrows, model_name)\n\nner_df.show()"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"5cc9a0fc-5187-4e7b-9fee-283b5fd343c9"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">ner_clinical_large download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n+--------------------+---------+---------+\n|               chunk|ner_label|assertion|\n+--------------------+---------+---------+\n|far-fieldoptical ...|  PROBLEM|  present|\n|           particles|  PROBLEM|  present|\n|           themethod|TREATMENT|  present|\n|a spherical particle|  PROBLEM| possible|\n|toMie scattering ...|TREATMENT|  present|\n|small-amplitude o...|  PROBLEM|  present|\n|the numerical method|TREATMENT|  present|\n|Numerical data fr...|  PROBLEM|  present|\n|hemispherical par...|  PROBLEM|  present|\n|ofhemispherical p...|  PROBLEM|  present|\n|       a calibration|     TEST|  present|\n|intensity-type op...|TREATMENT|  present|\n|  the esophageal ECG|     TEST|  present|\n|         tachycardia|  PROBLEM|  present|\n|            A method|TREATMENT|  present|\n|the Arzco pill el...|TREATMENT|  present|\n|       The electrode|TREATMENT|  present|\n| atelemetric monitor|     TEST|  present|\n|  the esophageal ECG|     TEST|  present|\n|transient, self-l...|  PROBLEM|  present|\n+--------------------+---------+---------+\nonly showing top 20 rows\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">ner_clinical_large download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n+--------------------+---------+---------+\n               chunk|ner_label|assertion|\n+--------------------+---------+---------+\nfar-fieldoptical ...|  PROBLEM|  present|\n           particles|  PROBLEM|  present|\n           themethod|TREATMENT|  present|\na spherical particle|  PROBLEM| possible|\ntoMie scattering ...|TREATMENT|  present|\nsmall-amplitude o...|  PROBLEM|  present|\nthe numerical method|TREATMENT|  present|\nNumerical data fr...|  PROBLEM|  present|\nhemispherical par...|  PROBLEM|  present|\nofhemispherical p...|  PROBLEM|  present|\n       a calibration|     TEST|  present|\nintensity-type op...|TREATMENT|  present|\n  the esophageal ECG|     TEST|  present|\n         tachycardia|  PROBLEM|  present|\n            A method|TREATMENT|  present|\nthe Arzco pill el...|TREATMENT|  present|\n       The electrode|TREATMENT|  present|\n atelemetric monitor|     TEST|  present|\n  the esophageal ECG|     TEST|  present|\ntransient, self-l...|  PROBLEM|  present|\n+--------------------+---------+---------+\nonly showing top 20 rows\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["embeddings = 'embeddings_clinical'\n\nmodel_name = 'ner_posology'\n\nnrows = 100\n\nner_df = get_clinical_assertion (embeddings, pubMedDF, nrows, model_name)\n\nner_df.show()"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"cb696bfc-6a34-4666-92e9-607683a57506"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">ner_posology download started this may take some time.\nApproximate size to download 13.8 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n+--------------------+---------+------------+\n|               chunk|ner_label|   assertion|\n+--------------------+---------+------------+\n|    lackedprotein VI|     DRUG|     present|\n|dehydroepiandrost...|     DRUG|     present|\n|            low-dose|   DOSAGE|      absent|\n|       dexamethasone|     DRUG|      absent|\n|withethinyloestra...|     DRUG|     present|\n| cyproterone acetate|     DRUG|     present|\n|   ethinyloestradiol|     DRUG|     present|\n| cyproterone acetate|     DRUG|     present|\n|withethinyloestra...|     DRUG|     present|\n| cyproterone acetate|     DRUG|     present|\n|                 BmK|     DRUG|     present|\n|                 BmK|     DRUG|     present|\n|ofbroad-spectrum ...|     DRUG|     present|\n|      blood-perfused|     DRUG|     present|\n|     isolatedcircuit|     DRUG|     present|\n|aspartate/glutama...|     DRUG|     present|\n|andaspartate/glut...|     DRUG|hypothetical|\n|          Superoxide|     DRUG|     present|\n|            monoxide|     DRUG|     present|\n|            modelsin|     DRUG|    possible|\n+--------------------+---------+------------+\nonly showing top 20 rows\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">ner_posology download started this may take some time.\nApproximate size to download 13.8 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n+--------------------+---------+------------+\n               chunk|ner_label|   assertion|\n+--------------------+---------+------------+\n    lackedprotein VI|     DRUG|     present|\ndehydroepiandrost...|     DRUG|     present|\n            low-dose|   DOSAGE|      absent|\n       dexamethasone|     DRUG|      absent|\nwithethinyloestra...|     DRUG|     present|\n cyproterone acetate|     DRUG|     present|\n   ethinyloestradiol|     DRUG|     present|\n cyproterone acetate|     DRUG|     present|\nwithethinyloestra...|     DRUG|     present|\n cyproterone acetate|     DRUG|     present|\n                 BmK|     DRUG|     present|\n                 BmK|     DRUG|     present|\nofbroad-spectrum ...|     DRUG|     present|\n      blood-perfused|     DRUG|     present|\n     isolatedcircuit|     DRUG|     present|\naspartate/glutama...|     DRUG|     present|\nandaspartate/glut...|     DRUG|hypothetical|\n          Superoxide|     DRUG|     present|\n            monoxide|     DRUG|     present|\n            modelsin|     DRUG|    possible|\n+--------------------+---------+------------+\nonly showing top 20 rows\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["embeddings = 'embeddings_clinical'\n\nmodel_name = 'ner_posology_greedy'\n\nentry_data = spark.createDataFrame([[\"The patient did not take a capsule of Advil\"]]).toDF(\"text\")\n\nner_df = get_clinical_assertion (embeddings, entry_data, nrows, model_name)\n\nner_df.show()"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"93fe3c92-4d0e-4e96-99d8-08e41f7aeb6b"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">ner_posology_greedy download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n+----------------+---------+---------+\n|           chunk|ner_label|assertion|\n+----------------+---------+---------+\n|capsule of Advil|     DRUG|   absent|\n+----------------+---------+---------+\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">ner_posology_greedy download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n+----------------+---------+---------+\n           chunk|ner_label|assertion|\n+----------------+---------+---------+\ncapsule of Advil|     DRUG|   absent|\n+----------------+---------+---------+\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["embeddings = 'embeddings_clinical'\n\nmodel_name = 'ner_clinical'\n\nentry_data = spark.createDataFrame([[\"The patient has no fever\"]]).toDF(\"text\")\n\nner_df = get_clinical_assertion (embeddings, entry_data, nrows, model_name)\n\nner_df.show()"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"4daa7794-be27-4259-a538-d3a36f37aa3a"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">ner_clinical download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n+-----+---------+---------+\n|chunk|ner_label|assertion|\n+-----+---------+---------+\n|fever|  PROBLEM|   absent|\n+-----+---------+---------+\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">ner_clinical download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n+-----+---------+---------+\nchunk|ner_label|assertion|\n+-----+---------+---------+\nfever|  PROBLEM|   absent|\n+-----+---------+---------+\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["import pandas as pd\n\ndef get_clinical_assertion_light (light_model, text):\n\n  light_result = light_model.fullAnnotate(text)[0]\n\n  chunks=[]\n  entities=[]\n  status=[]\n\n  for n,m in zip(light_result['ner_chunk'],light_result['assertion']):\n      \n      chunks.append(n.result)\n      entities.append(n.metadata['entity']) \n      status.append(m.result)\n          \n  df = pd.DataFrame({'chunks':chunks, 'entities':entities, 'assertion':status})\n\n  return df"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"925cdb03-42c3-43d5-ab5f-f6544fd9b5c3"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["clinical_text = \"\"\"\nPatient with severe fever and sore throat. \nHe shows no stomach pain and he maintained on an epidural and PCA for pain control.\nHe also became short of breath with climbing a flight of stairs.\nAfter CT, lung tumor located at the right lower lobe. Father with Alzheimer.\n\"\"\"\n\nlight_model = LightPipeline(model)\n\nget_clinical_assertion_light (light_model, clinical_text)\n"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"2f849a87-7851-4d0b-add0-540c0527d6ba"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">Out[23]: </div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Out[23]: </div>"]}},{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chunks</th>\n      <th>entities</th>\n      <th>assertion</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>severe fever</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>sore throat</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>stomach pain</td>\n      <td>PROBLEM</td>\n      <td>absent</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>an epidural</td>\n      <td>TREATMENT</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>PCA</td>\n      <td>TREATMENT</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>pain control</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>short of breath</td>\n      <td>PROBLEM</td>\n      <td>conditional</td>\n    </tr>\n    <tr>\n      <th>7</th>\n      <td>CT</td>\n      <td>TEST</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>8</th>\n      <td>lung tumor</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>9</th>\n      <td>Alzheimer</td>\n      <td>PROBLEM</td>\n      <td>associated_with_someone_else</td>\n    </tr>\n  </tbody>\n</table>\n</div>","textData":null,"removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"htmlSandbox","arguments":{}}},"output_type":"display_data","data":{"text/html":["<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chunks</th>\n      <th>entities</th>\n      <th>assertion</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>severe fever</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>sore throat</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>stomach pain</td>\n      <td>PROBLEM</td>\n      <td>absent</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>an epidural</td>\n      <td>TREATMENT</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>PCA</td>\n      <td>TREATMENT</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>pain control</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>short of breath</td>\n      <td>PROBLEM</td>\n      <td>conditional</td>\n    </tr>\n    <tr>\n      <th>7</th>\n      <td>CT</td>\n      <td>TEST</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>8</th>\n      <td>lung tumor</td>\n      <td>PROBLEM</td>\n      <td>present</td>\n    </tr>\n    <tr>\n      <th>9</th>\n      <td>Alzheimer</td>\n      <td>PROBLEM</td>\n      <td>associated_with_someone_else</td>\n    </tr>\n  </tbody>\n</table>\n</div>"]}}],"execution_count":0},{"cell_type":"markdown","source":["## Assertion with BioNLP (Cancer Genetics) NER"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"bdd0c509-8006-4c5c-9092-40256f47e34c"}}},{"cell_type":"code","source":["embeddings = 'embeddings_clinical'\n\nmodel_name = 'ner_bionlp'\n\nnrows = 100\n\nner_df = get_clinical_assertion (embeddings, pubMedDF, nrows, model_name)\n\nner_df.show(truncate = False)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"2ed9f618-83f3-4225-8e4c-79984a17c6ac"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">ner_bionlp download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n+----------------+----------------------+---------+\n|chunk           |ner_label             |assertion|\n+----------------+----------------------+---------+\n|optic axis      |Multi-tissue_structure|possible |\n|esophageal ECG  |Multi-tissue_structure|present  |\n|nose            |Organ                 |present  |\n|esophageal ECG  |Multi-tissue_structure|present  |\n|E5              |Gene_or_gene_product  |present  |\n|Tyr             |Amino_acid            |present  |\n|Asn             |Amino_acid            |present  |\n|Asp             |Amino_acid            |present  |\n|Escherichia coli|Organism              |present  |\n|RNases Aand T1  |Gene_or_gene_product  |present  |\n|E5-CRD          |Gene_or_gene_product  |absent   |\n|usualRNases     |Gene_or_gene_product  |absent   |\n|RNase           |Gene_or_gene_product  |present  |\n|Lys25           |Simple_chemical       |present  |\n|Gln29           |Simple_chemical       |present  |\n|Lys60           |Simple_chemical       |present  |\n|Ile94           |Gene_or_gene_product  |present  |\n|Arg33           |Gene_or_gene_product  |possible |\n|Arg33           |Gene_or_gene_product  |possible |\n|Lys60           |Simple_chemical       |present  |\n+----------------+----------------------+---------+\nonly showing top 20 rows\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">ner_bionlp download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\nembeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n+----------------+----------------------+---------+\nchunk           |ner_label             |assertion|\n+----------------+----------------------+---------+\noptic axis      |Multi-tissue_structure|possible |\nesophageal ECG  |Multi-tissue_structure|present  |\nnose            |Organ                 |present  |\nesophageal ECG  |Multi-tissue_structure|present  |\nE5              |Gene_or_gene_product  |present  |\nTyr             |Amino_acid            |present  |\nAsn             |Amino_acid            |present  |\nAsp             |Amino_acid            |present  |\nEscherichia coli|Organism              |present  |\nRNases Aand T1  |Gene_or_gene_product  |present  |\nE5-CRD          |Gene_or_gene_product  |absent   |\nusualRNases     |Gene_or_gene_product  |absent   |\nRNase           |Gene_or_gene_product  |present  |\nLys25           |Simple_chemical       |present  |\nGln29           |Simple_chemical       |present  |\nLys60           |Simple_chemical       |present  |\nIle94           |Gene_or_gene_product  |present  |\nArg33           |Gene_or_gene_product  |possible |\nArg33           |Gene_or_gene_product  |possible |\nLys60           |Simple_chemical       |present  |\n+----------------+----------------------+---------+\nonly showing top 20 rows\n\n</div>"]}}],"execution_count":0},{"cell_type":"markdown","source":["# Assertion Filterer\nAssertionFilterer will allow you to filter out the named entities by the list of acceptable assertion statuses. This annotator would be quite handy if you want to set a white list for the acceptable assertion statuses like present or conditional; and do not want absent conditions get out of your pipeline."],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"5b6f9442-9a54-49cc-ad1d-186fc5ea326e"}}},{"cell_type":"code","source":["clinical_ner = MedicalNerModel.pretrained(\"ner_clinical\", \"en\", \"clinical/models\") \\\n      .setInputCols([\"sentence\", \"token\", \"embeddings\"]) \\\n      .setOutputCol(\"ner\")\n\nner_converter = NerConverter() \\\n      .setInputCols([\"sentence\", \"token\", \"ner\"]) \\\n      .setOutputCol(\"ner_chunk\")\n\nclinical_assertion = AssertionDLModel.pretrained(\"assertion_dl\", \"en\", \"clinical/models\") \\\n      .setInputCols([\"sentence\", \"ner_chunk\", \"embeddings\"]) \\\n      .setOutputCol(\"assertion\")\n\nassertion_filterer = AssertionFilterer()\\\n      .setInputCols(\"sentence\",\"ner_chunk\",\"assertion\")\\\n      .setOutputCol(\"assertion_filtered\")\\\n      .setWhiteList([\"present\"])\n\nnlpPipeline = Pipeline(stages=[\n      documentAssembler, \n      sentenceDetector,\n      tokenizer,\n      word_embeddings,\n      clinical_ner,\n      ner_converter,\n      clinical_assertion,\n      assertion_filterer\n    ])\n\nempty_data = spark.createDataFrame([[\"\"]]).toDF(\"text\")\nassertionFilter_model = nlpPipeline.fit(empty_data)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"f50d1b8e-edff-4d4c-839f-d0ba3d0b8923"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">ner_clinical download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">ner_clinical download started this may take some time.\nApproximate size to download 13.9 MB\n\r[ | ]\r[OK!]\nassertion_dl download started this may take some time.\nApproximate size to download 1.3 MB\n\r[ | ]\r[OK!]\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["text = 'Patient has a headache for the last 2 weeks, needs to get a head CT, and appears anxious when she walks fast. Alopecia noted. She denies pain.'\n\nlight_model = LightPipeline(assertionFilter_model)\nlight_result = light_model.annotate(text)\n\nlight_result.keys()"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"f3800714-30a3-4515-9932-5a8c0c2021a6"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">Out[26]: dict_keys([&#39;assertion_filtered&#39;, &#39;document&#39;, &#39;ner_chunk&#39;, &#39;assertion&#39;, &#39;token&#39;, &#39;ner&#39;, &#39;embeddings&#39;, &#39;sentence&#39;])</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Out[26]: dict_keys([&#39;assertion_filtered&#39;, &#39;document&#39;, &#39;ner_chunk&#39;, &#39;assertion&#39;, &#39;token&#39;, &#39;ner&#39;, &#39;embeddings&#39;, &#39;sentence&#39;])</div>"]}}],"execution_count":0},{"cell_type":"code","source":["list(zip(light_result['ner_chunk'], light_result['assertion']))"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"b8986c91-53c4-4a08-a36b-1fea93c1c762"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">Out[27]: [(&#39;a headache&#39;, &#39;present&#39;),\n (&#39;a head CT&#39;, &#39;present&#39;),\n (&#39;anxious&#39;, &#39;present&#39;),\n (&#39;fast&#39;, &#39;present&#39;),\n (&#39;Alopecia&#39;, &#39;present&#39;),\n (&#39;pain&#39;, &#39;absent&#39;)]</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Out[27]: [(&#39;a headache&#39;, &#39;present&#39;),\n (&#39;a head CT&#39;, &#39;present&#39;),\n (&#39;anxious&#39;, &#39;present&#39;),\n (&#39;fast&#39;, &#39;present&#39;),\n (&#39;Alopecia&#39;, &#39;present&#39;),\n (&#39;pain&#39;, &#39;absent&#39;)]</div>"]}}],"execution_count":0},{"cell_type":"code","source":["assertion_filterer.getWhiteList()"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"44113efd-939b-4fd1-95f2-110e957a6635"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">Out[28]: [&#39;present&#39;]</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Out[28]: [&#39;present&#39;]</div>"]}}],"execution_count":0},{"cell_type":"code","source":["light_result['assertion_filtered']"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"aa81ca8c-7573-42cb-a895-3981c4ff1bdc"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">Out[29]: [&#39;a headache&#39;, &#39;a head CT&#39;, &#39;anxious&#39;, &#39;fast&#39;, &#39;Alopecia&#39;]</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Out[29]: [&#39;a headache&#39;, &#39;a head CT&#39;, &#39;anxious&#39;, &#39;fast&#39;, &#39;Alopecia&#39;]</div>"]}}],"execution_count":0},{"cell_type":"markdown","source":["# Train a custom Assertion Model"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"647435eb-a456-488f-ade2-f97e4c8de429"}}},{"cell_type":"markdown","source":["**WARNING:** For training an Assertion model, please use TensorFlow version 2.3"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"9aa4951b-abd4-44a7-bc32-51719afb01a6"}}},{"cell_type":"code","source":["!wget -q https://raw.githubusercontent.com/JohnSnowLabs/spark-nlp-workshop/master/tutorials/Certification_Trainings/Healthcare/data/i2b2_assertion_sample_short.csv -P /dbfs/"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"2d72802c-1c83-4572-a52e-c7e1074e69cb"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["assertion_df = spark.read.option(\"header\", True).option(\"inferSchema\", \"True\").csv(\"/i2b2_assertion_sample_short.csv\")\n"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"711aa642-d5ab-4f57-8063-4451b6ee0d11"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["assertion_df.show(3, truncate=100)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"d9f89d7b-ce06-43dc-bdff-e53cbb6f8ada"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">+-------------------------------------------------+-------------------+-------+-----+---+\n|                                             text|             target|  label|start|end|\n+-------------------------------------------------+-------------------+-------+-----+---+\n|She has no history of liver disease , hepatitis .|      liver disease| absent|    5|  6|\n|                         1. Undesired fertility .|undesired fertility|present|    1|  2|\n|                            3) STATUS POST FALL .|               fall|present|    3|  3|\n+-------------------------------------------------+-------------------+-------+-----+---+\nonly showing top 3 rows\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">+-------------------------------------------------+-------------------+-------+-----+---+\n                                             text|             target|  label|start|end|\n+-------------------------------------------------+-------------------+-------+-----+---+\nShe has no history of liver disease , hepatitis .|      liver disease| absent|    5|  6|\n                         1. Undesired fertility .|undesired fertility|present|    1|  2|\n                            3) STATUS POST FALL .|               fall|present|    3|  3|\n+-------------------------------------------------+-------------------+-------+-----+---+\nonly showing top 3 rows\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["(training_data, test_data) = assertion_df.randomSplit([0.8, 0.2], seed = 100)\nprint(\"Training Dataset Count: \" + str(training_data.count()))\nprint(\"Test Dataset Count: \" + str(test_data.count()))"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"eba6ba43-dbd4-4403-b967-8ecde9e0461a"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">Training Dataset Count: 721\nTest Dataset Count: 170\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Training Dataset Count: 721\nTest Dataset Count: 170\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["training_data.groupBy('label').count().orderBy('count', ascending=False).show(truncate=False)\n"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"208b749b-13fa-4780-93a1-d7cd316e81f6"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">+-------+-----+\n|label  |count|\n+-------+-----+\n|present|546  |\n|absent |175  |\n+-------+-----+\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">+-------+-----+\nlabel  |count|\n+-------+-----+\npresent|546  |\nabsent |175  |\n+-------+-----+\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["document = DocumentAssembler()\\\n    .setInputCol(\"text\")\\\n    .setOutputCol(\"document\")\n\nchunk = Doc2Chunk()\\\n    .setInputCols(\"document\")\\\n    .setOutputCol(\"chunk\")\\\n    .setChunkCol(\"target\")\\\n    .setStartCol(\"start\")\\\n    .setStartColByTokenIndex(True)\\\n    .setFailOnMissing(False)\\\n    .setLowerCase(True)\n\ntoken = Tokenizer()\\\n    .setInputCols(['document'])\\\n    .setOutputCol('token')\n\nembeddings = WordEmbeddingsModel.pretrained(\"embeddings_clinical\", \"en\", \"clinical/models\")\\\n      .setInputCols([\"document\", \"token\"])\\\n      .setOutputCol(\"embeddings\")\n"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"1dd09bfb-5a71-49b0-8e5a-ce40820d154a"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">embeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">embeddings_clinical download started this may take some time.\nApproximate size to download 1.6 GB\n\r[ | ]\r[OK!]\n</div>"]}}],"execution_count":0},{"cell_type":"markdown","source":["#### With a custom tf graph"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"90a7fafa-35e2-4437-a473-7f7c482ba351"}}},{"cell_type":"code","source":["import tensorflow\nfrom sparknlp_jsl.training import tf_graph"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"11bead5c-9e4b-4e5b-b1cf-164e59decbfb"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["%fs mkdirs file:/dbfs/assertion_tf_graphs"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"26545bcf-6cf6-4d85-a521-b70df503c635"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">res0: Boolean = true\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{"isDbfsCommandResult":false},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">res0: Boolean = true\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["%fs mkdirs file:/dbfs/assertion_logs"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"febac50c-d17f-466f-85ee-87b24b54449d"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">res1: Boolean = true\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{"isDbfsCommandResult":false},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">res1: Boolean = true\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["# Create custom graph\n\n\ntf_graph.print_model_params(\"assertion_dl\")\n\nfeat_size = 200\nn_classes = 2\n\ntf_graph.build(\"assertion_dl\",\n               build_params={\"n_classes\": n_classes},\n               model_location= \"/dbfs/assertion_tf_graphs\", \n               model_filename=\"blstm_34_32_30_{}_{}.pb\".format(feat_size, n_classes))"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"3426663d-2b24-49ea-9b1e-3e154b0e260a"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">assertion_dl parameters.\nParameter            Required   Default value        Description\nmax_seq_len          no         250                  Maximum sequence length.\nfeat_size            no         200                  Feature size.\nn_classes            yes        -                    Number of classes.\ndevice               no         /cpu:0               Device for training.\nn_hidden             no         34                   Number of hidden units.\n/databricks/python/lib/python3.8/site-packages/tensorflow_addons/utils/ensure_tf_install.py:53: UserWarning: Tensorflow Addons supports using Python ops for all Tensorflow versions above or equal to 2.6.0 and strictly below 2.9.0 (nightly versions are not supported). \n The versions of TensorFlow you are currently using is 2.3.0 and is not supported. \nSome things might work, some things might not.\nIf you were to encounter a bug, do not file an issue.\nIf you want to make sure you&#39;re using a tested and supported configuration, either change the TensorFlow version or the TensorFlow Addons&#39;s version. \nYou can find the compatibility matrix in TensorFlow Addon&#39;s readme:\nhttps://github.com/tensorflow/addons\n  warnings.warn(\nWARNING:tensorflow:From /databricks/python/lib/python3.8/site-packages/tensorflow/python/compat/v2_compat.py:96: disable_resource_variables (from tensorflow.python.ops.variable_scope) is deprecated and will be removed in a future version.\nInstructions for updating:\nnon-resource variables are not supported in the long term\nDevice mapping:\n/job:localhost/replica:0/task:0/device:XLA_CPU:0 -&gt; device: XLA_CPU device\n\nWARNING:tensorflow:From /databricks/python/lib/python3.8/site-packages/sparknlp_jsl/_tf_graph_builders/assertion_dl/assertion_model.py:49: BasicLSTMCell.__init__ (from tensorflow.python.keras.layers.legacy_rnn.rnn_cell_impl) is deprecated and will be removed in a future version.\nInstructions for updating:\nThis class is equivalent as tf.keras.layers.LSTMCell, and will be replaced by that in Tensorflow 2.0.\nWARNING:tensorflow:From /databricks/python/lib/python3.8/site-packages/sparknlp_jsl/_tf_graph_builders/tf2contrib/rnn.py:229: bidirectional_dynamic_rnn (from tensorflow.python.ops.rnn) is deprecated and will be removed in a future version.\nInstructions for updating:\nPlease use `keras.layers.Bidirectional(keras.layers.RNN(cell))`, which is equivalent to this API\nWARNING:tensorflow:From /databricks/python/lib/python3.8/site-packages/tensorflow/python/ops/rnn.py:438: dynamic_rnn (from tensorflow.python.ops.rnn) is deprecated and will be removed in a future version.\nInstructions for updating:\nPlease use `keras.layers.RNN(cell)`, which is equivalent to this API\nWARNING:tensorflow:From /databricks/python/lib/python3.8/site-packages/tensorflow/python/keras/layers/legacy_rnn/rnn_cell_impl.py:738: Layer.add_variable (from tensorflow.python.keras.engine.base_layer_v1) is deprecated and will be removed in a future version.\nInstructions for updating:\nPlease use `layer.add_weight` method instead.\nWARNING:tensorflow:From /databricks/python/lib/python3.8/site-packages/tensorflow/python/keras/layers/legacy_rnn/rnn_cell_impl.py:744: calling Zeros.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.\nInstructions for updating:\nCall initializer instance with the dtype argument instead of passing it to the constructor\nDevice mapping:\n/job:localhost/replica:0/task:0/device:XLA_CPU:0 -&gt; device: XLA_CPU device\n\nassertion_dl graph exported to /dbfs/assertion_tf_graphs/blstm_34_32_30_200_2.pb\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">assertion_dl parameters.\nParameter            Required   Default value        Description\nmax_seq_len          no         250                  Maximum sequence length.\nfeat_size            no         200                  Feature size.\nn_classes            yes        -                    Number of classes.\ndevice               no         /cpu:0               Device for training.\nn_hidden             no         34                   Number of hidden units.\n/databricks/python/lib/python3.8/site-packages/tensorflow_addons/utils/ensure_tf_install.py:53: UserWarning: Tensorflow Addons supports using Python ops for all Tensorflow versions above or equal to 2.6.0 and strictly below 2.9.0 (nightly versions are not supported). \n The versions of TensorFlow you are currently using is 2.3.0 and is not supported. \nSome things might work, some things might not.\nIf you were to encounter a bug, do not file an issue.\nIf you want to make sure you&#39;re using a tested and supported configuration, either change the TensorFlow version or the TensorFlow Addons&#39;s version. \nYou can find the compatibility matrix in TensorFlow Addon&#39;s readme:\nhttps://github.com/tensorflow/addons\n  warnings.warn(\nWARNING:tensorflow:From /databricks/python/lib/python3.8/site-packages/tensorflow/python/compat/v2_compat.py:96: disable_resource_variables (from tensorflow.python.ops.variable_scope) is deprecated and will be removed in a future version.\nInstructions for updating:\nnon-resource variables are not supported in the long term\nDevice mapping:\n/job:localhost/replica:0/task:0/device:XLA_CPU:0 -&gt; device: XLA_CPU device\n\nWARNING:tensorflow:From /databricks/python/lib/python3.8/site-packages/sparknlp_jsl/_tf_graph_builders/assertion_dl/assertion_model.py:49: BasicLSTMCell.__init__ (from tensorflow.python.keras.layers.legacy_rnn.rnn_cell_impl) is deprecated and will be removed in a future version.\nInstructions for updating:\nThis class is equivalent as tf.keras.layers.LSTMCell, and will be replaced by that in Tensorflow 2.0.\nWARNING:tensorflow:From /databricks/python/lib/python3.8/site-packages/sparknlp_jsl/_tf_graph_builders/tf2contrib/rnn.py:229: bidirectional_dynamic_rnn (from tensorflow.python.ops.rnn) is deprecated and will be removed in a future version.\nInstructions for updating:\nPlease use `keras.layers.Bidirectional(keras.layers.RNN(cell))`, which is equivalent to this API\nWARNING:tensorflow:From /databricks/python/lib/python3.8/site-packages/tensorflow/python/ops/rnn.py:438: dynamic_rnn (from tensorflow.python.ops.rnn) is deprecated and will be removed in a future version.\nInstructions for updating:\nPlease use `keras.layers.RNN(cell)`, which is equivalent to this API\nWARNING:tensorflow:From /databricks/python/lib/python3.8/site-packages/tensorflow/python/keras/layers/legacy_rnn/rnn_cell_impl.py:738: Layer.add_variable (from tensorflow.python.keras.engine.base_layer_v1) is deprecated and will be removed in a future version.\nInstructions for updating:\nPlease use `layer.add_weight` method instead.\nWARNING:tensorflow:From /databricks/python/lib/python3.8/site-packages/tensorflow/python/keras/layers/legacy_rnn/rnn_cell_impl.py:744: calling Zeros.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.\nInstructions for updating:\nCall initializer instance with the dtype argument instead of passing it to the constructor\nDevice mapping:\n/job:localhost/replica:0/task:0/device:XLA_CPU:0 -&gt; device: XLA_CPU device\n\nassertion_dl graph exported to /dbfs/assertion_tf_graphs/blstm_34_32_30_200_2.pb\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["# if you want you can use existing graph\n\n# !wget -q https://raw.githubusercontent.com/JohnSnowLabs/spark-nlp-workshop/master/tutorials/Certification_Trainings/Healthcare/tf_graphs/blstm_34_32_30_200_2.pb -P /dbfs/assertion_tf_graphs"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"ca963d9f-ddf4-4de7-b424-42b25747e110"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"markdown","source":["**Setting the Scope Window (Target Area) Dynamically in Assertion Status Detection Models**\n\n\nThis parameter allows you to train the Assertion Status Models to focus on specific context windows when resolving the status of a NER chunk. The window is in format `[X,Y]` being `X` the number of tokens to consider on the left of the chunk, and `Y` the max number of tokens to consider on the right. Lets take a look at what different windows mean:\n\n\n*   By default, the window is `[-1,-1]` which means that the Assertion Status will look at all of the tokens in the sentence/document (up to a maximum of tokens set in `setMaxSentLen()` ).\n*   `[0,0]` means dont pay attention to any token except the ner_chunk, what basically is not considering any context for the Assertion resolution.\n*   `[9,15]` is what empirically seems to be the best baseline, meaning that we look up to 9 tokens on the left and 15 on the right of the ner chunk to understand the context and resolve the status.\n\n\nCheck this [Scope Window Tuning Assertion Status Detection notebook](https://github.com/JohnSnowLabs/spark-nlp-workshop/blob/master/tutorials/Certification_Trainings/Healthcare/2.1.Scope_window_tuning_assertion_status_detection.ipynb)  that illustrates the effect of the different windows and how to properly fine-tune your AssertionDLModels to get the best of them.\n\nIn our case, the best Scope Window is around [10,10]"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"c86c03d7-8f0e-440c-a7bc-32c922470b6a"}}},{"cell_type":"code","source":["scope_window = [10,10]\n\nassertionStatus = AssertionDLApproach()\\\n    .setLabelCol(\"label\")\\\n    .setInputCols(\"document\", \"chunk\", \"embeddings\")\\\n    .setOutputCol(\"assertion\")\\\n    .setBatchSize(128)\\\n    .setDropout(0.1)\\\n    .setLearningRate(0.001)\\\n    .setEpochs(50)\\\n    .setValidationSplit(0.2)\\\n    .setStartCol(\"start\")\\\n    .setEndCol(\"end\")\\\n    .setMaxSentLen(250)\\\n    .setEnableOutputLogs(True)\\\n    .setOutputLogsPath('dbfs:/assertion_logs')\\\n    .setGraphFolder('dbfs:/assertion_tf_graphs')\\\n    .setScopeWindow(scope_window)\n\n\n'''\nIf .setTestDataset parameter is employed, raw test data cannot be fitted. .setTestDataset only works for dataframes which are correctly transformed\nby a pipeline consisting of document, chunk, embeddings stages.\n'''"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"19b3315c-86e7-4b86-9d0a-18dfd1b1ccc5"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">Out[39]: &#39;\\nIf .setTestDataset parameter is employed, raw test data cannot be fitted. .setTestDataset only works for dataframes which are correctly transformed\\nby a pipeline consisting of document, chunk, embeddings stages.\\n&#39;</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Out[39]: &#39;\\nIf .setTestDataset parameter is employed, raw test data cannot be fitted. .setTestDataset only works for dataframes which are correctly transformed\\nby a pipeline consisting of document, chunk, embeddings stages.\\n&#39;</div>"]}}],"execution_count":0},{"cell_type":"code","source":["clinical_assertion_pipeline = Pipeline(\n    stages = [\n    document,\n    chunk,\n    token,\n    embeddings,\n    assertionStatus])"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"315c9be0-9cd7-4bf4-b4c0-d21165acf919"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["assertion_model = clinical_assertion_pipeline.fit(training_data)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"9327a04d-1037-4261-b766-951d9a902ee0"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["assertion_model.stages"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"88d0ad2e-b6ef-4141-9387-1bd7692eb125"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">Out[42]: [DocumentAssembler_13477929893f,\n Doc2Chunk_7a67300c0bcb,\n REGEX_TOKENIZER_076956dda637,\n WORD_EMBEDDINGS_MODEL_9004b1d00302,\n ASSERTION_DL_6a55acf140c8]</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Out[42]: [DocumentAssembler_13477929893f,\n Doc2Chunk_7a67300c0bcb,\n REGEX_TOKENIZER_076956dda637,\n WORD_EMBEDDINGS_MODEL_9004b1d00302,\n ASSERTION_DL_6a55acf140c8]</div>"]}}],"execution_count":0},{"cell_type":"code","source":["preds = assertion_model.transform(test_data).select('label','assertion.result')\n\npreds.show()"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"0bda53e4-8351-4171-b994-8e9f93b2e93c"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">+-------+---------+\n|  label|   result|\n+-------+---------+\n|present|[present]|\n| absent|[present]|\n|present|[present]|\n|present|[present]|\n|present|[present]|\n|present|[present]|\n|present|[present]|\n|present|[present]|\n|present|[present]|\n|present|[present]|\n|present|[present]|\n|present|[present]|\n|present|[present]|\n|present|[present]|\n|present|[present]|\n|present|[present]|\n|present|[present]|\n|present|[present]|\n|present|[present]|\n|present|[present]|\n+-------+---------+\nonly showing top 20 rows\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">+-------+---------+\n  label|   result|\n+-------+---------+\npresent|[present]|\n absent|[present]|\npresent|[present]|\npresent|[present]|\npresent|[present]|\npresent|[present]|\npresent|[present]|\npresent|[present]|\npresent|[present]|\npresent|[present]|\npresent|[present]|\npresent|[present]|\npresent|[present]|\npresent|[present]|\npresent|[present]|\npresent|[present]|\npresent|[present]|\npresent|[present]|\npresent|[present]|\npresent|[present]|\n+-------+---------+\nonly showing top 20 rows\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["preds_df = preds.toPandas()"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"43bcfc6a-4c3a-40b8-8b26-5da14ce8682a"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["preds_df['result'] = preds_df['result'].apply(lambda x : x[0])\n"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"69926665-71df-4ee2-ab7e-bc70099feeba"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0},{"cell_type":"code","source":["# We are going to use sklearn to evalute the results on test dataset\nfrom sklearn.metrics import classification_report\n\nprint (classification_report(preds_df['result'], preds_df['label']))"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"d5d83d2b-95a4-4a14-a356-f85201bc641d"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\">              precision    recall  f1-score   support\n\n      absent       0.68      0.90      0.77        40\n     present       0.97      0.87      0.91       130\n\n    accuracy                           0.88       170\n   macro avg       0.82      0.88      0.84       170\nweighted avg       0.90      0.88      0.88       170\n\n</div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">              precision    recall  f1-score   support\n\n      absent       0.68      0.90      0.77        40\n     present       0.97      0.87      0.91       130\n\n    accuracy                           0.88       170\n   macro avg       0.82      0.88      0.84       170\nweighted avg       0.90      0.88      0.88       170\n\n</div>"]}}],"execution_count":0},{"cell_type":"code","source":["#saving the model that we've trained\nassertion_model.stages[-1].write().overwrite().save('/databricks/driver/models/custom_assertion_model')"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"ad145896-0369-40b2-8c05-85351675e7a0"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"datasetInfos":[],"data":"<div class=\"ansiout\"></div>","removedWidgets":[],"addedWidgets":{},"metadata":{},"type":"html","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":0}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"mimetype":"text/x-python","name":"python","pygments_lexer":"ipython3","codemirror_mode":{"name":"ipython","version":3},"version":"3.8.3","nbconvert_exporter":"python","file_extension":".py"},"application/vnd.databricks.v1+notebook":{"notebookName":"5. Training and Reusing Assertion Status Models","dashboards":[],"notebookMetadata":{"pythonIndentUnit":2},"language":"python","widgets":{},"notebookOrigID":4491977052999199}},"nbformat":4,"nbformat_minor":0}
