{"cells":[{"cell_type":"markdown","metadata":{"id":"K370gc-9YFR7"},"source":["![JohnSnowLabs](https://nlp.johnsnowlabs.com/assets/images/logo.png)\n"]},{"cell_type":"markdown","metadata":{"id":"Typ-TLc1YHgM"},"source":["[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/JohnSnowLabs/spark-nlp-workshop/blob/master/tutorials/Certification_Trainings/Healthcare/34.0.Model_Download_Helpers.ipynb)\n"]},{"cell_type":"markdown","metadata":{"id":"okhT7AcXxben"},"source":["## Colab Setup"]},{"cell_type":"markdown","metadata":{"id":"G7dOaR_TlgE-"},"source":["üìåTo run this yourself, you will need to upload your license keys to the notebook. Just Run The Cell Below in order to do that. Also You can open the file explorer on the left side of the screen and upload `license_keys.json` to the folder that opens.\n","Otherwise, you can look at the example outputs at the bottom of the notebook."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"NAC_DkSrhiYA"},"outputs":[],"source":["# Install the johnsnowlabs library to access Spark-OCR and Spark-NLP for Healthcare, Finance, and Legal.\n","! pip install -q johnsnowlabs==5.1.0"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"28bAyRW0hpQO"},"outputs":[],"source":["from google.colab import files\n","print('Please Upload your John Snow Labs License using the button below')\n","license_keys = files.upload()"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"K9HN9OBrhpWN"},"outputs":[],"source":["from johnsnowlabs import nlp, medical\n","\n","# After uploading your license run this to install all licensed Python Wheels and pre-download Jars the Spark Session JVM\n","nlp.install()"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"ixZPpowjhpa9"},"outputs":[],"source":["from johnsnowlabs import nlp, medical\n","\n","# Automatically load license data and start a session with all jars user has access to\n","spark = nlp.start()"]},{"cell_type":"code","source":["spark"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":219},"id":"54SQjAqeKB0C","executionInfo":{"status":"ok","timestamp":1695879181099,"user_tz":240,"elapsed":149,"user":{"displayName":"Sel√ßuk Meng√ºverdi","userId":"15361124602135819062"}},"outputId":"9f591d7a-ffbb-4233-9d7b-28e23f65038c"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["<pyspark.sql.session.SparkSession at 0x791a9d5c57e0>"],"text/html":["\n","            <div>\n","                <p><b>SparkSession - in-memory</b></p>\n","                \n","        <div>\n","            <p><b>SparkContext</b></p>\n","\n","            <p><a href=\"http://b85a2f63c4a3:4040\">Spark UI</a></p>\n","\n","            <dl>\n","              <dt>Version</dt>\n","                <dd><code>v3.1.2</code></dd>\n","              <dt>Master</dt>\n","                <dd><code>local[*]</code></dd>\n","              <dt>AppName</dt>\n","                <dd><code>John-Snow-Labs-Spark-Session üöÄ with Jars for: üöÄSpark-NLP==5.1.0, üíäSpark-Healthcare==5.1.0, running on ‚ö° PySpark==3.1.2</code></dd>\n","            </dl>\n","        </div>\n","        \n","            </div>\n","        "]},"metadata":{},"execution_count":5}]},{"cell_type":"code","execution_count":null,"metadata":{"id":"08XeTlbnGrNw"},"outputs":[],"source":["import glob\n","import string\n","import numpy as np"]},{"cell_type":"markdown","metadata":{"id":"k3Mnf126YOKK"},"source":["# ResourceDownloader"]},{"cell_type":"markdown","metadata":{"id":"JcPtpx0Dcn6x"},"source":["This notebook will cover the different parameters and usages of `ResourceDownloader` annotator.\n","\n","**üìñ Learning Objectives:**\n","\n","1. Understand how to use `ResourceDownloader`.\n","\n","2. Become comfortable using the different parameters of the annotator.\n","\n","\n","\n","\n","**üîó Helpful Links:**\n","\n","\n","- Python Docs : [ResourceDownloader](https://sparknlp.org/api/python/reference/autosummary/sparknlp/pretrained/resource_downloader/index.html#sparknlp.pretrained.resource_downloader.ResourceDownloader)\n","\n","- Scala Docs : [ResourceDownloader](https://sparknlp.org/api/com/johnsnowlabs/nlp/pretrained/ResourceDownloader.html)\n","\n","- For extended examples of usage, see the [Spark NLP Workshop repository]()."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"wviX423GJ7O7"},"outputs":[],"source":["from sparknlp.pretrained import ResourceDownloader"]},{"cell_type":"markdown","metadata":{"id":"pkQEnLGnp2pO"},"source":["## showPublicModels"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":5785,"status":"ok","timestamp":1695879290209,"user":{"displayName":"Sel√ßuk Meng√ºverdi","userId":"15361124602135819062"},"user_tz":240},"id":"H2wbsoeEnu6C","outputId":"d72bd39f-b925-4ba9-f7a1-c795cff1dab8"},"outputs":[{"output_type":"stream","name":"stdout","text":["+------------------------+------+---------+\n","| Model                  | lang | version |\n","+------------------------+------+---------+\n","| token_rules            |  en  | 2.1.0   |\n","| onto_100               |  en  | 2.1.0   |\n","| onto_300               |  en  | 2.1.0   |\n","| bert_base_cased        |  en  | 2.2.0   |\n","| bert_uncased           |  en  | 2.2.0   |\n","| bert_base_uncased      |  en  | 2.2.0   |\n","| bert_large_cased       |  en  | 2.2.0   |\n","| bert_large_uncased     |  en  | 2.2.0   |\n","| ner_dl_bert            |  en  | 2.2.0   |\n","| pos_ud_ewt             |  en  | 2.2.2   |\n","| glove_100d             |  en  | 2.4.0   |\n","| onto_100               |  en  | 2.4.0   |\n","| onto_300               |  en  | 2.4.0   |\n","| ner_dl                 |  en  | 2.4.0   |\n","| ner_dl_sentence        |  en  | 2.4.0   |\n","| elmo                   |  en  | 2.4.0   |\n","| bert_base_cased        |  en  | 2.4.0   |\n","| bert_base_uncased      |  en  | 2.4.0   |\n","| bert_large_cased       |  en  | 2.4.0   |\n","| bert_large_uncased     |  en  | 2.4.0   |\n","| ner_dl_bert            |  en  | 2.4.0   |\n","| ner_dl_bert_base_cased |  en  | 2.4.0   |\n","| tfhub_use              |  en  | 2.4.0   |\n","| tfhub_use_lg           |  en  | 2.4.0   |\n","+------------------------+------+---------+\n","\n"]}],"source":["ResourceDownloader.showPublicModels(lang=\"en\", version=\"2.4.0\")"]},{"cell_type":"markdown","metadata":{"id":"Bu-RpEd0l7UI"},"source":["## downloadModel\n","\n","    Downloads and loads a model with the default downloader. Usually this method does not need to be called directly, as it is called by the `pretrained()`\n","    method of the annotator. Return Loaded pretrained annotator/pipeline\n","\n","\n","    Parameters:\n","      reader    : Name of the class to read the model for\n","      name      : Name of the pretrained model\n","      language  : Language of the model\n","      remote_loc: Directory of the Spark NLP Folder, by default None\n"]},{"cell_type":"markdown","metadata":{"id":"cUXditGFCx4W"},"source":["[**glove_100d**](https://sparknlp.org/2020/01/22/glove_100d.html)\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":11030,"status":"ok","timestamp":1695879301221,"user":{"displayName":"Sel√ßuk Meng√ºverdi","userId":"15361124602135819062"},"user_tz":240},"id":"LGyVnr10CW76","outputId":"d89b66b9-5f31-4c87-87e1-d18448f18b76"},"outputs":[{"output_type":"stream","name":"stdout","text":["glove_100d download started this may take some time.\n","Approximate size to download 145.3 MB\n","[OK!]\n"]},{"output_type":"execute_result","data":{"text/plain":["WORD_EMBEDDINGS_MODEL_48cffc8b9a76"]},"metadata":{},"execution_count":9}],"source":["# You can download the model directly.\n","# parameters are class name, model name, lang, and remote location\n","ResourceDownloader.downloadModel(nlp.WordEmbeddingsModel, \"glove_100d\", \"en\",remote_loc=\"public/models\")"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":3809,"status":"ok","timestamp":1695879305027,"user":{"displayName":"Sel√ßuk Meng√ºverdi","userId":"15361124602135819062"},"user_tz":240},"id":"43JOQ6q-E1j5","outputId":"920a896e-2a4f-4723-fcda-fb4b711f59ab"},"outputs":[{"output_type":"stream","name":"stdout","text":["glove_100d download started this may take some time.\n","Approximate size to download 145.3 MB\n","[OK!]\n"]}],"source":["# You can directly download the model and assign it to the corresponding variable\n","\n","embeddings = ResourceDownloader.downloadModel(nlp.WordEmbeddingsModel, \"glove_100d\", \"en\",remote_loc=\"public/models\")"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":19,"status":"ok","timestamp":1695879305028,"user":{"displayName":"Sel√ßuk Meng√ºverdi","userId":"15361124602135819062"},"user_tz":240},"id":"smJ3G5RzEukO","outputId":"06529b5c-a93a-46dd-f349-c9fd376c56d7"},"outputs":[{"output_type":"execute_result","data":{"text/plain":["['/root/cache_pretrained/glove_100d_en_2.4.0_2.4_1579690104032']"]},"metadata":{},"execution_count":11}],"source":["# check the cache folder path\n","\n","path=glob.glob(\"/root/cache_pretrained/glove_100d*\")\n","path"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"ewd87W4lCWzU"},"outputs":[],"source":["# you can load the downloaded model with the corresponding annotator\n","\n","embeddings = nlp.WordEmbeddingsModel.load(\"/root/cache_pretrained/glove_100d_en_2.4.0_2.4_1579690104032\") \\\n","  .setInputCols(\"sentence\", \"token\") \\\n","  .setOutputCol(\"embeddings\")"]},{"cell_type":"markdown","metadata":{"id":"589SVWyIC1Mq"},"source":["[**nerdl_restaurant_100d**](https://sparknlp.org/2021/12/31/nerdl_restaurant_100d_en.html)\n"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":11139,"status":"ok","timestamp":1695879316315,"user":{"displayName":"Sel√ßuk Meng√ºverdi","userId":"15361124602135819062"},"user_tz":240},"id":"5Iv41MSllB3s","outputId":"8ce820b8-d37d-4b53-ddea-c923e8851560"},"outputs":[{"output_type":"stream","name":"stdout","text":["nerdl_restaurant_100d download started this may take some time.\n","Approximate size to download 13.6 MB\n","[OK!]\n"]},{"output_type":"execute_result","data":{"text/plain":["NerDLModel_ffcdc63f298c"]},"metadata":{},"execution_count":13}],"source":["ResourceDownloader.downloadModel(nlp.NerDLModel, \"nerdl_restaurant_100d\", \"en\",remote_loc=\"public/models\")"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":10,"status":"ok","timestamp":1695879316316,"user":{"displayName":"Sel√ßuk Meng√ºverdi","userId":"15361124602135819062"},"user_tz":240},"id":"pRoEPgc-nVdy","outputId":"a433dd03-d802-46b8-def8-08ac590e9cf6"},"outputs":[{"output_type":"execute_result","data":{"text/plain":["['/root/cache_pretrained/nerdl_restaurant_100d_en_3.3.4_3.0_1640949258750']"]},"metadata":{},"execution_count":14}],"source":["path=glob.glob(\"/root/cache_pretrained/nerdl_restaurant_100d*\")\n","path"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"B1gxDu0RlfwY"},"outputs":[],"source":["# you can use absolute path\n","ner_model = nlp.NerDLModel.load(path[0]) \\\n","    .setInputCols(['document', 'token', 'embeddings']) \\\n","    .setOutputCol('ner')"]},{"cell_type":"markdown","metadata":{"id":"EDOeu9q1EIk5"},"source":["**pipeline**"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"jPwcEsqsDDyb"},"outputs":[],"source":["document_assembler = nlp.DocumentAssembler() \\\n","    .setInputCol('text') \\\n","    .setOutputCol('document')\n","\n","sentence_detector = nlp.SentenceDetector()\\\n","  .setInputCols([\"document\"])\\\n","  .setOutputCol(\"sentence\")\n","\n","tokenizer = nlp.Tokenizer() \\\n","    .setInputCols(['sentence']) \\\n","    .setOutputCol('token')\n","\n","\n","embeddings = nlp.WordEmbeddingsModel.load(\"/root/cache_pretrained/glove_100d_en_2.4.0_2.4_1579690104032\") \\\n","    .setInputCols(\"sentence\", \"token\") \\\n","    .setOutputCol(\"embeddings\")\n","\n","nerdl = nlp.NerDLModel.load(\"/root/cache_pretrained/nerdl_restaurant_100d_en_3.3.4_3.0_1640949258750\")\\\n","    .setInputCols([\"sentence\", \"token\", \"embeddings\"])\\\n","    .setOutputCol(\"ner\")\n","\n","ner_converter = nlp.NerConverter()\\\n","    .setInputCols([\"sentence\", \"token\", \"ner\"])\\\n","    .setOutputCol(\"ner_chunk\")\n","\n","nlp_pipeline = nlp.Pipeline(\n","    stages=[\n","        document_assembler,\n","        sentence_detector,\n","        tokenizer,\n","        embeddings,\n","        nerdl,\n","        ner_converter\n","])\n","\n","text = [\"\"\"Hong Kong‚Äôs favourite pasta bar also offers one of the most reasonably priced lunch sets in town! With locations spread out all over the territory Sha Tin ‚Äì Pici‚Äôs formidable lunch menu reads like a highlight reel of the restaurant. Choose from starters like the burrata and arugula salad or freshly tossed tuna tartare, and reliable handmade pasta dishes like pappardelle. Finally, round out your effortless Italian meal with a tidy one-pot tiramisu, of course, an espresso to power you through the rest of the day.\"\"\"]\n","\n","data = spark.createDataFrame([text]).toDF(\"text\")\n","\n","result = nlp_pipeline.fit(data).transform(data)"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":301},"executionInfo":{"elapsed":4244,"status":"ok","timestamp":1695879328315,"user":{"displayName":"Sel√ßuk Meng√ºverdi","userId":"15361124602135819062"},"user_tz":240},"id":"ciT9NGYWDu9F","outputId":"e0f1a2f8-e238-4183-c37c-32d17706518b"},"outputs":[{"output_type":"display_data","data":{"text/plain":["<IPython.core.display.HTML object>"],"text/html":["\n","<style>\n","    @import url('https://fonts.googleapis.com/css2?family=Montserrat:wght@300;400;500;600;700&display=swap');\n","    @import url('https://fonts.googleapis.com/css2?family=Vistol Regular:wght@300;400;500;600;700&display=swap');\n","    \n","    .spark-nlp-display-scroll-entities {\n","        border: 1px solid #E7EDF0;\n","        border-radius: 3px;\n","        text-align: justify;\n","        \n","    }\n","    .spark-nlp-display-scroll-entities span {  \n","        font-size: 14px;\n","        line-height: 24px;\n","        color: #536B76;\n","        font-family: 'Montserrat', sans-serif !important;\n","    }\n","    \n","    .spark-nlp-display-entity-wrapper{\n","    \n","        display: inline-grid;\n","        text-align: center;\n","        border-radius: 4px;\n","        margin: 0 2px 5px 2px;\n","        padding: 1px\n","    }\n","    .spark-nlp-display-entity-name{\n","        font-size: 14px;\n","        line-height: 24px;\n","        font-family: 'Montserrat', sans-serif !important;\n","        \n","        background: #f1f2f3;\n","        border-width: medium;\n","        text-align: center;\n","        \n","        font-weight: 400;\n","        \n","        border-radius: 5px;\n","        padding: 2px 5px;\n","        display: block;\n","        margin: 3px 2px;\n","    \n","    }\n","    .spark-nlp-display-entity-type{\n","        font-size: 14px;\n","        line-height: 24px;\n","        color: #ffffff;\n","        font-family: 'Montserrat', sans-serif !important;\n","        \n","        text-transform: uppercase;\n","        \n","        font-weight: 500;\n","\n","        display: block;\n","        padding: 3px 5px;\n","    }\n","    \n","    .spark-nlp-display-entity-resolution{\n","        font-size: 14px;\n","        line-height: 24px;\n","        color: #ffffff;\n","        font-family: 'Vistol Regular', sans-serif !important;\n","        \n","        text-transform: uppercase;\n","        \n","        font-weight: 500;\n","\n","        display: block;\n","        padding: 3px 5px;\n","    }\n","    \n","    .spark-nlp-display-others{\n","        font-size: 14px;\n","        line-height: 24px;\n","        font-family: 'Montserrat', sans-serif !important;\n","        \n","        font-weight: 400;\n","    }\n","\n","</style>\n"," <span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #5A6D19\"><span class=\"spark-nlp-display-entity-name\">Hong Kong‚Äôs </span><span class=\"spark-nlp-display-entity-type\">Restaurant_Name</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #A2C842\"><span class=\"spark-nlp-display-entity-name\">favourite </span><span class=\"spark-nlp-display-entity-type\">Rating</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #1E89AF\"><span class=\"spark-nlp-display-entity-name\">pasta bar </span><span class=\"spark-nlp-display-entity-type\">Dish</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> also offers one of the </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #48A07A\"><span class=\"spark-nlp-display-entity-name\">most reasonably </span><span class=\"spark-nlp-display-entity-type\">Price</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> priced </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #C3332D\"><span class=\"spark-nlp-display-entity-name\">lunch </span><span class=\"spark-nlp-display-entity-type\">Hours</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> sets </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #7FB9C2\"><span class=\"spark-nlp-display-entity-name\">in town! </span><span class=\"spark-nlp-display-entity-type\">Location</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> With locations spread out all over the territory </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #5A6D19\"><span class=\"spark-nlp-display-entity-name\">Sha Tin ‚Äì Pici‚Äôs </span><span class=\"spark-nlp-display-entity-type\">Restaurant_Name</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> formidable lunch menu reads like a highlight reel of the restaurant. Choose from starters like the </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #1E89AF\"><span class=\"spark-nlp-display-entity-name\">burrata </span><span class=\"spark-nlp-display-entity-type\">Dish</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> and </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #1E89AF\"><span class=\"spark-nlp-display-entity-name\">arugula salad </span><span class=\"spark-nlp-display-entity-type\">Dish</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> or </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #1E89AF\"><span class=\"spark-nlp-display-entity-name\">freshly tossed tuna tartare </span><span class=\"spark-nlp-display-entity-type\">Dish</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\">, and </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #48A07A\"><span class=\"spark-nlp-display-entity-name\">reliable </span><span class=\"spark-nlp-display-entity-type\">Price</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #1E89AF\"><span class=\"spark-nlp-display-entity-name\">handmade pasta </span><span class=\"spark-nlp-display-entity-type\">Dish</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> dishes like </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #1E89AF\"><span class=\"spark-nlp-display-entity-name\">pappardelle </span><span class=\"spark-nlp-display-entity-type\">Dish</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\">. Finally, round out your </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #1B665E\"><span class=\"spark-nlp-display-entity-name\">effortless </span><span class=\"spark-nlp-display-entity-type\">Amenity</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #561405\"><span class=\"spark-nlp-display-entity-name\">Italian </span><span class=\"spark-nlp-display-entity-type\">Cuisine</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> meal with a </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #1B665E\"><span class=\"spark-nlp-display-entity-name\">tidy one-pot </span><span class=\"spark-nlp-display-entity-type\">Amenity</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> tiramisu, of course, an </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #1E89AF\"><span class=\"spark-nlp-display-entity-name\">espresso </span><span class=\"spark-nlp-display-entity-type\">Dish</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> to power you through the rest of the day.</span></div>"]},"metadata":{}}],"source":["from sparknlp_display import NerVisualizer\n","\n","for i in range(len(text)):\n","  NerVisualizer().display(\n","      result = result.collect()[i],\n","      label_col = 'ner_chunk',\n","      document_col = 'document'\n","  )"]},{"cell_type":"markdown","metadata":{"id":"iZ1cUVdVp9nV"},"source":["## showPublicPipelines"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":921,"status":"ok","timestamp":1695879329231,"user":{"displayName":"Sel√ßuk Meng√ºverdi","userId":"15361124602135819062"},"user_tz":240},"id":"XlLFbKSnfWIE","outputId":"be28db56-027d-4ff4-b1b5-317ba772eee4"},"outputs":[{"output_type":"stream","name":"stdout","text":["+-------------------------------------+------+---------+\n","| Pipeline                            | lang | version |\n","+-------------------------------------+------+---------+\n","| dependency_parse                    |  en  | 2.0.2   |\n","| check_spelling                      |  en  | 2.1.0   |\n","| match_datetime                      |  en  | 2.1.0   |\n","| match_pattern                       |  en  | 2.1.0   |\n","| clean_pattern                       |  en  | 2.1.0   |\n","| clean_stop                          |  en  | 2.1.0   |\n","| match_phrases                       |  en  | 2.1.0   |\n","| movies_sentiment_analysis           |  en  | 2.1.0   |\n","| explain_document_ml                 |  en  | 2.1.0   |\n","| clean_slang                         |  en  | 2.1.0   |\n","| analyze_sentiment                   |  en  | 2.1.0   |\n","| explain_document_dl                 |  en  | 2.1.0   |\n","| explain_document_dl_fast            |  en  | 2.1.0   |\n","| recognize_entities_dl               |  en  | 2.1.0   |\n","| recognize_entities_bert             |  en  | 2.1.0   |\n","| explain_document_dl_noncontrib      |  en  | 2.1.0   |\n","| explain_document_dl_fast_noncontrib |  en  | 2.1.0   |\n","| recognize_entities_dl_noncontrib    |  en  | 2.1.0   |\n","| recognize_entities_bert_noncontrib  |  en  | 2.1.0   |\n","| analyze_sentiment                   |  en  | 2.1.0   |\n","| movies_sentiment_analysis           |  en  | 2.1.0   |\n","| explain_document_ml                 |  en  | 2.1.0   |\n","| analyze_sentiment                   |  en  | 2.1.0   |\n","| dependency_parse                    |  en  | 2.1.0   |\n","| onto_recognize_entities_sm          |  en  | 2.1.0   |\n","| onto_recognize_entities_lg          |  en  | 2.1.0   |\n","| match_chunks                        |  en  | 2.1.0   |\n","| explain_document_dl_noncontrib      |  en  | 2.1.0   |\n","| explain_document_dl                 |  en  | 2.1.0   |\n","| match_chunks                        |  en  | 2.1.0   |\n","| match_chunks                        |  en  | 2.2.0   |\n","| onto_recognize_entities_lg          |  en  | 2.4.0   |\n","| onto_recognize_entities_sm          |  en  | 2.4.0   |\n","| check_spelling                      |  en  | 2.4.0   |\n","| match_datetime                      |  en  | 2.4.0   |\n","| clean_pattern                       |  en  | 2.4.0   |\n","| match_chunks                        |  en  | 2.4.0   |\n","| recognize_entities_dl               |  en  | 2.4.0   |\n","| explain_document_ml                 |  en  | 2.4.0   |\n","| dependency_parse                    |  en  | 2.4.0   |\n","| clean_stop                          |  en  | 2.4.0   |\n","| explain_document_dl                 |  en  | 2.4.0   |\n","| match_phrases                       |  en  | 2.4.0   |\n","| clean_slang                         |  en  | 2.4.0   |\n","| analyze_sentiment                   |  en  | 2.4.0   |\n","| match_pattern                       |  en  | 2.4.0   |\n","| match_datetime                      |  en  | 2.4.0   |\n","+-------------------------------------+------+---------+\n","\n"]}],"source":["from sparknlp.pretrained import ResourceDownloader\n","\n","ResourceDownloader.showPublicPipelines(lang=\"en\", version=\"2.4.0\")"]},{"cell_type":"markdown","metadata":{"id":"csWTlU4ql_H0"},"source":["## downloadPipeline\n","\n","    Downloads and loads a pipeline with the default downloader.\n","\n","    Parameters:\n","        name      :  Name of the pipeline\n","        language  :  Language of the pipeline\n","        remote_loc:  Directory of the remote Spark NLP Folder, by default None"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":20543,"status":"ok","timestamp":1695879349771,"user":{"displayName":"Sel√ßuk Meng√ºverdi","userId":"15361124602135819062"},"user_tz":240},"id":"GEP0XyxucQMD","outputId":"db15b823-71aa-4d41-da2c-e0f662c462ed"},"outputs":[{"output_type":"stream","name":"stdout","text":["explain_document_ml download started this may take some time.\n","Approx size to download 9 MB\n","[OK!]\n"]}],"source":["explain_document_ml = ResourceDownloader.downloadPipeline(\"explain_document_ml\", language =\"en\", remote_loc = \"public/models\"  )"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":884,"status":"ok","timestamp":1695879350640,"user":{"displayName":"Sel√ßuk Meng√ºverdi","userId":"15361124602135819062"},"user_tz":240},"id":"iQVjEARefMGI","outputId":"7245032c-ab54-466c-8fae-2e5f77be2b6c"},"outputs":[{"output_type":"execute_result","data":{"text/plain":["[('Peter', 'Peter', 'peter', 'Peter'),\n"," ('Parker', 'Parker', 'parker', 'Parker'),\n"," ('is', 'be', 'i', 'is'),\n"," ('a', 'a', 'a', 'a'),\n"," ('nice', 'nice', 'nice', 'nice'),\n"," ('guy', 'guy', 'gui', 'guy'),\n"," ('and', 'and', 'and', 'and'),\n"," ('lives', 'life', 'live', 'lives'),\n"," ('in', 'in', 'in', 'in'),\n"," ('New', 'New', 'new', 'New'),\n"," ('York', 'York', 'york', 'York')]"]},"metadata":{},"execution_count":20}],"source":["pipeline = nlp.LightPipeline(explain_document_ml)\n","\n","text = 'Peter Parker is a nice guy and lives in New York'\n","\n","result = pipeline.annotate(text)\n","\n","list(zip(result['token'], result['lemmas'], result['stems'], result['spell']))"]},{"cell_type":"markdown","metadata":{"id":"Rvg5Yuukl8xq"},"source":["## downloadModelDirectly\n","\n","    Downloads a model directly to the cache folder.\n","    You can use to copy-paste the s3 URI from the model hub and download the model.\n","    For available s3 URI and models,\n","\n","please see the [Models Hub](https://nlp.johnsnowlabs.com/models).\n"]},{"cell_type":"markdown","metadata":{"id":"Kv8RQ6-HJMJ-"},"source":["    Parameters:\n","        name       : Name of the model or s3 URI\n","        remote_loc : Directory of the remote Spark NLP Folder, by default \"public/models\"\n","        unzip      : Used to unzip model, by default 'True'\n"]},{"cell_type":"markdown","metadata":{"id":"NfUhSu4iOwDq"},"source":["### with **S3 URI**"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"Au0UxTUgRdI8"},"outputs":[],"source":["# You can download the model directly.\n","# parameters are model name, remote location, and unzip\n","\n","# model link:  https://sparknlp.org/2022/06/01/nerdl_conll_elmo_en_3_0.html\n","\n","s3_uri = \"s3://auxdata.johnsnowlabs.com/public/models/nerdl_conll_elmo_en_4.0.0_3.0_1654103884644.zip\"\n","\n","ResourceDownloader.downloadModelDirectly(s3_uri,  \"public/models\", unzip=True)\n"]},{"cell_type":"markdown","metadata":{"id":"nGU8sceZbUNB"},"source":["### with **model name**"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"TAPaoUSfJN4j"},"outputs":[],"source":["# You can download the model directly.\n","# parameters are model name, remote location, and unzip\n","\n","model_name = \"public/models/nerdl_restaurant_100d_en_3.3.4_3.0_1640949258750.zip\"\n","\n","ResourceDownloader.downloadModelDirectly(model_name,  \"public/models\", unzip=True)\n"]},{"cell_type":"markdown","metadata":{"id":"ErPktcnEmBbN"},"source":["## clearCache\n","\n","    Clears the cache entry of a model\n","\n","    parameters:\n","      name      : Name of the model\n","      language  : Language of the model\n","      remote_loc: Directory of the remote Spark NLP Folder, by default None\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"d7ycKn_bmFs9"},"outputs":[],"source":["# ResourceDownloader.clearCache(\"nerdl_conll_elmo\",\"en\",\"public/models\")\n","\n","#Note: if S3 Amazon throws credential error ignore it, just check your cache folder, your model will be deleted"]},{"cell_type":"markdown","metadata":{"id":"5YyC_l5dl2Sp"},"source":["# InternalResourceDownloader\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"REgLeuVDrUek"},"outputs":[],"source":["from sparknlp_jsl.pretrained import InternalResourceDownloader"]},{"cell_type":"markdown","metadata":{"id":"_pYx5HsXrjoe"},"source":["### showPrivateModels\n","\n","    show private models available for download\n","\n","    Parameters:\n","        annotator : The annotator to filter by. Defaults to None.\n","        lang      : The language to filter by. Defaults to None.\n","        version   : The version to filter by. Defaults to None.\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":1104,"status":"ok","timestamp":1695879354027,"user":{"displayName":"Sel√ßuk Meng√ºverdi","userId":"15361124602135819062"},"user_tz":240},"id":"J45qdAKDrY8k","outputId":"130a1f33-df74-4643-fce2-c670cb20dee0"},"outputs":[{"output_type":"stream","name":"stdout","text":["+----------------------------------------+------+---------+\n","| Model                                  | lang | version |\n","+----------------------------------------+------+---------+\n","| nerdl_tumour_demo                      |  en  | 1.7.3   |\n","| nerdl_tumour_demo                      |  en  | 1.8.0   |\n","| nerdl_tumour_demo                      |  en  | 2.0.2   |\n","| ner_healthcare                         |  en  | 2.4.4   |\n","| ner_radiology                          |  en  | 2.7.0   |\n","| ner_deid_augmented                     |  en  | 2.7.1   |\n","| ner_deidentify_dl                      |  en  | 2.7.2   |\n","| ner_events_admission_clinical          |  en  | 2.7.4   |\n","| ner_clinical                           |  en  | 3.0.0   |\n","| ner_radiology                          |  en  | 3.0.0   |\n","| ner_bionlp                             |  en  | 3.0.0   |\n","| ner_posology                           |  en  | 3.0.0   |\n","| ner_deid_augmented                     |  en  | 3.0.0   |\n","| ner_anatomy                            |  en  | 3.0.0   |\n","| ner_risk_factors                       |  en  | 3.0.0   |\n","| ner_chemprot_clinical                  |  en  | 3.0.0   |\n","| ner_posology_small                     |  en  | 3.0.0   |\n","| ner_posology_greedy                    |  en  | 3.0.0   |\n","| ner_deid_enriched                      |  en  | 3.0.0   |\n","| ner_drugs_greedy                       |  en  | 3.0.0   |\n","| jsl_ner_wip_clinical                   |  en  | 3.0.0   |\n","| ner_posology_large                     |  en  | 3.0.0   |\n","| jsl_ner_wip_greedy_clinical            |  en  | 3.0.0   |\n","| ner_clinical_large                     |  en  | 3.0.0   |\n","| ner_diseases                           |  en  | 3.0.0   |\n","| ner_aspect_based_sentiment             |  en  | 3.0.0   |\n","| ner_cellular                           |  en  | 3.0.0   |\n","| ner_drugs                              |  en  | 3.0.0   |\n","| ner_jsl                                |  en  | 3.0.0   |\n","| ner_anatomy_coarse_biobert             |  en  | 3.0.0   |\n","| ner_cancer_genetics                    |  en  | 3.0.0   |\n","| ner_drugs_large                        |  en  | 3.0.0   |\n","| ner_events_admission_clinical          |  en  | 3.0.0   |\n","| ner_deidentify_dl                      |  en  | 3.0.0   |\n","| ner_deid_synthetic                     |  en  | 3.0.0   |\n","| ner_jsl_enriched                       |  en  | 3.0.0   |\n","| ner_deid_large                         |  en  | 3.0.0   |\n","| ner_human_phenotype_gene_clinical      |  en  | 3.0.0   |\n","| ner_human_phenotype_go_clinical        |  en  | 3.0.0   |\n","| ner_events_clinical                    |  en  | 3.0.0   |\n","| ner_anatomy_coarse                     |  en  | 3.0.0   |\n","| ner_bionlp_biobert                     |  en  | 3.0.0   |\n","| ner_deid_sd_large                      |  en  | 3.0.0   |\n","| ner_risk_factors_biobert               |  en  | 3.0.0   |\n","| ner_posology_healthcare                |  en  | 3.0.0   |\n","| ner_diseases_large                     |  en  | 3.0.0   |\n","| ner_ade_biobert                        |  en  | 3.0.0   |\n","| ner_jsl_enriched_biobert               |  en  | 3.0.0   |\n","| ner_ade_healthcare                     |  en  | 3.0.0   |\n","| ner_deid_sd                            |  en  | 3.0.0   |\n","| ner_events_healthcare                  |  en  | 3.0.0   |\n","| ner_financial_contract                 |  en  | 3.0.0   |\n","| ner_chemprot_biobert                   |  en  | 3.0.0   |\n","| ner_ade_clinicalbert                   |  en  | 3.0.0   |\n","| ner_posology_large_biobert             |  en  | 3.0.0   |\n","| ner_radiology_wip_clinical             |  en  | 3.0.0   |\n","| ner_jsl_biobert                        |  en  | 3.0.0   |\n","| ner_clinical_biobert                   |  en  | 3.0.0   |\n","| ner_deid_enriched_biobert              |  en  | 3.0.0   |\n","| ner_posology_biobert                   |  en  | 3.0.0   |\n","| ner_cellular_biobert                   |  en  | 3.0.0   |\n","| jsl_ner_wip_modifier_clinical          |  en  | 3.0.0   |\n","| ner_measurements_clinical              |  en  | 3.0.0   |\n","| ner_medmentions_coarse                 |  en  | 3.0.0   |\n","| ner_chemicals                          |  en  | 3.0.0   |\n","| ner_events_biobert                     |  en  | 3.0.0   |\n","| ner_bacterial_species                  |  en  | 3.0.0   |\n","| ner_diseases_biobert                   |  en  | 3.0.0   |\n","| ner_human_phenotype_gene_biobert       |  en  | 3.0.0   |\n","| ner_deid_biobert                       |  en  | 3.0.0   |\n","| nerdl_tumour_demo                      |  en  | 3.0.0   |\n","| ner_human_phenotype_go_biobert         |  en  | 3.0.0   |\n","| ner_anatomy_biobert                    |  en  | 3.0.0   |\n","| ner_ade_clinical                       |  en  | 3.0.0   |\n","| jsl_rd_ner_wip_greedy_clinical         |  en  | 3.0.0   |\n","| ner_healthcare                         |  en  | 3.0.0   |\n","| ner_deid_generic_augmented             |  en  | 3.0.3   |\n","| ner_deid_subentity_augmented           |  en  | 3.0.3   |\n","| ner_deid_generic_glove                 |  en  | 3.0.4   |\n","| ner_deid_subentity_glove               |  en  | 3.0.4   |\n","| ner_jsl                                |  en  | 3.1.0   |\n","| ner_genetic_variants                   |  en  | 3.1.0   |\n","| ner_deid_subentity_augmented           |  en  | 3.1.0   |\n","| ner_deid_generic_augmented             |  en  | 3.1.0   |\n","| jsl_ner_wip_greedy_biobert             |  en  | 3.1.3   |\n","| jsl_rd_ner_wip_greedy_biobert          |  en  | 3.1.3   |\n","| ner_jsl_greedy_biobert                 |  en  | 3.2.0   |\n","| ner_jsl_slim                           |  en  | 3.2.0   |\n","| ner_posology_experimental              |  en  | 3.1.3   |\n","| ner_deid_subentity_augmented           |  en  | 3.2.0   |\n","| ner_jsl_biobert                        |  en  | 3.2.0   |\n","| ner_chexpert                           |  en  | 3.3.0   |\n","| ner_jsl_enriched                       |  en  | 3.3.0   |\n","| ner_covid_trials                       |  en  | 3.3.2   |\n","| ner_chemd_clinical                     |  en  | 3.3.0   |\n","| ner_nihss                              |  en  | 3.3.2   |\n","| ner_biomarker                          |  en  | 3.3.3   |\n","| ner_deid_subentity_augmented_i2b2      |  en  | 3.3.2   |\n","| ner_drugprot_clinical                  |  en  | 3.3.3   |\n","| ner_abbreviation_clinical              |  en  | 3.3.4   |\n","| ner_supplement_clinical                |  en  | 3.3.4   |\n","| ner_nature_nero_clinical               |  en  | 3.3.4   |\n","| ner_biomedical_bc2gm                   |  en  | 3.5.1   |\n","| ner_living_species                     |  en  | 3.5.3   |\n","| ner_living_species_biobert             |  en  | 3.5.3   |\n","| ner_clinical_trials_abstracts          |  en  | 3.5.3   |\n","| ner_pathogen                           |  en  | 4.0.0   |\n","| ner_oncology_wip                       |  en  | 3.5.0   |\n","| ner_ade_binary                         |  en  | 4.0.0   |\n","| ner_oncology_unspecific_posology_wip   |  en  | 4.0.0   |\n","| ner_oncology_wip                       |  en  | 4.0.0   |\n","| ner_oncology_therapy_wip               |  en  | 4.0.0   |\n","| ner_oncology_posology_wip              |  en  | 4.0.0   |\n","| ner_oncology_anatomy_general_wip       |  en  | 4.0.0   |\n","| ner_oncology_tnm_wip                   |  en  | 4.0.0   |\n","| ner_oncology_demographics_wip          |  en  | 4.0.0   |\n","| ner_oncology_biomarker_wip             |  en  | 4.0.0   |\n","| ner_oncology_anatomy_granular_wip      |  en  | 4.0.0   |\n","| ner_oncology_test_wip                  |  en  | 4.0.0   |\n","| ner_oncology_diagnosis_wip             |  en  | 4.0.0   |\n","| ner_oncology_response_to_treatment_wip |  en  | 4.0.0   |\n","| ner_oncology_unspecific_posology       |  en  | 4.0.0   |\n","| ner_oncology                           |  en  | 4.0.0   |\n","| ner_oncology_tnm                       |  en  | 4.0.0   |\n","| ner_oncology_anatomy_general           |  en  | 4.0.0   |\n","| ner_oncology_therapy                   |  en  | 4.0.0   |\n","| ner_oncology_test                      |  en  | 4.0.0   |\n","| ner_oncology_diagnosis                 |  en  | 4.0.0   |\n","| ner_oncology_demographics              |  en  | 4.0.0   |\n","| ner_oncology_anatomy_granular          |  en  | 4.0.0   |\n","| ner_oncology_response_to_treatment     |  en  | 4.0.0   |\n","| ner_oncology_posology                  |  en  | 4.0.0   |\n","| ner_oncology_biomarker                 |  en  | 4.0.0   |\n","+----------------------------------------+------+---------+\n","\n"]}],"source":["InternalResourceDownloader.showPrivateModels(\"MedicalNerModel\",\"en\", \"4.0.0\")"]},{"cell_type":"markdown","metadata":{"id":"s-04xNqgZ5ZP"},"source":["### showPrivatePipelines\n","\n","    show private models available for download\n","\n","    Parameters:\n","        lang    : The language to filter by. Defaults to None.\n","        version : The version to filter by. Defaults to None.\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":610,"status":"ok","timestamp":1695879354631,"user":{"displayName":"Sel√ßuk Meng√ºverdi","userId":"15361124602135819062"},"user_tz":240},"id":"LRerv5-EZ5ZQ","outputId":"19e06c8d-e151-4340-f8ce-d24076cb9b0e"},"outputs":[{"output_type":"stream","name":"stdout","text":["+--------------------------------------------------------+------+---------+\n","| Pipeline                                               | lang | version |\n","+--------------------------------------------------------+------+---------+\n","| clinical_analysis                                      |  en  | 2.4.0   |\n","| clinical_ner_assertion                                 |  en  | 2.4.0   |\n","| clinical_deidentification                              |  en  | 2.4.0   |\n","| explain_clinical_doc_ade                               |  en  | 2.7.3   |\n","| recognize_entities_posology                            |  en  | 3.0.0   |\n","| explain_clinical_doc_carp                              |  en  | 3.0.0   |\n","| explain_clinical_doc_ade                               |  en  | 3.0.0   |\n","| explain_clinical_doc_era                               |  en  | 3.0.0   |\n","| icd10cm_snomed_mapping                                 |  en  | 3.0.2   |\n","| snomed_icd10cm_mapping                                 |  en  | 3.0.2   |\n","| icd10cm_umls_mapping                                   |  en  | 3.0.2   |\n","| snomed_umls_mapping                                    |  en  | 3.0.2   |\n","| rxnorm_umls_mapping                                    |  en  | 3.0.2   |\n","| mesh_umls_mapping                                      |  en  | 3.0.2   |\n","| rxnorm_mesh_mapping                                    |  en  | 3.0.2   |\n","| clinical_deidentification                              |  en  | 3.0.3   |\n","| clinical_deidentification_glove                        |  en  | 3.0.4   |\n","| rxnorm_mesh_mapping                                    |  en  | 3.1.0   |\n","| snomed_umls_mapping                                    |  en  | 3.1.0   |\n","| rxnorm_umls_mapping                                    |  en  | 3.1.0   |\n","| mesh_umls_mapping                                      |  en  | 3.1.0   |\n","| icd10cm_umls_mapping                                   |  en  | 3.1.0   |\n","| explain_clinical_doc_ade                               |  en  | 3.1.2   |\n","| ner_profiling_biobert                                  |  en  | 3.2.3   |\n","| ner_profiling_clinical                                 |  en  | 3.2.3   |\n","| ner_profiling_biobert                                  |  en  | 3.3.1   |\n","| ner_profiling_clinical                                 |  en  | 3.3.1   |\n","| ner_model_finder                                       |  en  | 3.3.2   |\n","| icd10_icd9_mapping                                     |  en  | 3.3.4   |\n","| clinical_deidentification                              |  en  | 3.4.1   |\n","| clinical_deidentification_glove                        |  en  | 3.4.1   |\n","| bert_token_classifier_ner_cellular_pipeline            |  en  | 3.4.1   |\n","| bert_token_classifier_ner_chemprot_pipeline            |  en  | 3.4.1   |\n","| bert_token_classifier_ner_clinical_pipeline            |  en  | 3.4.1   |\n","| bert_token_classifier_ner_chemicals_pipeline           |  en  | 3.4.1   |\n","| bert_token_classifier_ner_anatomy_pipeline             |  en  | 3.4.1   |\n","| bert_token_classifier_ner_ade_pipeline                 |  en  | 3.4.1   |\n","| clinical_deidentification_glove_augmented              |  en  | 3.4.1   |\n","| ner_bacterial_species_pipeline                         |  en  | 3.4.1   |\n","| ner_measurements_clinical_pipeline                     |  en  | 3.4.1   |\n","| ner_medmentions_coarse_pipeline                        |  en  | 3.4.1   |\n","| bert_token_classifier_drug_development_trials_pipeline |  en  | 3.4.1   |\n","| ner_events_healthcare_pipeline                         |  en  | 3.4.1   |\n","| ner_healthcare_pipeline                                |  en  | 3.4.1   |\n","| ner_posology_healthcare_pipeline                       |  en  | 3.4.1   |\n","| ner_clinical_biobert_pipeline                          |  en  | 3.4.1   |\n","| ner_abbreviation_clinical_pipeline                     |  en  | 3.4.1   |\n","| ner_ade_biobert_pipeline                               |  en  | 3.4.1   |\n","| ner_radiology_wip_clinical_pipeline                    |  en  | 3.4.1   |\n","| ner_ade_clinicalbert_pipeline                          |  en  | 3.4.1   |\n","| ner_events_biobert_pipeline                            |  en  | 3.4.1   |\n","| ner_posology_large_biobert_pipeline                    |  en  | 3.4.1   |\n","| ner_anatomy_coarse_biobert_pipeline                    |  en  | 3.4.1   |\n","| ner_posology_small_pipeline                            |  en  | 3.4.1   |\n","| ner_anatomy_coarse_pipeline                            |  en  | 3.4.1   |\n","| ner_drugs_greedy_pipeline                              |  en  | 3.4.1   |\n","| ner_posology_large_pipeline                            |  en  | 3.4.1   |\n","| ner_clinical_large_pipeline                            |  en  | 3.4.1   |\n","| ner_drugs_large_pipeline                               |  en  | 3.4.1   |\n","| ner_drugs_pipeline                                     |  en  | 3.4.1   |\n","| ner_posology_greedy_pipeline                           |  en  | 3.4.1   |\n","| ner_drugprot_clinical_pipeline                         |  en  | 3.4.1   |\n","| ner_genetic_variants_pipeline                          |  en  | 3.4.1   |\n","| ner_posology_experimental_pipeline                     |  en  | 3.4.1   |\n","| ner_biomarker_pipeline                                 |  en  | 3.4.1   |\n","| ner_posology_biobert_pipeline                          |  en  | 3.4.1   |\n","| ner_posology_pipeline                                  |  en  | 3.4.1   |\n","| ner_risk_factors_biobert_pipeline                      |  en  | 3.4.1   |\n","| ner_deidentify_dl_pipeline                             |  en  | 3.4.1   |\n","| ner_deid_subentity_augmented_i2b2_pipeline             |  en  | 3.4.1   |\n","| ner_deid_subentity_augmented_pipeline                  |  en  | 3.4.1   |\n","| ner_cellular_biobert_pipeline                          |  en  | 3.4.1   |\n","| ner_jsl_slim_pipeline                                  |  en  | 3.4.1   |\n","| ner_cellular_pipeline                                  |  en  | 3.4.1   |\n","| ner_deid_sd_large_pipeline                             |  en  | 3.4.1   |\n","| ner_nihss_pipeline                                     |  en  | 3.4.1   |\n","| ner_jsl_greedy_biobert_pipeline                        |  en  | 3.4.1   |\n","| ner_deid_sd_pipeline                                   |  en  | 3.4.1   |\n","| ner_chemicals_pipeline                                 |  en  | 3.4.1   |\n","| ner_jsl_greedy_pipeline                                |  en  | 3.4.1   |\n","| ner_jsl_enriched_biobert_pipeline                      |  en  | 3.4.1   |\n","| ner_jsl_enriched_pipeline                              |  en  | 3.4.1   |\n","| ner_jsl_pipeline                                       |  en  | 3.4.1   |\n","| ner_chemprot_biobert_pipeline                          |  en  | 3.4.1   |\n","| ner_deid_enriched_biobert_pipeline                     |  en  | 3.4.1   |\n","| ner_human_phenotype_go_biobert_pipeline                |  en  | 3.4.1   |\n","| ner_human_phenotype_go_clinical_pipeline               |  en  | 3.4.1   |\n","| ner_chemprot_clinical_pipeline                         |  en  | 3.4.1   |\n","| ner_human_phenotype_gene_clinical_pipeline             |  en  | 3.4.1   |\n","| jsl_rd_ner_wip_greedy_clinical_pipeline                |  en  | 3.4.1   |\n","| ner_human_phenotype_gene_biobert_pipeline              |  en  | 3.4.1   |\n","| jsl_rd_ner_wip_greedy_biobert_pipeline                 |  en  | 3.4.1   |\n","| jsl_ner_wip_modifier_clinical_pipeline                 |  en  | 3.4.1   |\n","| ner_deid_biobert_pipeline                              |  en  | 3.4.1   |\n","| jsl_ner_wip_greedy_clinical_pipeline                   |  en  | 3.4.1   |\n","| jsl_ner_wip_greedy_biobert_pipeline                    |  en  | 3.4.1   |\n","| jsl_ner_wip_clinical_pipeline                          |  en  | 3.4.1   |\n","| bert_token_classifier_ner_jsl_slim_pipeline            |  en  | 3.4.1   |\n","| ner_deid_augmented_pipeline                            |  en  | 3.4.1   |\n","| bert_token_classifier_ner_drugs_pipeline               |  en  | 3.4.1   |\n","| bert_token_classifier_ner_deid_pipeline                |  en  | 3.4.1   |\n","| bert_token_classifier_ner_bionlp_pipeline              |  en  | 3.4.1   |\n","| ner_cancer_genetics_pipeline                           |  en  | 3.4.1   |\n","| ner_clinical_pipeline                                  |  en  | 3.4.1   |\n","| ner_deid_enriched_pipeline                             |  en  | 3.4.1   |\n","| ner_deid_generic_augmented_pipeline                    |  en  | 3.4.1   |\n","| ner_deid_large_pipeline                                |  en  | 3.4.1   |\n","| ner_diseases_pipeline                                  |  en  | 3.4.1   |\n","| ner_diseases_large_pipeline                            |  en  | 3.4.1   |\n","| ner_diseases_biobert_pipeline                          |  en  | 3.4.1   |\n","| ner_events_admission_clinical_pipeline                 |  en  | 3.4.1   |\n","| ner_events_clinical_pipeline                           |  en  | 3.4.1   |\n","| bert_token_classifier_ner_jsl_pipeline                 |  en  | 3.4.1   |\n","| ner_jsl_biobert_pipeline                               |  en  | 3.4.1   |\n","| ner_anatomy_pipeline                                   |  en  | 3.4.1   |\n","| ner_bionlp_pipeline                                    |  en  | 3.4.1   |\n","| ner_bionlp_biobert_pipeline                            |  en  | 3.4.1   |\n","| ner_chexpert_pipeline                                  |  en  | 3.4.1   |\n","| ner_radiology_pipeline                                 |  en  | 3.4.1   |\n","| ner_risk_factors_pipeline                              |  en  | 3.4.1   |\n","| bert_token_classifier_ner_bacteria_pipeline            |  en  | 3.4.1   |\n","| ner_anatomy_biobert_pipeline                           |  en  | 3.4.1   |\n","| ner_ade_healthcare_pipeline                            |  en  | 3.4.1   |\n","| ner_ade_clinical_pipeline                              |  en  | 3.4.1   |\n","| re_temporal_events_enriched_clinical_pipeline          |  en  | 3.4.1   |\n","| re_date_clinical_pipeline                              |  en  | 3.4.1   |\n","| re_human_phenotype_gene_clinical_pipeline              |  en  | 3.4.1   |\n","| re_test_result_date_pipeline                           |  en  | 3.4.1   |\n","| re_temporal_events_clinical_pipeline                   |  en  | 3.4.1   |\n","| re_bodypart_proceduretest_pipeline                     |  en  | 3.4.1   |\n","| re_test_problem_finding_pipeline                       |  en  | 3.4.1   |\n","| re_bodypart_directions_pipeline                        |  en  | 3.4.1   |\n","| explain_clinical_doc_radiology                         |  en  | 3.4.2   |\n","| explain_clinical_doc_medication                        |  en  | 3.4.2   |\n","| spellcheck_clinical_pipeline                           |  en  | 3.4.1   |\n","| spellcheck_clinical_pipeline                           |  en  | 3.4.2   |\n","| rct_binary_classifier_biobert_pipeline                 |  en  | 3.4.2   |\n","| bert_sequence_classifier_binary_rct_biobert_pipeline   |  en  | 3.4.2   |\n","| rct_binary_classifier_use_pipeline                     |  en  | 3.4.2   |\n","| ner_biomedical_bc2gm_pipeline                          |  en  | 3.5.3   |\n","| ner_clinical_trials_abstracts_pipeline                 |  en  | 3.5.3   |\n","| icd10cm_snomed_mapping                                 |  en  | 3.5.3   |\n","| rxnorm_ndc_mapping                                     |  en  | 3.5.3   |\n","| snomed_umls_mapping                                    |  en  | 3.5.3   |\n","| rxnorm_umls_mapping                                    |  en  | 3.5.3   |\n","| mesh_umls_mapping                                      |  en  | 3.5.3   |\n","| icd10cm_umls_mapping                                   |  en  | 3.5.3   |\n","| snomed_icdo_mapping                                    |  en  | 3.5.3   |\n","| icdo_snomed_mapping                                    |  en  | 3.5.3   |\n","| snomed_icd10cm_mapping                                 |  en  | 3.5.3   |\n","| ner_pathogen_pipeline                                  |  en  | 4.0.0   |\n","| explain_clinical_doc_ade                               |  en  | 4.0.0   |\n","| clinical_deidentification_glove_augmented              |  en  | 4.0.0   |\n","| clinical_deidentification_wip                          |  en  | 4.0.0   |\n","| clinical_deidentification                              |  en  | 4.0.0   |\n","| clinical_deidentification_slim                         |  en  | 4.0.0   |\n","| umls_drug_substance_resolver_pipeline                  |  en  | 4.0.0   |\n","| umls_disease_syndrome_resolver_pipeline                |  en  | 4.0.0   |\n","| umls_clinical_findings_resolver_pipeline               |  en  | 4.0.0   |\n","| umls_drug_resolver_pipeline                            |  en  | 4.0.0   |\n","| umls_major_concepts_resolver_pipeline                  |  en  | 4.0.0   |\n","| ner_medication_pipeline                                |  en  | 4.0.0   |\n","| explain_clinical_doc_radiology                         |  en  | 4.0.0   |\n","| explain_clinical_doc_medication                        |  en  | 4.0.0   |\n","+--------------------------------------------------------+------+---------+\n","\n"]}],"source":["InternalResourceDownloader.showPrivatePipelines(\"en\", \"4.0.0\")"]},{"cell_type":"markdown","metadata":{"id":"wPxjuKjtrtjl"},"source":["### downloadModel\n","    Download a model from S3\n","\n","    parameters:\n","        Args:\n","            reader    : The reader class to use to load the model.\n","            name      : The name of the model to download.\n","            language  : The language of the model to download.\n","            remote_loc: The remote location of the model. Defaults to None.\n"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":7552,"status":"ok","timestamp":1695879362166,"user":{"displayName":"Sel√ßuk Meng√ºverdi","userId":"15361124602135819062"},"user_tz":240},"id":"5Kk4ERiklfqy","outputId":"6eb60804-7147-4681-c909-b1e578df4336"},"outputs":[{"output_type":"stream","name":"stdout","text":["ner_clinical download started this may take some time.\n","[OK!]\n"]}],"source":["ner_clinical = InternalResourceDownloader.downloadModel(medical.NerModel,\"ner_clinical\",\"en\",remote_loc=\"clinical/models\")"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":80437,"status":"ok","timestamp":1695879442585,"user":{"displayName":"Sel√ßuk Meng√ºverdi","userId":"15361124602135819062"},"user_tz":240},"id":"nm01vIdoaccf","outputId":"c232172f-2376-4cf9-d7b1-c50084c77c69"},"outputs":[{"output_type":"stream","name":"stdout","text":["embeddings_clinical download started this may take some time.\n","Approximate size to download 1.6 GB\n","[OK!]\n"]}],"source":["embeddings = ResourceDownloader.downloadModel(nlp.WordEmbeddingsModel,\"embeddings_clinical\",\"en\",remote_loc=\"clinical/models\")"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"yu3MGLOnsqAc"},"outputs":[],"source":["document_assembler = nlp.DocumentAssembler() \\\n","    .setInputCol('text') \\\n","    .setOutputCol('document')\n","\n","sentence_detector = nlp.SentenceDetector()\\\n","    .setInputCols([\"document\"])\\\n","    .setOutputCol(\"sentence\")\n","\n","tokenizer = nlp.Tokenizer() \\\n","    .setInputCols(['sentence']) \\\n","    .setOutputCol('token')\n","\n","embeddings = nlp.WordEmbeddingsModel.load(\"/root/cache_pretrained/embeddings_clinical_en_2.4.0_2.4_1580237286004\") \\\n","    .setInputCols(\"sentence\", \"token\") \\\n","    .setOutputCol(\"embeddings\")\n","\n","# or just set Input and Output columns\n","# embeddings\\\n","#     .setInputCols(\"sentence\", \"token\") \\\n","#     .setOutputCol(\"embeddings\")\n","\n","ner_clinical = medical.NerModel.load(\"/root/cache_pretrained/ner_clinical_en_3.0.0_3.0_1617208419368\")\\\n","    .setInputCols([\"sentence\", \"token\", \"embeddings\"])\\\n","    .setOutputCol(\"ner\")\n","\n","# or just set Input and Output columns\n","# ner_clinical\\\n","#     .setInputCols([\"sentence\", \"token\", \"embeddings\"])\\\n","#     .setOutputCol(\"ner\")\n","\n","ner_converter = medical.NerConverterInternal()\\\n","    .setInputCols([\"sentence\", \"token\", \"ner\"])\\\n","    .setOutputCol(\"ner_chunk\")\n","\n","nlp_pipeline = nlp.Pipeline(\n","    stages=[\n","        document_assembler,\n","        sentence_detector,\n","        tokenizer,\n","        embeddings,\n","        ner_clinical,\n","        ner_converter\n","])\n","\n","text = [\"\"\"The human KCNJ9 (Kir 3.3, GIRK3) is a member of the G-protein-activated inwardly rectifying potassium (GIRK) channel family. Here we describe the genomicorganization of the KCNJ9 locus on chromosome 1q21-23 as a candidate gene forType II diabetes mellitus in the Pima Indian population. The gene spansapproximately 7.6 kb and contains one noncoding and two coding exons separated byapproximately 2.2 and approximately 2.6 kb introns, respectively. We identified14 single nucleotide polymorphisms (SNPs), including one that predicts aVal366Ala substitution, and an 8 base-pair (bp) insertion/deletion. Ourexpression studies revealed the presence of the transcript in various humantissues including pancreas, and two major insulin-responsive tissues: fat andskeletal muscle. The characterization of the KCNJ9 gene should facilitate furtherstudies on the function of the KCNJ9 protein and allow evaluation of thepotential role of the locus in Type II diabetes. BACKGROUND: At present, it is one of the most important issues for the treatment of breast cancer to develop the standard therapy for patients previously treated with anthracyclines and taxanes\"\"\"]\n","\n","data = spark.createDataFrame([text]).toDF(\"text\")\n","\n","result = nlp_pipeline.fit(data).transform(data)"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":585},"executionInfo":{"elapsed":4089,"status":"ok","timestamp":1695879452391,"user":{"displayName":"Sel√ßuk Meng√ºverdi","userId":"15361124602135819062"},"user_tz":240},"id":"i7OYLoSKcPhL","outputId":"08c4f8cd-fd5d-44fe-fc7c-3a177b705cd6"},"outputs":[{"output_type":"display_data","data":{"text/plain":["<IPython.core.display.HTML object>"],"text/html":["\n","<style>\n","    @import url('https://fonts.googleapis.com/css2?family=Montserrat:wght@300;400;500;600;700&display=swap');\n","    @import url('https://fonts.googleapis.com/css2?family=Vistol Regular:wght@300;400;500;600;700&display=swap');\n","    \n","    .spark-nlp-display-scroll-entities {\n","        border: 1px solid #E7EDF0;\n","        border-radius: 3px;\n","        text-align: justify;\n","        \n","    }\n","    .spark-nlp-display-scroll-entities span {  \n","        font-size: 14px;\n","        line-height: 24px;\n","        color: #536B76;\n","        font-family: 'Montserrat', sans-serif !important;\n","    }\n","    \n","    .spark-nlp-display-entity-wrapper{\n","    \n","        display: inline-grid;\n","        text-align: center;\n","        border-radius: 4px;\n","        margin: 0 2px 5px 2px;\n","        padding: 1px\n","    }\n","    .spark-nlp-display-entity-name{\n","        font-size: 14px;\n","        line-height: 24px;\n","        font-family: 'Montserrat', sans-serif !important;\n","        \n","        background: #f1f2f3;\n","        border-width: medium;\n","        text-align: center;\n","        \n","        font-weight: 400;\n","        \n","        border-radius: 5px;\n","        padding: 2px 5px;\n","        display: block;\n","        margin: 3px 2px;\n","    \n","    }\n","    .spark-nlp-display-entity-type{\n","        font-size: 14px;\n","        line-height: 24px;\n","        color: #ffffff;\n","        font-family: 'Montserrat', sans-serif !important;\n","        \n","        text-transform: uppercase;\n","        \n","        font-weight: 500;\n","\n","        display: block;\n","        padding: 3px 5px;\n","    }\n","    \n","    .spark-nlp-display-entity-resolution{\n","        font-size: 14px;\n","        line-height: 24px;\n","        color: #ffffff;\n","        font-family: 'Vistol Regular', sans-serif !important;\n","        \n","        text-transform: uppercase;\n","        \n","        font-weight: 500;\n","\n","        display: block;\n","        padding: 3px 5px;\n","    }\n","    \n","    .spark-nlp-display-others{\n","        font-size: 14px;\n","        line-height: 24px;\n","        font-family: 'Montserrat', sans-serif !important;\n","        \n","        font-weight: 400;\n","    }\n","\n","</style>\n"," <span class=\"spark-nlp-display-others\" style=\"background-color: white\">The human KCNJ9 (Kir 3.3, GIRK3) is a member of </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #8b6673\"><span class=\"spark-nlp-display-entity-name\">the G-protein-activated inwardly rectifying potassium (GIRK </span><span class=\"spark-nlp-display-entity-type\">TREATMENT</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\">) channel family. Here we describe </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #8b6673\"><span class=\"spark-nlp-display-entity-name\">the genomicorganization </span><span class=\"spark-nlp-display-entity-type\">TREATMENT</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> of the KCNJ9 locus on chromosome 1q21-23 as </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #800080\"><span class=\"spark-nlp-display-entity-name\">a candidate gene forType II diabetes mellitus </span><span class=\"spark-nlp-display-entity-type\">PROBLEM</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> in the Pima Indian population. The gene spansapproximately 7.6 kb and contains one noncoding and two coding exons separated </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #8b6673\"><span class=\"spark-nlp-display-entity-name\">byapproximately </span><span class=\"spark-nlp-display-entity-type\">TREATMENT</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> 2.2 and approximately 2.6 kb introns, respectively. We identified14 </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #8b6673\"><span class=\"spark-nlp-display-entity-name\">single nucleotide polymorphisms </span><span class=\"spark-nlp-display-entity-type\">TREATMENT</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> (SNPs), including one that predicts </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #8b6673\"><span class=\"spark-nlp-display-entity-name\">aVal366Ala substitution </span><span class=\"spark-nlp-display-entity-type\">TREATMENT</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\">, and </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #8b6673\"><span class=\"spark-nlp-display-entity-name\">an 8 base-pair </span><span class=\"spark-nlp-display-entity-type\">TREATMENT</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> (bp) </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #800080\"><span class=\"spark-nlp-display-entity-name\">insertion/deletion </span><span class=\"spark-nlp-display-entity-type\">PROBLEM</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\">. </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #77b5fe\"><span class=\"spark-nlp-display-entity-name\">Ourexpression studies </span><span class=\"spark-nlp-display-entity-type\">TEST</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> revealed the presence of </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #800080\"><span class=\"spark-nlp-display-entity-name\">the transcript in various humantissues </span><span class=\"spark-nlp-display-entity-type\">PROBLEM</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> including pancreas, and two major insulin-responsive tissues: </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #800080\"><span class=\"spark-nlp-display-entity-name\">fat andskeletal muscle </span><span class=\"spark-nlp-display-entity-type\">PROBLEM</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\">. The characterization of the KCNJ9 gene should facilitate </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #800080\"><span class=\"spark-nlp-display-entity-name\">furtherstudies </span><span class=\"spark-nlp-display-entity-type\">PROBLEM</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> on the function of </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #8b6673\"><span class=\"spark-nlp-display-entity-name\">the KCNJ9 protein </span><span class=\"spark-nlp-display-entity-type\">TREATMENT</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> and allow </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #77b5fe\"><span class=\"spark-nlp-display-entity-name\">evaluation </span><span class=\"spark-nlp-display-entity-type\">TEST</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> of thepotential role of the locus in </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #800080\"><span class=\"spark-nlp-display-entity-name\">Type II diabetes </span><span class=\"spark-nlp-display-entity-type\">PROBLEM</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\">. BACKGROUND: At present, it is one of the most important issues for </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #8b6673\"><span class=\"spark-nlp-display-entity-name\">the treatment </span><span class=\"spark-nlp-display-entity-type\">TREATMENT</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> of </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #800080\"><span class=\"spark-nlp-display-entity-name\">breast cancer </span><span class=\"spark-nlp-display-entity-type\">PROBLEM</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> to develop </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #8b6673\"><span class=\"spark-nlp-display-entity-name\">the standard therapy </span><span class=\"spark-nlp-display-entity-type\">TREATMENT</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> for patients previously treated with </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #8b6673\"><span class=\"spark-nlp-display-entity-name\">anthracyclines </span><span class=\"spark-nlp-display-entity-type\">TREATMENT</span></span><span class=\"spark-nlp-display-others\" style=\"background-color: white\"> and </span><span class=\"spark-nlp-display-entity-wrapper\" style=\"background-color: #8b6673\"><span class=\"spark-nlp-display-entity-name\">taxanes </span><span class=\"spark-nlp-display-entity-type\">TREATMENT</span></span></div>"]},"metadata":{}}],"source":["from sparknlp_display import NerVisualizer\n","\n","for i in range(len(text)):\n","  NerVisualizer().display(\n","      result = result.collect()[i],\n","      label_col = 'ner_chunk',\n","      document_col = 'document'\n","  )"]},{"cell_type":"markdown","metadata":{"id":"EN7rCdXia5U0"},"source":["### Control where model downloaded"]},{"cell_type":"markdown","metadata":{"id":"l0vVGWVXa_7o"},"source":["By setting the cache_folder_path, you can control where the downloaded resources are stored, enabling easy access and reuse of the downloaded models in subsequent operations or workflows"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"lWTuJQCIa4q7"},"outputs":[],"source":["#The first argument is the path to the zip file and the second one is the folder.\n","InternalResourceDownloader.downloadModelDirectly(\"clinical/models/ner_clinical_large_en_2.5.0_2.4_1590021302624.zip\",\n","                                                 \"clinical/models\",\n","                                                 unzip=False,\n","                                                 cache_folder_path=\"/content/models\")"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":13,"status":"ok","timestamp":1695879453307,"user":{"displayName":"Sel√ßuk Meng√ºverdi","userId":"15361124602135819062"},"user_tz":240},"id":"RaXeJgU3fJcQ","outputId":"9c75cb3e-c720-42a8-a3a7-1e2acf3dfb6c"},"outputs":[{"output_type":"stream","name":"stdout","text":["/content/models\n"]}],"source":["cd models"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":179,"status":"ok","timestamp":1695879453478,"user":{"displayName":"Sel√ßuk Meng√ºverdi","userId":"15361124602135819062"},"user_tz":240},"id":"J67lUYkFfKv9","outputId":"4effbc88-54db-42a5-94a0-15730c0c05de"},"outputs":[{"output_type":"stream","name":"stdout","text":["ner_clinical_large_en_2.5.0_2.4_1590021302624.zip\n"]}],"source":["ls"]},{"cell_type":"markdown","metadata":{"id":"2c-5uGzVsS3b"},"source":["### returnPrivateModels\n","\n","    Return private models available for download.\n","\n","    parameter:\n","        annotator : The annotator to filter by. Defaults to None.\n","        lang      : The language to filter by. Defaults to None.\n","        version   : The version to filter by. Defaults to None."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":801,"status":"ok","timestamp":1695879454277,"user":{"displayName":"Sel√ßuk Meng√ºverdi","userId":"15361124602135819062"},"user_tz":240},"id":"nhNP7NMvsSJl","outputId":"ed23a45c-1258-44b6-a431-48c47a9fb4ab"},"outputs":[{"output_type":"execute_result","data":{"text/plain":["[['nerdl_tumour_demo', 'en', '1.7.3'],\n"," ['nerdl_tumour_demo', 'en', '1.8.0'],\n"," ['nerdl_tumour_demo', 'en', '2.0.2'],\n"," ['ner_healthcare', 'en', '2.4.4'],\n"," ['ner_radiology', 'en', '2.7.0'],\n"," ['ner_deid_augmented', 'en', '2.7.1'],\n"," ['ner_deidentify_dl', 'en', '2.7.2'],\n"," ['ner_events_admission_clinical', 'en', '2.7.4'],\n"," ['ner_clinical', 'en', '3.0.0'],\n"," ['ner_radiology', 'en', '3.0.0'],\n"," ['ner_bionlp', 'en', '3.0.0'],\n"," ['ner_posology', 'en', '3.0.0'],\n"," ['ner_deid_augmented', 'en', '3.0.0'],\n"," ['ner_anatomy', 'en', '3.0.0'],\n"," ['ner_risk_factors', 'en', '3.0.0'],\n"," ['ner_chemprot_clinical', 'en', '3.0.0'],\n"," ['ner_posology_small', 'en', '3.0.0'],\n"," ['ner_posology_greedy', 'en', '3.0.0'],\n"," ['ner_deid_enriched', 'en', '3.0.0'],\n"," ['ner_drugs_greedy', 'en', '3.0.0'],\n"," ['jsl_ner_wip_clinical', 'en', '3.0.0'],\n"," ['ner_posology_large', 'en', '3.0.0'],\n"," ['jsl_ner_wip_greedy_clinical', 'en', '3.0.0'],\n"," ['ner_clinical_large', 'en', '3.0.0'],\n"," ['ner_diseases', 'en', '3.0.0'],\n"," ['ner_aspect_based_sentiment', 'en', '3.0.0'],\n"," ['ner_cellular', 'en', '3.0.0'],\n"," ['ner_drugs', 'en', '3.0.0'],\n"," ['ner_jsl', 'en', '3.0.0'],\n"," ['ner_anatomy_coarse_biobert', 'en', '3.0.0'],\n"," ['ner_cancer_genetics', 'en', '3.0.0'],\n"," ['ner_drugs_large', 'en', '3.0.0'],\n"," ['ner_events_admission_clinical', 'en', '3.0.0'],\n"," ['ner_deidentify_dl', 'en', '3.0.0'],\n"," ['ner_deid_synthetic', 'en', '3.0.0'],\n"," ['ner_jsl_enriched', 'en', '3.0.0'],\n"," ['ner_deid_large', 'en', '3.0.0'],\n"," ['ner_human_phenotype_gene_clinical', 'en', '3.0.0'],\n"," ['ner_human_phenotype_go_clinical', 'en', '3.0.0'],\n"," ['ner_events_clinical', 'en', '3.0.0'],\n"," ['ner_anatomy_coarse', 'en', '3.0.0'],\n"," ['ner_bionlp_biobert', 'en', '3.0.0'],\n"," ['ner_deid_sd_large', 'en', '3.0.0'],\n"," ['ner_risk_factors_biobert', 'en', '3.0.0'],\n"," ['ner_posology_healthcare', 'en', '3.0.0'],\n"," ['ner_diseases_large', 'en', '3.0.0'],\n"," ['ner_ade_biobert', 'en', '3.0.0'],\n"," ['ner_jsl_enriched_biobert', 'en', '3.0.0'],\n"," ['ner_ade_healthcare', 'en', '3.0.0'],\n"," ['ner_deid_sd', 'en', '3.0.0'],\n"," ['ner_events_healthcare', 'en', '3.0.0'],\n"," ['ner_financial_contract', 'en', '3.0.0'],\n"," ['ner_chemprot_biobert', 'en', '3.0.0'],\n"," ['ner_ade_clinicalbert', 'en', '3.0.0'],\n"," ['ner_posology_large_biobert', 'en', '3.0.0'],\n"," ['ner_radiology_wip_clinical', 'en', '3.0.0'],\n"," ['ner_jsl_biobert', 'en', '3.0.0'],\n"," ['ner_clinical_biobert', 'en', '3.0.0'],\n"," ['ner_deid_enriched_biobert', 'en', '3.0.0'],\n"," ['ner_posology_biobert', 'en', '3.0.0'],\n"," ['ner_cellular_biobert', 'en', '3.0.0'],\n"," ['jsl_ner_wip_modifier_clinical', 'en', '3.0.0'],\n"," ['ner_measurements_clinical', 'en', '3.0.0'],\n"," ['ner_medmentions_coarse', 'en', '3.0.0'],\n"," ['ner_chemicals', 'en', '3.0.0'],\n"," ['ner_events_biobert', 'en', '3.0.0'],\n"," ['ner_bacterial_species', 'en', '3.0.0'],\n"," ['ner_diseases_biobert', 'en', '3.0.0'],\n"," ['ner_human_phenotype_gene_biobert', 'en', '3.0.0'],\n"," ['ner_deid_biobert', 'en', '3.0.0'],\n"," ['nerdl_tumour_demo', 'en', '3.0.0'],\n"," ['ner_human_phenotype_go_biobert', 'en', '3.0.0'],\n"," ['ner_anatomy_biobert', 'en', '3.0.0'],\n"," ['ner_ade_clinical', 'en', '3.0.0'],\n"," ['jsl_rd_ner_wip_greedy_clinical', 'en', '3.0.0'],\n"," ['ner_healthcare', 'en', '3.0.0'],\n"," ['ner_deid_generic_augmented', 'en', '3.0.3'],\n"," ['ner_deid_subentity_augmented', 'en', '3.0.3'],\n"," ['ner_deid_generic_glove', 'en', '3.0.4'],\n"," ['ner_deid_subentity_glove', 'en', '3.0.4'],\n"," ['ner_jsl', 'en', '3.1.0'],\n"," ['ner_genetic_variants', 'en', '3.1.0'],\n"," ['ner_deid_subentity_augmented', 'en', '3.1.0'],\n"," ['ner_deid_generic_augmented', 'en', '3.1.0'],\n"," ['jsl_ner_wip_greedy_biobert', 'en', '3.1.3'],\n"," ['jsl_rd_ner_wip_greedy_biobert', 'en', '3.1.3'],\n"," ['ner_jsl_greedy_biobert', 'en', '3.2.0'],\n"," ['ner_jsl_slim', 'en', '3.2.0'],\n"," ['ner_posology_experimental', 'en', '3.1.3'],\n"," ['ner_deid_subentity_augmented', 'en', '3.2.0'],\n"," ['ner_jsl_biobert', 'en', '3.2.0'],\n"," ['ner_chexpert', 'en', '3.3.0'],\n"," ['ner_jsl_enriched', 'en', '3.3.0'],\n"," ['ner_covid_trials', 'en', '3.3.2'],\n"," ['ner_chemd_clinical', 'en', '3.3.0'],\n"," ['ner_nihss', 'en', '3.3.2'],\n"," ['ner_biomarker', 'en', '3.3.3'],\n"," ['ner_deid_subentity_augmented_i2b2', 'en', '3.3.2'],\n"," ['ner_drugprot_clinical', 'en', '3.3.3'],\n"," ['ner_abbreviation_clinical', 'en', '3.3.4'],\n"," ['ner_supplement_clinical', 'en', '3.3.4'],\n"," ['ner_nature_nero_clinical', 'en', '3.3.4'],\n"," ['ner_biomedical_bc2gm', 'en', '3.5.1'],\n"," ['ner_living_species', 'en', '3.5.3'],\n"," ['ner_living_species_biobert', 'en', '3.5.3'],\n"," ['ner_clinical_trials_abstracts', 'en', '3.5.3'],\n"," ['ner_pathogen', 'en', '4.0.0'],\n"," ['ner_oncology_wip', 'en', '3.5.0'],\n"," ['ner_ade_binary', 'en', '4.0.0'],\n"," ['ner_oncology_unspecific_posology_wip', 'en', '4.0.0'],\n"," ['ner_oncology_wip', 'en', '4.0.0'],\n"," ['ner_oncology_therapy_wip', 'en', '4.0.0'],\n"," ['ner_oncology_posology_wip', 'en', '4.0.0'],\n"," ['ner_oncology_anatomy_general_wip', 'en', '4.0.0'],\n"," ['ner_oncology_tnm_wip', 'en', '4.0.0'],\n"," ['ner_oncology_demographics_wip', 'en', '4.0.0'],\n"," ['ner_oncology_biomarker_wip', 'en', '4.0.0'],\n"," ['ner_oncology_anatomy_granular_wip', 'en', '4.0.0'],\n"," ['ner_oncology_test_wip', 'en', '4.0.0'],\n"," ['ner_oncology_diagnosis_wip', 'en', '4.0.0'],\n"," ['ner_oncology_response_to_treatment_wip', 'en', '4.0.0'],\n"," ['ner_oncology_unspecific_posology', 'en', '4.0.0'],\n"," ['ner_oncology', 'en', '4.0.0'],\n"," ['ner_oncology_tnm', 'en', '4.0.0'],\n"," ['ner_oncology_anatomy_general', 'en', '4.0.0'],\n"," ['ner_oncology_therapy', 'en', '4.0.0'],\n"," ['ner_oncology_test', 'en', '4.0.0'],\n"," ['ner_oncology_diagnosis', 'en', '4.0.0'],\n"," ['ner_oncology_demographics', 'en', '4.0.0'],\n"," ['ner_oncology_anatomy_granular', 'en', '4.0.0'],\n"," ['ner_oncology_response_to_treatment', 'en', '4.0.0'],\n"," ['ner_oncology_posology', 'en', '4.0.0'],\n"," ['ner_oncology_biomarker', 'en', '4.0.0']]"]},"metadata":{},"execution_count":34}],"source":["ner_models = InternalResourceDownloader.returnPrivateModels(\"MedicalNerModel\",\"en\", \"4.0.0\")\n","ner_models"]},{"cell_type":"markdown","metadata":{"id":"baFJoHJCYqJQ"},"source":["# UpdateModels"]},{"cell_type":"markdown","metadata":{"id":"2dDbrWe7YsUI"},"source":["## updateCacheModels\n","\n","    Refreshes all pretrained models located in the cache pretrained folder.\n","    Checks the existing models in the cache pretrained folder and if there is are new\n","    version for each model. If there is a new version, it will be downloaded and\n","    overwrite the existing one.\n","\n","    Parameters:\n","        cache_folder : Path where the models will be refreshed. i.e (\"hdfs:..\",\"file:...\")"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"uXjQ4LgAdwoL"},"outputs":[],"source":["# clear cache folder\n","!rm -rf /root/cache_pretrained"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"DfCHZDUab3ij"},"outputs":[],"source":["from sparknlp.pretrained import ResourceDownloader\n","\n","#The first argument is the path to the zip file and the second one is the folder.\n","ResourceDownloader.downloadModelDirectly(\"clinical/models/embeddings_clinical_en_2.0.2_2.4_1558454742956.zip\", \"clinical/models\")\n","ResourceDownloader.downloadModelDirectly(\"clinical/models/ner_clinical_large_en_2.5.0_2.4_1590021302624.zip\", \"clinical/models\")"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":154,"status":"ok","timestamp":1695879539045,"user":{"displayName":"Sel√ßuk Meng√ºverdi","userId":"15361124602135819062"},"user_tz":240},"id":"GpPXlj6ib457","outputId":"3da43c9d-87ab-49ce-8837-6f59f336099b"},"outputs":[{"output_type":"stream","name":"stdout","text":["\u001b[0m\u001b[01;34membeddings_clinical_en_2.0.2_2.4_1558454742956\u001b[0m/\n","\u001b[01;34mner_clinical_large_en_2.5.0_2.4_1590021302624\u001b[0m/\n"]}],"source":["ls ~/cache_pretrained"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"AS1h72wYb6cE"},"outputs":[],"source":["from sparknlp_jsl.updateModels import UpdateModels\n","UpdateModels.updateCacheModels()"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":169,"status":"ok","timestamp":1695879617600,"user":{"displayName":"Sel√ßuk Meng√ºverdi","userId":"15361124602135819062"},"user_tz":240},"id":"1f_B8qAAb7wU","outputId":"9bbf4327-d5a6-4b0e-a452-bc66a202c19e"},"outputs":[{"output_type":"stream","name":"stdout","text":["\u001b[0m\u001b[01;34membeddings_clinical_en_2.0.2_2.4_1558454742956\u001b[0m/\n","\u001b[01;34membeddings_clinical_en_2.4.0_2.4_1580237286004\u001b[0m/\n","\u001b[01;34mner_clinical_large_en_2.5.0_2.4_1590021302624\u001b[0m/\n","\u001b[01;34mner_clinical_large_en_3.0.0_3.0_1617206114650\u001b[0m/\n"]}],"source":["ls ~/cache_pretrained"]},{"cell_type":"markdown","metadata":{"id":"op4cZM91faaA"},"source":["## updateModels\n","\n","    downloads all the new pretrained models that have been released since the specified date interval.\n","\n","    parameter:\n","       model_names: A list of names of the models to be downloaded.\n","       language: The language of the models, with a default value of \"en\".\n","       start_date: The starting date used to filter the models, in the format \"yyyy-MM-dd\".\n","       end_date: The ending date used to filter the models, in the format \"yyyy-MM-dd\".\n","       cache_folder: The path indicating where the models will be downloaded and stored."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"P4O6RTh3f_sr"},"outputs":[],"source":["UpdateModels.updateModels(start_date = \"2021-01-01\",\n","                          end_date = \"2023-07-07\",\n","                          model_names=[\"ner_clinical\",\"ner_jsl\"],\n","                          language=\"en\",\n","                          remote_loc=\"clinical/models\",\n","                          cache_folder=\"/content/jsl_models\"\n","                          )"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":186,"status":"ok","timestamp":1695879620509,"user":{"displayName":"Sel√ßuk Meng√ºverdi","userId":"15361124602135819062"},"user_tz":240},"id":"uo2Qo2rzg9W4","outputId":"abde6bbb-ca95-4cf9-f816-24381c4e129c"},"outputs":[{"output_type":"stream","name":"stdout","text":["\u001b[0m\u001b[01;34mner_clinical_en_3.0.0_3.0_1617208419368\u001b[0m/  \u001b[01;34mner_jsl_en_4.2.0_3.0_1666181370373\u001b[0m/\n"]}],"source":["ls /content/jsl_models"]}],"metadata":{"colab":{"provenance":[]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python"}},"nbformat":4,"nbformat_minor":0}