{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "360RD6-wzNiR"
   },
   "source": [
    "![JohnSnowLabs](https://nlp.johnsnowlabs.com/assets/images/logo.png)\n",
    "\n",
    "[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/JohnSnowLabs/nlu/blob/master/examples/colab/healthcare/assertion/assertion_overview.ipynb)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "uv_9xSABoHKk"
   },
   "source": [
    "# Assert statuses of medical entities \n",
    "\n",
    "`Assert` for each `entity` the status into one out of `C` classes.  These classes usually are : ``hypothetical``, ``present``, ``absent``, ``possible``, ``conditional``, ``associated_with_someone_else``.\n",
    "\n",
    "\n",
    "All the models avaiable are :\n",
    "\n",
    "| Language | nlu.load() reference                                         | Spark NLP Model reference          |\n",
    "| -------- | ------------------------------------------------------------ | ------------------------------------------------------------ |\n",
    "| English  | [assert](https://nlp.johnsnowlabs.com/2021/01/26/assertion_dl_en.html) | [assertion_dl](https://nlp.johnsnowlabs.com/2021/01/26/assertion_dl_en.html)                   |\n",
    "| English  | [assert.biobert](https://nlp.johnsnowlabs.com/2021/01/26/assertion_dl_biobert_en.html) | [assertion_dl_biobert](https://nlp.johnsnowlabs.com/2021/01/26/assertion_dl_biobert_en.html)                   |\n",
    "| English  | [assert.healthcare](https://nlp.johnsnowlabs.com/2020/09/23/assertion_dl_healthcare_en.html) | [assertion_dl_healthcare](https://nlp.johnsnowlabs.com/2020/09/23/assertion_dl_healthcare_en.html)                   |\n",
    "| English  | [assert.large](https://nlp.johnsnowlabs.com/2020/05/21/assertion_dl_large_en.html) | [assertion_dl_large](https://nlp.johnsnowlabs.com/2020/05/21/assertion_dl_large_en.html)                   |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 264038,
     "status": "ok",
     "timestamp": 1649994526620,
     "user": {
      "displayName": "ahmed lone",
      "userId": "02458088882398909889"
     },
     "user_tz": -300
    },
    "id": "OMiHjCtQhF_r",
    "outputId": "fbea9408-414a-4ac2-8cb9-72855ee8849c"
   },
   "outputs": [],
   "source": [
    "%%capture\n",
    "!wget https://setup.johnsnowlabs.com/nlu/colab.sh -O - | bash\n",
    "\n",
    "import nlu\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "rSUTBWKi6Hb-"
   },
   "source": [
    "#### [Detect Assertion Status (assertion_dl) - supports confidence scores.](https://nlp.johnsnowlabs.com/2021/01/26/assertion_dl_en.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 998
    },
    "executionInfo": {
     "elapsed": 186066,
     "status": "ok",
     "timestamp": 1649994712670,
     "user": {
      "displayName": "ahmed lone",
      "userId": "02458088882398909889"
     },
     "user_tz": -300
    },
    "id": "emvKNnXE6HBv",
    "outputId": "6ed81aed-5e4e-44e8-bcd9-c3324f76509b"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ner_clinical download started this may take some time.\n",
      "Approximate size to download 13.9 MB\n",
      "[OK!]\n",
      "assertion_dl download started this may take some time.\n",
      "Approximate size to download 1.3 MB\n",
      "[OK!]\n",
      "embeddings_clinical download started this may take some time.\n",
      "Approximate size to download 1.6 GB\n",
      "[OK!]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "  <div id=\"df-4f6f0a65-292e-4041-a02c-160eb1e30b4f\">\n",
       "    <div class=\"colab-df-container\">\n",
       "      <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assertion</th>\n",
       "      <th>document</th>\n",
       "      <th>entities_clinical</th>\n",
       "      <th>entities_clinical_class</th>\n",
       "      <th>entities_clinical_confidence</th>\n",
       "      <th>word_embedding_glove</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>conditional</td>\n",
       "      <td>The human KCNJ9 (Kir 3.3, GIRK3) is a member o...</td>\n",
       "      <td>the G-protein-activated inwardly rectifying po...</td>\n",
       "      <td>TREATMENT</td>\n",
       "      <td>0.77290004</td>\n",
       "      <td>[[-0.0823686420917511, -0.3178570866584778, 0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>present</td>\n",
       "      <td>Here we describe the genomicorganization of th...</td>\n",
       "      <td>the genomicorganization</td>\n",
       "      <td>TREATMENT</td>\n",
       "      <td>0.96985</td>\n",
       "      <td>[[0.40810859203338623, -0.34884893894195557, 0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>present</td>\n",
       "      <td>Here we describe the genomicorganization of th...</td>\n",
       "      <td>a candidate gene forType II diabetes mellitus</td>\n",
       "      <td>PROBLEM</td>\n",
       "      <td>0.7840571</td>\n",
       "      <td>[[0.40810859203338623, -0.34884893894195557, 0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>present</td>\n",
       "      <td>The gene spansapproximately 7.6 kb and contain...</td>\n",
       "      <td>byapproximately</td>\n",
       "      <td>TREATMENT</td>\n",
       "      <td>0.9692</td>\n",
       "      <td>[[-0.0823686420917511, -0.3178570866584778, 0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>present</td>\n",
       "      <td>We identified14 single nucleotide polymorphism...</td>\n",
       "      <td>single nucleotide polymorphisms</td>\n",
       "      <td>TREATMENT</td>\n",
       "      <td>0.9097667</td>\n",
       "      <td>[[0.2036844938993454, -0.39434725046157837, 0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>present</td>\n",
       "      <td>We identified14 single nucleotide polymorphism...</td>\n",
       "      <td>aVal366Ala substitution</td>\n",
       "      <td>PROBLEM</td>\n",
       "      <td>0.77129996</td>\n",
       "      <td>[[0.2036844938993454, -0.39434725046157837, 0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>present</td>\n",
       "      <td>We identified14 single nucleotide polymorphism...</td>\n",
       "      <td>an 8 base-pair</td>\n",
       "      <td>PROBLEM</td>\n",
       "      <td>0.5048</td>\n",
       "      <td>[[0.2036844938993454, -0.39434725046157837, 0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>present</td>\n",
       "      <td>(bp) insertion/deletion.</td>\n",
       "      <td>insertion/deletion</td>\n",
       "      <td>PROBLEM</td>\n",
       "      <td>0.9999</td>\n",
       "      <td>[[-0.058002080768346786, 0.14185668528079987, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>present</td>\n",
       "      <td>Ourexpression studies revealed the presence of...</td>\n",
       "      <td>Ourexpression studies</td>\n",
       "      <td>TEST</td>\n",
       "      <td>0.8843</td>\n",
       "      <td>[[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>present</td>\n",
       "      <td>Ourexpression studies revealed the presence of...</td>\n",
       "      <td>the transcript in various humantissues</td>\n",
       "      <td>PROBLEM</td>\n",
       "      <td>0.92882</td>\n",
       "      <td>[[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>possible</td>\n",
       "      <td>Ourexpression studies revealed the presence of...</td>\n",
       "      <td>fat andskeletal muscle</td>\n",
       "      <td>PROBLEM</td>\n",
       "      <td>0.8903666</td>\n",
       "      <td>[[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>present</td>\n",
       "      <td>The characterization of the KCNJ9 gene should ...</td>\n",
       "      <td>furtherstudies</td>\n",
       "      <td>PROBLEM</td>\n",
       "      <td>0.9981</td>\n",
       "      <td>[[-0.0823686420917511, -0.3178570866584778, 0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>present</td>\n",
       "      <td>The characterization of the KCNJ9 gene should ...</td>\n",
       "      <td>the KCNJ9 protein</td>\n",
       "      <td>TREATMENT</td>\n",
       "      <td>0.7142</td>\n",
       "      <td>[[-0.0823686420917511, -0.3178570866584778, 0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>possible</td>\n",
       "      <td>The characterization of the KCNJ9 gene should ...</td>\n",
       "      <td>evaluation</td>\n",
       "      <td>TEST</td>\n",
       "      <td>0.9987</td>\n",
       "      <td>[[-0.0823686420917511, -0.3178570866584778, 0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>present</td>\n",
       "      <td>The characterization of the KCNJ9 gene should ...</td>\n",
       "      <td>Type II diabetes</td>\n",
       "      <td>PROBLEM</td>\n",
       "      <td>0.9191</td>\n",
       "      <td>[[-0.0823686420917511, -0.3178570866584778, 0....</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-4f6f0a65-292e-4041-a02c-160eb1e30b4f')\"\n",
       "              title=\"Convert this dataframe to an interactive table.\"\n",
       "              style=\"display:none;\">\n",
       "        \n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "       width=\"24px\">\n",
       "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
       "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
       "  </svg>\n",
       "      </button>\n",
       "      \n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      flex-wrap:wrap;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "      <script>\n",
       "        const buttonEl =\n",
       "          document.querySelector('#df-4f6f0a65-292e-4041-a02c-160eb1e30b4f button.colab-df-convert');\n",
       "        buttonEl.style.display =\n",
       "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "        async function convertToInteractive(key) {\n",
       "          const element = document.querySelector('#df-4f6f0a65-292e-4041-a02c-160eb1e30b4f');\n",
       "          const dataTable =\n",
       "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                     [key], {});\n",
       "          if (!dataTable) return;\n",
       "\n",
       "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "            + ' to learn more about interactive tables.';\n",
       "          element.innerHTML = '';\n",
       "          dataTable['output_type'] = 'display_data';\n",
       "          await google.colab.output.renderOutput(dataTable, element);\n",
       "          const docLink = document.createElement('div');\n",
       "          docLink.innerHTML = docLinkHtml;\n",
       "          element.appendChild(docLink);\n",
       "        }\n",
       "      </script>\n",
       "    </div>\n",
       "  </div>\n",
       "  "
      ],
      "text/plain": [
       "     assertion                                           document  \\\n",
       "0  conditional  The human KCNJ9 (Kir 3.3, GIRK3) is a member o...   \n",
       "1      present  Here we describe the genomicorganization of th...   \n",
       "1      present  Here we describe the genomicorganization of th...   \n",
       "2      present  The gene spansapproximately 7.6 kb and contain...   \n",
       "3      present  We identified14 single nucleotide polymorphism...   \n",
       "3      present  We identified14 single nucleotide polymorphism...   \n",
       "3      present  We identified14 single nucleotide polymorphism...   \n",
       "4      present                           (bp) insertion/deletion.   \n",
       "5      present  Ourexpression studies revealed the presence of...   \n",
       "5      present  Ourexpression studies revealed the presence of...   \n",
       "5     possible  Ourexpression studies revealed the presence of...   \n",
       "6      present  The characterization of the KCNJ9 gene should ...   \n",
       "6      present  The characterization of the KCNJ9 gene should ...   \n",
       "6     possible  The characterization of the KCNJ9 gene should ...   \n",
       "6      present  The characterization of the KCNJ9 gene should ...   \n",
       "\n",
       "                                   entities_clinical entities_clinical_class  \\\n",
       "0  the G-protein-activated inwardly rectifying po...               TREATMENT   \n",
       "1                            the genomicorganization               TREATMENT   \n",
       "1      a candidate gene forType II diabetes mellitus                 PROBLEM   \n",
       "2                                    byapproximately               TREATMENT   \n",
       "3                    single nucleotide polymorphisms               TREATMENT   \n",
       "3                            aVal366Ala substitution                 PROBLEM   \n",
       "3                                     an 8 base-pair                 PROBLEM   \n",
       "4                                 insertion/deletion                 PROBLEM   \n",
       "5                              Ourexpression studies                    TEST   \n",
       "5             the transcript in various humantissues                 PROBLEM   \n",
       "5                             fat andskeletal muscle                 PROBLEM   \n",
       "6                                     furtherstudies                 PROBLEM   \n",
       "6                                  the KCNJ9 protein               TREATMENT   \n",
       "6                                         evaluation                    TEST   \n",
       "6                                   Type II diabetes                 PROBLEM   \n",
       "\n",
       "  entities_clinical_confidence  \\\n",
       "0                   0.77290004   \n",
       "1                      0.96985   \n",
       "1                    0.7840571   \n",
       "2                       0.9692   \n",
       "3                    0.9097667   \n",
       "3                   0.77129996   \n",
       "3                       0.5048   \n",
       "4                       0.9999   \n",
       "5                       0.8843   \n",
       "5                      0.92882   \n",
       "5                    0.8903666   \n",
       "6                       0.9981   \n",
       "6                       0.7142   \n",
       "6                       0.9987   \n",
       "6                       0.9191   \n",
       "\n",
       "                                word_embedding_glove  \n",
       "0  [[-0.0823686420917511, -0.3178570866584778, 0....  \n",
       "1  [[0.40810859203338623, -0.34884893894195557, 0...  \n",
       "1  [[0.40810859203338623, -0.34884893894195557, 0...  \n",
       "2  [[-0.0823686420917511, -0.3178570866584778, 0....  \n",
       "3  [[0.2036844938993454, -0.39434725046157837, 0....  \n",
       "3  [[0.2036844938993454, -0.39434725046157837, 0....  \n",
       "3  [[0.2036844938993454, -0.39434725046157837, 0....  \n",
       "4  [[-0.058002080768346786, 0.14185668528079987, ...  \n",
       "5  [[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...  \n",
       "5  [[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...  \n",
       "5  [[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...  \n",
       "6  [[-0.0823686420917511, -0.3178570866584778, 0....  \n",
       "6  [[-0.0823686420917511, -0.3178570866584778, 0....  \n",
       "6  [[-0.0823686420917511, -0.3178570866584778, 0....  \n",
       "6  [[-0.0823686420917511, -0.3178570866584778, 0....  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import nlu\n",
    "\n",
    "nlu.load(\"en.med_ner.clinical en.assert\").predict(['The human KCNJ9 (Kir 3.3, GIRK3) is a member of the G-protein-activated inwardly rectifying potassium (GIRK) channel family.',\n",
    " 'Here we describe the genomicorganization of the KCNJ9 locus on chromosome 1q21-23 as a candidate gene forType II diabetes mellitus in the Pima Indian population.',\n",
    "  'The gene spansapproximately 7.6 kb and contains one noncoding and two coding exons separated byapproximately 2.2 and approximately 2.6 kb introns, respectively.',\n",
    "   'We identified14 single nucleotide polymorphisms (SNPs), including one that predicts aVal366Ala substitution, and an 8 base-pair', \n",
    "   '(bp) insertion/deletion.', 'Ourexpression studies revealed the presence of the transcript in various humantissues including pancreas, and two major insulin-responsive tissues: fat andskeletal muscle.', \n",
    "   'The characterization of the KCNJ9 gene should facilitate furtherstudies on the function of the KCNJ9 protein and allow evaluation of thepotential role of the locus in Type II diabetes.'],output_level =  \"chunk\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "OXTlByOz6TLg"
   },
   "source": [
    "#### [Detect Assertion Status (assertion_dl_biobert) - supports confidence scores.](https://nlp.johnsnowlabs.com/2021/01/26/assertion_dl_biobert_en.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "executionInfo": {
     "elapsed": 74593,
     "status": "ok",
     "timestamp": 1649994787225,
     "user": {
      "displayName": "ahmed lone",
      "userId": "02458088882398909889"
     },
     "user_tz": -300
    },
    "id": "EQUUl_Sk6TLh",
    "outputId": "19bfb4bb-2bba-4904-ec21-9c79068ffb2c"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ner_clinical_biobert download started this may take some time.\n",
      "Approximate size to download 15.5 MB\n",
      "[OK!]\n",
      "assertion_dl_biobert download started this may take some time.\n",
      "Approximate size to download 3 MB\n",
      "[OK!]\n",
      "biobert_pubmed_base_cased download started this may take some time.\n",
      "Approximate size to download 386.4 MB\n",
      "[OK!]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "  <div id=\"df-4df68e2e-ea3e-4e29-91e2-d25d3878fb22\">\n",
       "    <div class=\"colab-df-container\">\n",
       "      <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assertion</th>\n",
       "      <th>document</th>\n",
       "      <th>entities_clinical</th>\n",
       "      <th>entities_clinical_class</th>\n",
       "      <th>entities_clinical_confidence</th>\n",
       "      <th>word_embedding_biobert</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>present</td>\n",
       "      <td>The human KCNJ9 (Kir 3.3, GIRK3) is a member o...</td>\n",
       "      <td>The human KCNJ9 (Kir</td>\n",
       "      <td>TEST</td>\n",
       "      <td>0.56672</td>\n",
       "      <td>[[-0.248008593916893, 0.17330053448677063, 0.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>present</td>\n",
       "      <td>The human KCNJ9 (Kir 3.3, GIRK3) is a member o...</td>\n",
       "      <td>the G-protein-activated inwardly rectifying po...</td>\n",
       "      <td>PROBLEM</td>\n",
       "      <td>0.56248</td>\n",
       "      <td>[[-0.248008593916893, 0.17330053448677063, 0.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>present</td>\n",
       "      <td>Here we describe the genomicorganization of th...</td>\n",
       "      <td>the KCNJ9 locus</td>\n",
       "      <td>TEST</td>\n",
       "      <td>0.43453333</td>\n",
       "      <td>[[0.01339302584528923, 0.005851167719811201, 0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>associated_with_someone_else</td>\n",
       "      <td>Here we describe the genomicorganization of th...</td>\n",
       "      <td>a candidate gene forType II diabetes mellitus</td>\n",
       "      <td>PROBLEM</td>\n",
       "      <td>0.7657142</td>\n",
       "      <td>[[0.01339302584528923, 0.005851167719811201, 0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>The gene spansapproximately 7.6 kb and contain...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[[-0.542584240436554, 0.050951749086380005, -0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>present</td>\n",
       "      <td>We identified14 single nucleotide polymorphism...</td>\n",
       "      <td>single nucleotide polymorphisms (SNPs),</td>\n",
       "      <td>PROBLEM</td>\n",
       "      <td>0.56495005</td>\n",
       "      <td>[[0.08011732995510101, -0.11200879514217377, -...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>present</td>\n",
       "      <td>We identified14 single nucleotide polymorphism...</td>\n",
       "      <td>aVal366Ala substitution</td>\n",
       "      <td>PROBLEM</td>\n",
       "      <td>0.66485</td>\n",
       "      <td>[[0.08011732995510101, -0.11200879514217377, -...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>present</td>\n",
       "      <td>We identified14 single nucleotide polymorphism...</td>\n",
       "      <td>base-pair</td>\n",
       "      <td>TEST</td>\n",
       "      <td>0.5886</td>\n",
       "      <td>[[0.08011732995510101, -0.11200879514217377, -...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>(bp) insertion/deletion.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[[-0.025541141629219055, -0.29619351029396057,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>present</td>\n",
       "      <td>Ourexpression studies revealed the presence of...</td>\n",
       "      <td>Ourexpression studies</td>\n",
       "      <td>TEST</td>\n",
       "      <td>0.9911</td>\n",
       "      <td>[[0.4464523494243622, -0.1416597217321396, 0.5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>present</td>\n",
       "      <td>The characterization of the KCNJ9 gene should ...</td>\n",
       "      <td>The characterization of the KCNJ9 gene</td>\n",
       "      <td>TEST</td>\n",
       "      <td>0.65316665</td>\n",
       "      <td>[[-0.5280818939208984, 0.03617645055055618, 0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>present</td>\n",
       "      <td>The characterization of the KCNJ9 gene should ...</td>\n",
       "      <td>furtherstudies on the function of the KCNJ9 pr...</td>\n",
       "      <td>TEST</td>\n",
       "      <td>0.77554995</td>\n",
       "      <td>[[-0.5280818939208984, 0.03617645055055618, 0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>associated_with_someone_else</td>\n",
       "      <td>The characterization of the KCNJ9 gene should ...</td>\n",
       "      <td>evaluation</td>\n",
       "      <td>TEST</td>\n",
       "      <td>0.7549</td>\n",
       "      <td>[[-0.5280818939208984, 0.03617645055055618, 0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>associated_with_someone_else</td>\n",
       "      <td>The characterization of the KCNJ9 gene should ...</td>\n",
       "      <td>Type II diabetes</td>\n",
       "      <td>PROBLEM</td>\n",
       "      <td>0.9324</td>\n",
       "      <td>[[-0.5280818939208984, 0.03617645055055618, 0....</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-4df68e2e-ea3e-4e29-91e2-d25d3878fb22')\"\n",
       "              title=\"Convert this dataframe to an interactive table.\"\n",
       "              style=\"display:none;\">\n",
       "        \n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "       width=\"24px\">\n",
       "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
       "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
       "  </svg>\n",
       "      </button>\n",
       "      \n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      flex-wrap:wrap;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "      <script>\n",
       "        const buttonEl =\n",
       "          document.querySelector('#df-4df68e2e-ea3e-4e29-91e2-d25d3878fb22 button.colab-df-convert');\n",
       "        buttonEl.style.display =\n",
       "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "        async function convertToInteractive(key) {\n",
       "          const element = document.querySelector('#df-4df68e2e-ea3e-4e29-91e2-d25d3878fb22');\n",
       "          const dataTable =\n",
       "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                     [key], {});\n",
       "          if (!dataTable) return;\n",
       "\n",
       "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "            + ' to learn more about interactive tables.';\n",
       "          element.innerHTML = '';\n",
       "          dataTable['output_type'] = 'display_data';\n",
       "          await google.colab.output.renderOutput(dataTable, element);\n",
       "          const docLink = document.createElement('div');\n",
       "          docLink.innerHTML = docLinkHtml;\n",
       "          element.appendChild(docLink);\n",
       "        }\n",
       "      </script>\n",
       "    </div>\n",
       "  </div>\n",
       "  "
      ],
      "text/plain": [
       "                      assertion  \\\n",
       "0                       present   \n",
       "0                       present   \n",
       "1                       present   \n",
       "1  associated_with_someone_else   \n",
       "2                           NaN   \n",
       "3                       present   \n",
       "3                       present   \n",
       "3                       present   \n",
       "4                           NaN   \n",
       "5                       present   \n",
       "6                       present   \n",
       "6                       present   \n",
       "6  associated_with_someone_else   \n",
       "6  associated_with_someone_else   \n",
       "\n",
       "                                            document  \\\n",
       "0  The human KCNJ9 (Kir 3.3, GIRK3) is a member o...   \n",
       "0  The human KCNJ9 (Kir 3.3, GIRK3) is a member o...   \n",
       "1  Here we describe the genomicorganization of th...   \n",
       "1  Here we describe the genomicorganization of th...   \n",
       "2  The gene spansapproximately 7.6 kb and contain...   \n",
       "3  We identified14 single nucleotide polymorphism...   \n",
       "3  We identified14 single nucleotide polymorphism...   \n",
       "3  We identified14 single nucleotide polymorphism...   \n",
       "4                           (bp) insertion/deletion.   \n",
       "5  Ourexpression studies revealed the presence of...   \n",
       "6  The characterization of the KCNJ9 gene should ...   \n",
       "6  The characterization of the KCNJ9 gene should ...   \n",
       "6  The characterization of the KCNJ9 gene should ...   \n",
       "6  The characterization of the KCNJ9 gene should ...   \n",
       "\n",
       "                                   entities_clinical entities_clinical_class  \\\n",
       "0                               The human KCNJ9 (Kir                    TEST   \n",
       "0  the G-protein-activated inwardly rectifying po...                 PROBLEM   \n",
       "1                                    the KCNJ9 locus                    TEST   \n",
       "1      a candidate gene forType II diabetes mellitus                 PROBLEM   \n",
       "2                                                NaN                     NaN   \n",
       "3            single nucleotide polymorphisms (SNPs),                 PROBLEM   \n",
       "3                            aVal366Ala substitution                 PROBLEM   \n",
       "3                                          base-pair                    TEST   \n",
       "4                                                NaN                     NaN   \n",
       "5                              Ourexpression studies                    TEST   \n",
       "6             The characterization of the KCNJ9 gene                    TEST   \n",
       "6  furtherstudies on the function of the KCNJ9 pr...                    TEST   \n",
       "6                                         evaluation                    TEST   \n",
       "6                                   Type II diabetes                 PROBLEM   \n",
       "\n",
       "  entities_clinical_confidence  \\\n",
       "0                      0.56672   \n",
       "0                      0.56248   \n",
       "1                   0.43453333   \n",
       "1                    0.7657142   \n",
       "2                          NaN   \n",
       "3                   0.56495005   \n",
       "3                      0.66485   \n",
       "3                       0.5886   \n",
       "4                          NaN   \n",
       "5                       0.9911   \n",
       "6                   0.65316665   \n",
       "6                   0.77554995   \n",
       "6                       0.7549   \n",
       "6                       0.9324   \n",
       "\n",
       "                              word_embedding_biobert  \n",
       "0  [[-0.248008593916893, 0.17330053448677063, 0.1...  \n",
       "0  [[-0.248008593916893, 0.17330053448677063, 0.1...  \n",
       "1  [[0.01339302584528923, 0.005851167719811201, 0...  \n",
       "1  [[0.01339302584528923, 0.005851167719811201, 0...  \n",
       "2  [[-0.542584240436554, 0.050951749086380005, -0...  \n",
       "3  [[0.08011732995510101, -0.11200879514217377, -...  \n",
       "3  [[0.08011732995510101, -0.11200879514217377, -...  \n",
       "3  [[0.08011732995510101, -0.11200879514217377, -...  \n",
       "4  [[-0.025541141629219055, -0.29619351029396057,...  \n",
       "5  [[0.4464523494243622, -0.1416597217321396, 0.5...  \n",
       "6  [[-0.5280818939208984, 0.03617645055055618, 0....  \n",
       "6  [[-0.5280818939208984, 0.03617645055055618, 0....  \n",
       "6  [[-0.5280818939208984, 0.03617645055055618, 0....  \n",
       "6  [[-0.5280818939208984, 0.03617645055055618, 0....  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nlu.load(\"en.med_ner.clinical.biobert en.assert.biobert\").predict(['The human KCNJ9 (Kir 3.3, GIRK3) is a member of the G-protein-activated inwardly rectifying potassium (GIRK) channel family.',\n",
    "'Here we describe the genomicorganization of the KCNJ9 locus on chromosome 1q21-23 as a candidate gene forType II diabetes mellitus in the Pima Indian population.', \n",
    "'The gene spansapproximately 7.6 kb and contains one noncoding and two coding exons separated byapproximately 2.2 and approximately 2.6 kb introns, respectively.', \n",
    "'We identified14 single nucleotide polymorphisms (SNPs), including one that predicts aVal366Ala substitution, and an 8 base-pair', '(bp) insertion/deletion.', \n",
    "'Ourexpression studies revealed the presence of the transcript in various humantissues including pancreas, and two major insulin-responsive tissues: fat andskeletal muscle.', \n",
    "'The characterization of the KCNJ9 gene should facilitate furtherstudies on the function of the KCNJ9 protein and allow evaluation of thepotential role of the locus in Type II diabetes.'],output_level =  \"chunk\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "ReXrvJhH6TZA"
   },
   "source": [
    "#### [Detect Assertion Status (assertion_dl_healthcare)](https://nlp.johnsnowlabs.com/2020/09/23/assertion_dl_healthcare_en.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 863
    },
    "executionInfo": {
     "elapsed": 65718,
     "status": "ok",
     "timestamp": 1649994852918,
     "user": {
      "displayName": "ahmed lone",
      "userId": "02458088882398909889"
     },
     "user_tz": -300
    },
    "id": "vaUFr-RR6TZC",
    "outputId": "f072b52c-eb59-4918-a9da-66e44a6b0f42"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ner_clinical download started this may take some time.\n",
      "Approximate size to download 13.9 MB\n",
      "[OK!]\n",
      "assertion_dl_healthcare download started this may take some time.\n",
      "Approximate size to download 1 MB\n",
      "[OK!]\n",
      "embeddings_clinical download started this may take some time.\n",
      "Approximate size to download 1.6 GB\n",
      "[OK!]\n",
      "embeddings_healthcare_100d download started this may take some time.\n",
      "Approximate size to download 475.8 MB\n",
      "[OK!]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "  <div id=\"df-fb535000-7513-4b70-adc1-f53837f35c4c\">\n",
       "    <div class=\"colab-df-container\">\n",
       "      <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assertion</th>\n",
       "      <th>document</th>\n",
       "      <th>entities_clinical</th>\n",
       "      <th>entities_clinical_class</th>\n",
       "      <th>entities_clinical_confidence</th>\n",
       "      <th>word_embedding_glove_clinical</th>\n",
       "      <th>word_embedding_glove_healthcare_100d</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>present</td>\n",
       "      <td>Patient has a headache for the last 2 weeks an...</td>\n",
       "      <td>a headache</td>\n",
       "      <td>PROBLEM</td>\n",
       "      <td>0.6488</td>\n",
       "      <td>[[-0.08453157544136047, 0.20632991194725037, -...</td>\n",
       "      <td>[[-0.09624659270048141, 0.3514532148838043, -0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>present</td>\n",
       "      <td>Patient has a headache for the last 2 weeks an...</td>\n",
       "      <td>anxious</td>\n",
       "      <td>PROBLEM</td>\n",
       "      <td>0.9999</td>\n",
       "      <td>[[-0.08453157544136047, 0.20632991194725037, -...</td>\n",
       "      <td>[[-0.09624659270048141, 0.3514532148838043, -0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>present</td>\n",
       "      <td>Patient has a headache for the last 2 weeks an...</td>\n",
       "      <td>alopecia</td>\n",
       "      <td>PROBLEM</td>\n",
       "      <td>0.9893</td>\n",
       "      <td>[[-0.08453157544136047, 0.20632991194725037, -...</td>\n",
       "      <td>[[-0.09624659270048141, 0.3514532148838043, -0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>present</td>\n",
       "      <td>Patient has a headache for the last 2 weeks an...</td>\n",
       "      <td>pain</td>\n",
       "      <td>PROBLEM</td>\n",
       "      <td>0.9774</td>\n",
       "      <td>[[-0.08453157544136047, 0.20632991194725037, -...</td>\n",
       "      <td>[[-0.09624659270048141, 0.3514532148838043, -0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-fb535000-7513-4b70-adc1-f53837f35c4c')\"\n",
       "              title=\"Convert this dataframe to an interactive table.\"\n",
       "              style=\"display:none;\">\n",
       "        \n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "       width=\"24px\">\n",
       "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
       "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
       "  </svg>\n",
       "      </button>\n",
       "      \n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      flex-wrap:wrap;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "      <script>\n",
       "        const buttonEl =\n",
       "          document.querySelector('#df-fb535000-7513-4b70-adc1-f53837f35c4c button.colab-df-convert');\n",
       "        buttonEl.style.display =\n",
       "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "        async function convertToInteractive(key) {\n",
       "          const element = document.querySelector('#df-fb535000-7513-4b70-adc1-f53837f35c4c');\n",
       "          const dataTable =\n",
       "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                     [key], {});\n",
       "          if (!dataTable) return;\n",
       "\n",
       "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "            + ' to learn more about interactive tables.';\n",
       "          element.innerHTML = '';\n",
       "          dataTable['output_type'] = 'display_data';\n",
       "          await google.colab.output.renderOutput(dataTable, element);\n",
       "          const docLink = document.createElement('div');\n",
       "          docLink.innerHTML = docLinkHtml;\n",
       "          element.appendChild(docLink);\n",
       "        }\n",
       "      </script>\n",
       "    </div>\n",
       "  </div>\n",
       "  "
      ],
      "text/plain": [
       "  assertion                                           document  \\\n",
       "0   present  Patient has a headache for the last 2 weeks an...   \n",
       "0   present  Patient has a headache for the last 2 weeks an...   \n",
       "0   present  Patient has a headache for the last 2 weeks an...   \n",
       "0   present  Patient has a headache for the last 2 weeks an...   \n",
       "\n",
       "  entities_clinical entities_clinical_class entities_clinical_confidence  \\\n",
       "0        a headache                 PROBLEM                       0.6488   \n",
       "0           anxious                 PROBLEM                       0.9999   \n",
       "0          alopecia                 PROBLEM                       0.9893   \n",
       "0              pain                 PROBLEM                       0.9774   \n",
       "\n",
       "                       word_embedding_glove_clinical  \\\n",
       "0  [[-0.08453157544136047, 0.20632991194725037, -...   \n",
       "0  [[-0.08453157544136047, 0.20632991194725037, -...   \n",
       "0  [[-0.08453157544136047, 0.20632991194725037, -...   \n",
       "0  [[-0.08453157544136047, 0.20632991194725037, -...   \n",
       "\n",
       "                word_embedding_glove_healthcare_100d  \n",
       "0  [[-0.09624659270048141, 0.3514532148838043, -0...  \n",
       "0  [[-0.09624659270048141, 0.3514532148838043, -0...  \n",
       "0  [[-0.09624659270048141, 0.3514532148838043, -0...  \n",
       "0  [[-0.09624659270048141, 0.3514532148838043, -0...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nlu.load(\"en.med_ner.clinical en.assert.healthcare\").predict(['Patient has a headache for the last 2 weeks and appears anxious when she walks fast. No alopecia noted. She denies pain'],output_level =  \"chunk\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "VWzdXb8t6Tia"
   },
   "source": [
    "#### [Detect Assertion Status (DL Large)](https://nlp.johnsnowlabs.com/2020/05/21/assertion_dl_large_en.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 737
    },
    "executionInfo": {
     "elapsed": 19260,
     "status": "ok",
     "timestamp": 1649994872152,
     "user": {
      "displayName": "ahmed lone",
      "userId": "02458088882398909889"
     },
     "user_tz": -300
    },
    "id": "pKEpS_Tm6Tib",
    "outputId": "ade0ae16-2365-481d-af37-fd7f22e0c8cf"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ner_clinical download started this may take some time.\n",
      "Approximate size to download 13.9 MB\n",
      "[OK!]\n",
      "assertion_dl_large download started this may take some time.\n",
      "Approximate size to download 1.3 MB\n",
      "[OK!]\n",
      "embeddings_clinical download started this may take some time.\n",
      "Approximate size to download 1.6 GB\n",
      "[OK!]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "  <div id=\"df-fa4c4088-2e2f-462f-a8cf-3fdfbac1ba3f\">\n",
       "    <div class=\"colab-df-container\">\n",
       "      <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assertion</th>\n",
       "      <th>document</th>\n",
       "      <th>entities_clinical</th>\n",
       "      <th>entities_clinical_class</th>\n",
       "      <th>entities_clinical_confidence</th>\n",
       "      <th>word_embedding_glove</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>present</td>\n",
       "      <td>Patient with severe fever and sore throat. He ...</td>\n",
       "      <td>severe fever</td>\n",
       "      <td>PROBLEM</td>\n",
       "      <td>0.8215</td>\n",
       "      <td>[[-0.08453157544136047, 0.20632991194725037, -...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>present</td>\n",
       "      <td>Patient with severe fever and sore throat. He ...</td>\n",
       "      <td>sore throat</td>\n",
       "      <td>PROBLEM</td>\n",
       "      <td>0.99605</td>\n",
       "      <td>[[-0.08453157544136047, 0.20632991194725037, -...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>absent</td>\n",
       "      <td>Patient with severe fever and sore throat. He ...</td>\n",
       "      <td>stomach pain</td>\n",
       "      <td>PROBLEM</td>\n",
       "      <td>0.99705005</td>\n",
       "      <td>[[-0.08453157544136047, 0.20632991194725037, -...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>present</td>\n",
       "      <td>Patient with severe fever and sore throat. He ...</td>\n",
       "      <td>an epidural</td>\n",
       "      <td>TREATMENT</td>\n",
       "      <td>0.9779</td>\n",
       "      <td>[[-0.08453157544136047, 0.20632991194725037, -...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>present</td>\n",
       "      <td>Patient with severe fever and sore throat. He ...</td>\n",
       "      <td>PCA</td>\n",
       "      <td>TREATMENT</td>\n",
       "      <td>0.9994</td>\n",
       "      <td>[[-0.08453157544136047, 0.20632991194725037, -...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>present</td>\n",
       "      <td>Patient with severe fever and sore throat. He ...</td>\n",
       "      <td>pain control</td>\n",
       "      <td>PROBLEM</td>\n",
       "      <td>0.84575</td>\n",
       "      <td>[[-0.08453157544136047, 0.20632991194725037, -...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>conditional</td>\n",
       "      <td>Patient with severe fever and sore throat. He ...</td>\n",
       "      <td>short of breath</td>\n",
       "      <td>PROBLEM</td>\n",
       "      <td>0.93709993</td>\n",
       "      <td>[[-0.08453157544136047, 0.20632991194725037, -...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>present</td>\n",
       "      <td>Patient with severe fever and sore throat. He ...</td>\n",
       "      <td>CT</td>\n",
       "      <td>TEST</td>\n",
       "      <td>0.9995</td>\n",
       "      <td>[[-0.08453157544136047, 0.20632991194725037, -...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>present</td>\n",
       "      <td>Patient with severe fever and sore throat. He ...</td>\n",
       "      <td>lung tumor</td>\n",
       "      <td>PROBLEM</td>\n",
       "      <td>0.67785</td>\n",
       "      <td>[[-0.08453157544136047, 0.20632991194725037, -...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>present</td>\n",
       "      <td>Patient with severe fever and sore throat. He ...</td>\n",
       "      <td>Alzheimer</td>\n",
       "      <td>PROBLEM</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[[-0.08453157544136047, 0.20632991194725037, -...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-fa4c4088-2e2f-462f-a8cf-3fdfbac1ba3f')\"\n",
       "              title=\"Convert this dataframe to an interactive table.\"\n",
       "              style=\"display:none;\">\n",
       "        \n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "       width=\"24px\">\n",
       "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
       "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
       "  </svg>\n",
       "      </button>\n",
       "      \n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      flex-wrap:wrap;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "      <script>\n",
       "        const buttonEl =\n",
       "          document.querySelector('#df-fa4c4088-2e2f-462f-a8cf-3fdfbac1ba3f button.colab-df-convert');\n",
       "        buttonEl.style.display =\n",
       "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "        async function convertToInteractive(key) {\n",
       "          const element = document.querySelector('#df-fa4c4088-2e2f-462f-a8cf-3fdfbac1ba3f');\n",
       "          const dataTable =\n",
       "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                     [key], {});\n",
       "          if (!dataTable) return;\n",
       "\n",
       "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "            + ' to learn more about interactive tables.';\n",
       "          element.innerHTML = '';\n",
       "          dataTable['output_type'] = 'display_data';\n",
       "          await google.colab.output.renderOutput(dataTable, element);\n",
       "          const docLink = document.createElement('div');\n",
       "          docLink.innerHTML = docLinkHtml;\n",
       "          element.appendChild(docLink);\n",
       "        }\n",
       "      </script>\n",
       "    </div>\n",
       "  </div>\n",
       "  "
      ],
      "text/plain": [
       "     assertion                                           document  \\\n",
       "0      present  Patient with severe fever and sore throat. He ...   \n",
       "0      present  Patient with severe fever and sore throat. He ...   \n",
       "0       absent  Patient with severe fever and sore throat. He ...   \n",
       "0      present  Patient with severe fever and sore throat. He ...   \n",
       "0      present  Patient with severe fever and sore throat. He ...   \n",
       "0      present  Patient with severe fever and sore throat. He ...   \n",
       "0  conditional  Patient with severe fever and sore throat. He ...   \n",
       "0      present  Patient with severe fever and sore throat. He ...   \n",
       "0      present  Patient with severe fever and sore throat. He ...   \n",
       "0      present  Patient with severe fever and sore throat. He ...   \n",
       "\n",
       "  entities_clinical entities_clinical_class entities_clinical_confidence  \\\n",
       "0      severe fever                 PROBLEM                       0.8215   \n",
       "0       sore throat                 PROBLEM                      0.99605   \n",
       "0      stomach pain                 PROBLEM                   0.99705005   \n",
       "0       an epidural               TREATMENT                       0.9779   \n",
       "0               PCA               TREATMENT                       0.9994   \n",
       "0      pain control                 PROBLEM                      0.84575   \n",
       "0   short of breath                 PROBLEM                   0.93709993   \n",
       "0                CT                    TEST                       0.9995   \n",
       "0        lung tumor                 PROBLEM                      0.67785   \n",
       "0         Alzheimer                 PROBLEM                          1.0   \n",
       "\n",
       "                                word_embedding_glove  \n",
       "0  [[-0.08453157544136047, 0.20632991194725037, -...  \n",
       "0  [[-0.08453157544136047, 0.20632991194725037, -...  \n",
       "0  [[-0.08453157544136047, 0.20632991194725037, -...  \n",
       "0  [[-0.08453157544136047, 0.20632991194725037, -...  \n",
       "0  [[-0.08453157544136047, 0.20632991194725037, -...  \n",
       "0  [[-0.08453157544136047, 0.20632991194725037, -...  \n",
       "0  [[-0.08453157544136047, 0.20632991194725037, -...  \n",
       "0  [[-0.08453157544136047, 0.20632991194725037, -...  \n",
       "0  [[-0.08453157544136047, 0.20632991194725037, -...  \n",
       "0  [[-0.08453157544136047, 0.20632991194725037, -...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nlu.load(\"en.med_ner.clinical en.assert.large\").predict([\"\"\"Patient with severe fever and sore throat. He shows no stomach pain and he maintained on an epidural \n",
    "and PCA for pain control. He also became short of breath with climbing a flight of stairs. After CT, lung tumor located at the right lower lobe. Father with Alzheimer.\"\"\"],output_level =  \"chunk\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 289
    },
    "executionInfo": {
     "elapsed": 234863,
     "status": "ok",
     "timestamp": 1649995106902,
     "user": {
      "displayName": "ahmed lone",
      "userId": "02458088882398909889"
     },
     "user_tz": -300
    },
    "id": "6dVUUaOMyNSt",
    "outputId": "10badfca-792a-41b3-9302-51769a45afcd"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ner_wikiner_glove_840B_300 download started this may take some time.\n",
      "Approximate size to download 14.8 MB\n",
      "[OK!]\n",
      "assertion_dl_large download started this may take some time.\n",
      "Approximate size to download 1.3 MB\n",
      "[OK!]\n",
      "embeddings_clinical download started this may take some time.\n",
      "Approximate size to download 1.6 GB\n",
      "[OK!]\n",
      "glove_840B_300 download started this may take some time.\n",
      "Approximate size to download 2.3 GB\n",
      "[OK!]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "  <div id=\"df-9645257b-9c42-4d92-a5b8-c1280e2374a2\">\n",
       "    <div class=\"colab-df-container\">\n",
       "      <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>document</th>\n",
       "      <th>word_embedding_glove_840B_300</th>\n",
       "      <th>word_embedding_glove_clinical</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>He was cured of cancer! But is cough is still ...</td>\n",
       "      <td>[[0.02472599968314171, 0.273499995470047, 0.06...</td>\n",
       "      <td>[[-0.00917550828307867, -0.16478723287582397, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-9645257b-9c42-4d92-a5b8-c1280e2374a2')\"\n",
       "              title=\"Convert this dataframe to an interactive table.\"\n",
       "              style=\"display:none;\">\n",
       "        \n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "       width=\"24px\">\n",
       "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
       "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
       "  </svg>\n",
       "      </button>\n",
       "      \n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      flex-wrap:wrap;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "      <script>\n",
       "        const buttonEl =\n",
       "          document.querySelector('#df-9645257b-9c42-4d92-a5b8-c1280e2374a2 button.colab-df-convert');\n",
       "        buttonEl.style.display =\n",
       "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "        async function convertToInteractive(key) {\n",
       "          const element = document.querySelector('#df-9645257b-9c42-4d92-a5b8-c1280e2374a2');\n",
       "          const dataTable =\n",
       "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                     [key], {});\n",
       "          if (!dataTable) return;\n",
       "\n",
       "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "            + ' to learn more about interactive tables.';\n",
       "          element.innerHTML = '';\n",
       "          dataTable['output_type'] = 'display_data';\n",
       "          await google.colab.output.renderOutput(dataTable, element);\n",
       "          const docLink = document.createElement('div');\n",
       "          docLink.innerHTML = docLinkHtml;\n",
       "          element.appendChild(docLink);\n",
       "        }\n",
       "      </script>\n",
       "    </div>\n",
       "  </div>\n",
       "  "
      ],
      "text/plain": [
       "                                            document  \\\n",
       "0  He was cured of cancer! But is cough is still ...   \n",
       "\n",
       "                       word_embedding_glove_840B_300  \\\n",
       "0  [[0.02472599968314171, 0.273499995470047, 0.06...   \n",
       "\n",
       "                       word_embedding_glove_clinical  \n",
       "0  [[-0.00917550828307867, -0.16478723287582397, ...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = nlu.load(\"med_ner.clinical assert.large\").predict('He was cured of cancer! But is cough is still there',output_level =  \"chunk\")\n",
    "df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "UzASczdVoJxD"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "name": "assertion_overview.ipynb",
   "provenance": [
    {
     "file_id": "1pDt15LVNN_2GdpjDqFKy7cuxedJReSQJ",
     "timestamp": 1618841320853
    }
   ],
   "toc_visible": true
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
